// @apps-in-toss/framework/dist/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));

// src/index.ts
import { Analytics as InternalAnalytics } from "@apps-in-toss/analytics";

// src/core/registerApp.tsx
import { Analytics } from "@apps-in-toss/analytics";
import { isMinVersionSupported as isMinVersionSupported3, setIosSwipeGestureEnabled as setIosSwipeGestureEnabled2, eventLog as eventLog2, getSchemeUri as getSchemeUri5 } from "@apps-in-toss/native-modules";
import { Granite as Granite6 } from "@granite-js/react-native";
import { TDSProvider } from "@toss/tds-react-native";
import { AppRegistry } from "react-native";

// src/core/components/AppEvent.tsx
import { Granite, getSchemeUri as getSchemeUri2 } from "@granite-js/react-native";
import { useEffect } from "react";

// src/env.ts
var env = {
  getDeploymentId: () => __DEV__ ? "local" : global.__appsInToss?.deploymentId,
  getWebViewType: () => global.__appsInToss.webViewType,
  getAppName: () => global.__granite.app.name
};

// src/utils/isPrivateScheme.ts
import { getSchemeUri } from "@granite-js/react-native";
function isPrivateScheme() {
  try {
    return new URL(getSchemeUri()).protocol === "intoss-private:";
  } catch {
    return false;
  }
}

// src/utils/sendLog.ts
import { INTERNAL__module, eventLog } from "@apps-in-toss/native-modules";
var DEFAULT_PIPELINES = ["core", "partner"];
function sendLog(logData, options) {
  const { pipelines = DEFAULT_PIPELINES } = options ?? {};
  if (pipelines.includes("core")) {
    INTERNAL__module.tossCoreEventLog(logData);
  }
  if (pipelines.includes("partner")) {
    eventLog(logData);
  }
}

// src/core/components/AppEvent.tsx
function SystemAppEvent({ ...initialProps }) {
  useEffect(() => {
    sendLog(
      {
        log_name: "AppsInTossInitialProps",
        log_type: "debug",
        params: {
          ...initialProps,
          schemeUri: getSchemeUri2(),
          deployment_id: env.getDeploymentId(),
          app_name: Granite.appName,
          is_private: isPrivateScheme()
        }
      },
      { pipelines: ["core"] }
    );
  }, [initialProps]);
  return null;
}
var AppEvent = {
  System: SystemAppEvent
};

// src/core/components/AppUpdate.tsx
import { openURL } from "@granite-js/react-native";
import { useEffect as useEffect2 } from "react";
import { Fragment, jsx } from "react/jsx-runtime";
var UPDATE_SCHEME = "servicetoss://update/forced/j?updateType=required";
function AppUpdate() {
  useEffect2(() => {
    openURL(`supertoss://closePage?url=${UPDATE_SCHEME}`);
  }, []);
  return /* @__PURE__ */ jsx(Fragment, {});
}

// src/core/context/NavigationBarContext.tsx
import { createContext, useContext, useState } from "react";

// src/core/utils/getAppsInTossGlobals.ts
function getAppsInTossGlobals() {
  if (global.__appsInToss == null) {
    throw new Error("invalid apps-in-toss globals");
  }
  return global.__appsInToss;
}

// src/core/utils/safeParseNavigationBar.ts
function safeParseNavigationBar(navigationBar) {
  if (typeof navigationBar === "object") {
    return navigationBar;
  }
  try {
    return JSON.parse(navigationBar);
  } catch {
    return null;
  }
}

// src/core/context/NavigationBarContext.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
var NavigationBarContext = createContext(null);
function NavigationBarContextProvider({ children }) {
  const globals = getAppsInTossGlobals();
  const parsedNavigationBar = globals.navigationBar != null ? safeParseNavigationBar(globals.navigationBar) : null;
  const [navigationRightButton, setNavigationRightButton] = useState(
    parsedNavigationBar?.initialAccessoryButton
  );
  const addNavigationRightButton = (button) => {
    setNavigationRightButton(button);
  };
  const removeNavigationRightButton = () => {
    setNavigationRightButton(void 0);
  };
  return /* @__PURE__ */ jsx2(
    NavigationBarContext.Provider,
    {
      value: { navigationRightButton, addNavigationRightButton, removeNavigationRightButton },
      children
    }
  );
}
function useNavigationBarContext() {
  const context = useContext(NavigationBarContext);
  if (context == null) {
    throw Error("useNavigationBarContext must be used within NavigationBarContextProvider");
  }
  return context;
}

// src/components/RNAppContainer.tsx
import { setIosSwipeGestureEnabled } from "@apps-in-toss/native-modules";
import { useEffect as useEffect7 } from "react";
import { Platform as Platform3 } from "react-native";

// src/components/GameInitializer.tsx
import { isMinVersionSupported } from "@apps-in-toss/native-modules";
import { closeView as closeView2 } from "@granite-js/react-native";
import { Loader } from "@toss/tds-react-native";
import { josa } from "es-hangul";
import { useEffect as useEffect3, useRef } from "react";
import { Pressable, View } from "react-native";

// src/constant/game-center.ts
var GAME_PROFILE_WEBVIEW_URL = "servicetoss://game-center/profile";
var GAME_MIN_VERSION = {
  android: "5.232.0",
  ios: "5.232.0"
};

// src/hooks/useAppUpdateDialog.tsx
import { Granite as Granite2, openURL as openURL2 } from "@granite-js/react-native";
import { useDialog } from "@toss/tds-react-native";
import { useCallback } from "react";

// src/core/hooks/useReferrer.ts
import { getSchemeUri as getSchemeUri3 } from "@granite-js/react-native";
import { useMemo } from "react";
function useReferrer() {
  return useMemo(() => {
    try {
      return new URL(getSchemeUri3()).searchParams.get("referrer");
    } catch {
      return null;
    }
  }, []);
}

// src/utils/market.ts
import { Platform } from "react-native";
var PLAYSTORE_LINK = "https://play.google.com/store/apps/details?id=viva.republica.toss";
var APPSTORE_LINK = "https://itunes.apple.com/app/id839333328";
var getMarketLink = () => {
  return Platform.OS === "android" ? PLAYSTORE_LINK : APPSTORE_LINK;
};

// src/hooks/useAppUpdateDialog.tsx
function useAppUpdateDialog() {
  const { openConfirm } = useDialog();
  const logging = useAppUpdateDialogLogging();
  const openAppUpdateDialog = useCallback(
    async ({
      title,
      description,
      leftButton = "\uB2EB\uAE30",
      rightButton = "\uC5C5\uB370\uC774\uD2B8\uD558\uAE30",
      onDismiss
    }) => {
      logging.show();
      const isConfirmed = await openConfirm({
        title,
        description,
        leftButton,
        rightButton,
        closeOnDimmerClick: true
      });
      if (!isConfirmed) {
        logging.close();
        onDismiss?.();
        return;
      }
      logging.update();
      const STORE_SCHEME = getMarketLink();
      openURL2(`supertoss://web?url=${STORE_SCHEME}&external=browser`);
    },
    [logging, openConfirm]
  );
  return {
    open: openAppUpdateDialog
  };
}
var UPDATE_DIALOG_SCHEMA_ID = 1634992;
var UPDATE_DIALOG_LOG_NAME = "appsintoss_app_visit__common_module::bottomsheet__app_update";
var UPDATE_DIALOG_CTA_CLICK_SCHEMA_ID = 1634996;
var UPDATE_DIALOG_CTA_CLICK_LOG_NAME = "appsintoss_app_visit__common_module::bottomsheet__app_update::click__cta";
function useAppUpdateDialogLogging() {
  const referrer = useReferrer();
  const baseParams = {
    referrer,
    app_name: Granite2.appName
  };
  const logUpdateClick = () => {
    sendLog({
      log_name: UPDATE_DIALOG_CTA_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        schema_id: UPDATE_DIALOG_CTA_CLICK_SCHEMA_ID,
        event_type: "click",
        button_type: "update"
      }
    });
  };
  const logCloseClick = () => {
    sendLog({
      log_name: UPDATE_DIALOG_CTA_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        schema_id: UPDATE_DIALOG_CTA_CLICK_SCHEMA_ID,
        event_type: "click",
        button_type: "close"
      }
    });
  };
  const logDialogShow = () => {
    sendLog({
      log_name: UPDATE_DIALOG_LOG_NAME,
      log_type: "popup",
      params: {
        ...baseParams,
        schema_id: UPDATE_DIALOG_SCHEMA_ID
      }
    });
  };
  return { update: logUpdateClick, close: logCloseClick, show: logDialogShow };
}

// src/hooks/useGameCenterProfile.ts
import { getGameCenterGameProfile } from "@apps-in-toss/native-modules";
import { useCallback as useCallback5, useState as useState3 } from "react";

// src/hooks/useErrorAlert.ts
import { closeView } from "@granite-js/react-native";
import { useDialog as useDialog2 } from "@toss/tds-react-native";
import { useCallback as useCallback2 } from "react";

// src/utils/error.ts
var DEFAULT_ERROR = {
  title: "\uC7A0\uC2DC \uD6C4 \uB2E4\uC2DC \uC2DC\uB3C4\uD574\uC8FC\uC138\uC694",
  description: "\uBB38\uC81C\uAC00 \uACC4\uC18D\uB418\uBA74 \uD1A0\uC2A4 \uACE0\uAC1D\uC13C\uD130(1599-4905)\uB85C \uBB38\uC758\uD574\uC8FC\uC138\uC694."
};

// src/hooks/useErrorAlert.ts
var useErrorAlert = () => {
  const { openAlert } = useDialog2();
  const openErrorAlert = useCallback2(async () => {
    await openAlert({
      title: DEFAULT_ERROR.title,
      description: DEFAULT_ERROR.description
    });
    closeView();
  }, [openAlert]);
  return { open: openErrorAlert };
};

// src/hooks/useTransparentWebview.tsx
import { useCallback as useCallback3, useState as useState2 } from "react";

// src/utils/openTransparentWebView.ts
import { openURL as openURL3 } from "@granite-js/react-native";

// src/private.ts
import { onVisibilityChangedByTransparentServiceWeb } from "@apps-in-toss/native-modules";
var INTERNAL__onVisibilityChangedByTransparentServiceWeb = onVisibilityChangedByTransparentServiceWeb;

// src/utils/openTransparentWebView.ts
var openTransparentWebView = ({
  webUrl,
  cleanupWhenDismissed = true,
  onEvent,
  onError,
  callbackId = "fn",
  params
}) => {
  const url = new URL("supertoss://transparent-service-web");
  url.searchParams.set("url", webUrl);
  url.searchParams.set("onVisibilityChangeCallback", callbackId);
  Object.entries(params ?? {}).forEach(([key, value]) => {
    url.searchParams.set(key, value);
  });
  const cleanup = INTERNAL__onVisibilityChangedByTransparentServiceWeb({
    options: { callbackId },
    onError: (error) => {
      onError(error);
      cleanup();
    },
    onEvent: (value) => {
      onEvent(value);
      if (cleanupWhenDismissed && value === true) {
        cleanup();
      }
    }
  });
  openURL3(url.toString());
};

// src/hooks/useTransparentWebview.tsx
var useTransparentWebview = () => {
  const [isWebviewLoading, setIsWebviewLoading] = useState2(false);
  const { open: openErrorAlert } = useErrorAlert();
  const _openTransparentWebview = useCallback3(
    ({ webUrl, onClose, onError }) => {
      if (isWebviewLoading) {
        return;
      }
      setIsWebviewLoading(true);
      openTransparentWebView({
        webUrl,
        onEvent: async (isClosedTransparentWebView) => {
          if (isClosedTransparentWebView) {
            setIsWebviewLoading(false);
            onClose?.();
          }
        },
        onError: (error) => {
          setIsWebviewLoading(false);
          if (onError) {
            onError(error);
          } else {
            openErrorAlert();
          }
        }
      });
    },
    [isWebviewLoading, openErrorAlert]
  );
  return {
    open: _openTransparentWebview
  };
};

// src/components/GameProfileToast.tsx
import { Asset, Toast } from "@toss/tds-react-native";
import { AdaptiveColorProvider, ColorPreferenceProvider, useOverlay } from "@toss/tds-react-native/private";
import { useCallback as useCallback4 } from "react";
import { jsx as jsx3 } from "react/jsx-runtime";
var useGameProfileToast = () => {
  const overlay = useOverlay();
  const openGameProfileToast = useCallback4(
    (nickname, profileImageUri) => {
      return new Promise((resolve) => {
        overlay.open(({ isOpen, close, exit }) => {
          return /* @__PURE__ */ jsx3(ColorPreferenceProvider, { colorPreference: "dark", children: /* @__PURE__ */ jsx3(AdaptiveColorProvider, { children: /* @__PURE__ */ jsx3(
            Toast,
            {
              open: isOpen,
              onClose: () => {
                resolve();
                close();
              },
              onExited: exit,
              position: "top",
              text: `${nickname}\uB2D8 \uBC18\uAC00\uC6CC\uC694!`,
              icon: /* @__PURE__ */ jsx3(
                Asset.Image,
                {
                  style: { borderRadius: 64, overflow: "hidden" },
                  frameShape: Asset.frameShape.CleanW32,
                  source: { uri: profileImageUri }
                }
              )
            }
          ) }) });
        });
      });
    },
    [overlay]
  );
  return { openGameProfileToast };
};

// src/hooks/useGameCenterProfile.ts
var useGameCenterProfile = () => {
  const [profileData, setProfileData] = useState3(void 0);
  const [isProfileDataLoading, setIsProfileDataLoading] = useState3(true);
  const [isProfileDataRefetching, setIsProfileDataRefetching] = useState3(false);
  const shouldShowProfileLoadingOverlay = isProfileDataLoading || isProfileDataRefetching;
  const shouldShowProfileNotFoundOverlay = profileData?.statusCode === "PROFILE_NOT_FOUND" && !isProfileDataRefetching;
  const canShowBottomSheetOrToast = !isProfileDataLoading;
  const { openGameProfileToast } = useGameProfileToast();
  const { open: openErrorAlert } = useErrorAlert();
  const { open: openTransparentWebView2 } = useTransparentWebview();
  const fetchProfileData = useCallback5(async () => {
    try {
      const data = await getGameCenterGameProfile();
      setProfileData(data);
      setIsProfileDataLoading(false);
    } catch (_) {
      openErrorAlert();
      setIsProfileDataLoading(false);
    }
  }, [openErrorAlert]);
  const refetchProfileData = useCallback5(async () => {
    try {
      setIsProfileDataRefetching(true);
      const data = await getGameCenterGameProfile();
      setProfileData(data);
      setIsProfileDataRefetching(false);
      if (data?.statusCode === "SUCCESS") {
        openGameProfileToast(data.nickname, data.profileImageUri);
      }
    } catch (_) {
      setIsProfileDataRefetching(false);
      openErrorAlert();
    }
  }, [openErrorAlert, openGameProfileToast]);
  const openProfileWebview = useCallback5(() => {
    openTransparentWebView2({
      webUrl: `${GAME_PROFILE_WEBVIEW_URL}?appName=${getAppName()}&referrer=appsintoss.${getAppName()}`,
      onClose: async () => {
        refetchProfileData();
      }
    });
  }, [openTransparentWebView2, refetchProfileData]);
  return {
    profileData,
    isProfileDataLoading,
    shouldShowProfileLoadingOverlay,
    shouldShowProfileNotFoundOverlay,
    canShowBottomSheetOrToast,
    setIsProfileDataLoading,
    openProfileWebview,
    setProfileData,
    openErrorAlert,
    openGameProfileToast,
    fetchProfileData
  };
};

// src/utils/zIndex.ts
var Z_INDEX = {
  /* 게임 프로필을 위한 overlay
   */
  PROFILE_OVERLAY: 9998,
  // 게임을 종료할 수 있는 X 버튼
  CLOSE_BUTTON: 9999
};

// src/components/GameInitializer.tsx
import { Fragment as Fragment2, jsx as jsx4, jsxs } from "react/jsx-runtime";
var GameInitializer = ({ children }) => {
  const {
    profileData,
    shouldShowProfileLoadingOverlay,
    shouldShowProfileNotFoundOverlay,
    canShowBottomSheetOrToast,
    openProfileWebview,
    openGameProfileToast,
    fetchProfileData
  } = useGameCenterProfile();
  const isCompletedFlow = useRef(false);
  const { open: openAppUpdateDialog } = useAppUpdateDialog();
  useEffect3(() => {
    fetchProfileData();
  }, []);
  useEffect3(() => {
    const handleGameProfileFlow = async () => {
      if (!canShowBottomSheetOrToast) {
        return;
      }
      if (isCompletedFlow.current) {
        return;
      }
      isCompletedFlow.current = true;
      if (!isMinVersionSupported(GAME_MIN_VERSION)) {
        openAppUpdateDialog({
          title: `${josa(getAppsInTossGlobals().brandDisplayName, "\uC744/\uB97C")} \uD558\uB824\uBA74 
\uC571\uC744 \uC5C5\uB370\uC774\uD2B8\uD574\uC8FC\uC138\uC694`,
          onDismiss: () => {
            closeView2();
          }
        });
        return;
      }
      if (profileData?.statusCode === "SUCCESS") {
        openGameProfileToast(profileData.nickname, profileData.profileImageUri);
        return;
      }
      if (profileData?.statusCode === "PROFILE_NOT_FOUND") {
        openProfileWebview();
      }
    };
    handleGameProfileFlow();
  }, [
    canShowBottomSheetOrToast,
    isCompletedFlow,
    openAppUpdateDialog,
    openGameProfileToast,
    openProfileWebview,
    profileData
  ]);
  if (!isMinVersionSupported(GAME_MIN_VERSION)) {
    return /* @__PURE__ */ jsxs(Fragment2, { children: [
      /* @__PURE__ */ jsx4(View, { style: { flex: 1, position: "relative" }, children }),
      /* @__PURE__ */ jsx4(
        Pressable,
        {
          style: {
            ...overlayStyle
          },
          onPress: () => {
            openAppUpdateDialog({
              title: `${josa(getAppsInTossGlobals().brandDisplayName, "\uC744/\uB97C")} \uD558\uB824\uBA74 
\uC571\uC744 \uC5C5\uB370\uC774\uD2B8\uD574\uC8FC\uC138\uC694`,
              onDismiss: () => {
                closeView2();
              }
            });
          }
        }
      )
    ] });
  }
  if (shouldShowProfileLoadingOverlay) {
    return /* @__PURE__ */ jsxs(Fragment2, { children: [
      /* @__PURE__ */ jsx4(View, { style: { flex: 1, position: "relative" }, children }),
      /* @__PURE__ */ jsx4(
        View,
        {
          style: {
            ...overlayStyle,
            justifyContent: "center",
            alignItems: "center",
            backgroundColor: "rgba(0, 0, 0, 0.2)"
          },
          children: /* @__PURE__ */ jsx4(Loader, { size: "large", type: "light" })
        }
      )
    ] });
  }
  if (shouldShowProfileNotFoundOverlay) {
    return /* @__PURE__ */ jsxs(Fragment2, { children: [
      /* @__PURE__ */ jsx4(View, { style: { flex: 1, position: "relative" }, children }),
      shouldShowProfileNotFoundOverlay && /* @__PURE__ */ jsx4(
        Pressable,
        {
          style: {
            ...overlayStyle
          },
          onPress: () => {
            openProfileWebview();
          }
        }
      )
    ] });
  }
  return /* @__PURE__ */ jsx4(Fragment2, { children: /* @__PURE__ */ jsx4(View, { style: { flex: 1, position: "relative" }, children }) });
};
var overlayStyle = {
  position: "absolute",
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  zIndex: Z_INDEX.PROFILE_OVERLAY
};

// src/components/NavigationBar/RNNavigationBar/Default.tsx
import { NavigationBackButton, NavigationLeft, TopNavigation } from "@toss/tds-react-native/private";

// src/components/NavigationBar/RNNavigationBar/hooks/useHardwareBackPress.ts
import { useEffect as useEffect4 } from "react";
import { BackHandler } from "react-native";
function useHardwareBackPress(handler) {
  useEffect4(() => {
    const handleBackPress = () => {
      handler();
      return true;
    };
    const backHandler = BackHandler.addEventListener("hardwareBackPress", handleBackPress);
    return () => backHandler.remove();
  }, [handler]);
}

// src/components/NavigationBar/RNNavigationBar/hooks/useNavigationEvent.ts
import { closeView as closeView3, useBackEventContext, useNavigation } from "@granite-js/react-native";
import { useMemo as useMemo2 } from "react";

// src/components/NavigationBar/RNNavigationBar/hooks/useCloseConfirm.ts
import { useDialog as useDialog3 } from "@toss/tds-react-native";
import { josa as josa2 } from "es-hangul";
import { useCallback as useCallback6 } from "react";
function useCloseConfirm() {
  const { brandDisplayName } = getAppsInTossGlobals();
  const { openConfirm } = useDialog3();
  return useCallback6(
    async ({ onEntered }) => {
      return await openConfirm({
        title: `${josa2(brandDisplayName, "\uC744/\uB97C")} \uC885\uB8CC\uD560\uAE4C\uC694?`,
        leftButton: "\uB2EB\uAE30",
        rightButton: "\uC885\uB8CC\uD558\uAE30",
        closeOnDimmerClick: true,
        onEntered
      });
    },
    [brandDisplayName, openConfirm]
  );
}

// src/utils/eventEmitter.ts
var EventEmitter = class {
  listeners = {};
  subscribe(event, listener) {
    if (!this.listeners[event]) {
      this.listeners[event] = [];
    }
    this.listeners[event].push(listener);
    return () => {
      this.listeners[event] = this.listeners[event]?.filter((l) => l !== listener) ?? [];
    };
  }
  hasSubscriptions(event) {
    return this.listeners[event] != null && this.listeners[event].length > 0;
  }
  clearSubscriptions(event) {
    this.listeners[event] = [];
  }
  emit(event, ...args) {
    if (!this.listeners[event]) {
      return;
    }
    this.listeners[event].forEach((listener) => listener(...args));
  }
};
var eventEmitter = new EventEmitter();
function createEvent(event) {
  return {
    name: event,
    subscribe: (listener) => eventEmitter.subscribe(event, listener),
    hasSubscriptions: () => eventEmitter.hasSubscriptions(event),
    clearSubscriptions: () => eventEmitter.clearSubscriptions(event),
    emit: (data) => eventEmitter.emit(event, data)
  };
}

// src/events.ts
var homeEvent = createEvent("homeEvent");
var safeAreaInsetsChange = createEvent("safeAreaInsetsChange");

// src/components/NavigationBar/common/useNavigationBarLogging.tsx
import { Granite as Granite3 } from "@granite-js/react-native";
var NAVI_BAR_IMPRESSION_SCHEMA_ID = 1596837;
var NAVI_BAR_IMPRESSION_LOG_NAME = "appsintoss_app_visit__common_module::impression__navigation_bar";
var CLOSE_POPUP_SHOW_SCHEMA_ID = 1644490;
var CLOSE_POPUP_SHOW_LOG_NAME = "appsintoss_app_visit__common_module::popup__close_app";
var CLOSE_BUTTON_CLICK_SCHEMA_ID = 1596831;
var CLOSE_BUTTON_CLICK_LOG_NAME = "appsintoss_app_visit__common_module::click__icon_close";
var CLOSE_POPUP_CTA_CLICK_SCHEMA_ID = 1644492;
var CLOSE_POPUP_CTA_CLICK_LOG_NAME = "appsintoss_app_visit__common_module::popup__close_app::click__cta";
var HOME_BUTTON_CLICK_SCHEMA_ID = 1596839;
var HOME_BUTTON_CLICK_LOG_NAME = "appsintoss_app_visit__common_module::click__icon_home";
function useNavigationBarLogging() {
  const referrer = useReferrer();
  const baseParams = {
    referrer,
    app_name: Granite3.appName
  };
  const logNavBarImpression = (naviBarConfig) => {
    sendLog({
      log_name: NAVI_BAR_IMPRESSION_LOG_NAME,
      log_type: "event",
      params: {
        ...naviBarConfig,
        ...baseParams,
        event_type: "impression",
        schema_id: NAVI_BAR_IMPRESSION_SCHEMA_ID
      }
    });
  };
  const logHomeButtonClick = () => {
    sendLog({
      log_name: HOME_BUTTON_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        event_type: "click",
        schema_id: HOME_BUTTON_CLICK_SCHEMA_ID
      }
    });
  };
  const logCloseButtonClick = () => {
    sendLog({
      log_name: CLOSE_BUTTON_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        event_type: "click",
        schema_id: CLOSE_BUTTON_CLICK_SCHEMA_ID
      }
    });
  };
  const logClosePopupShow = () => {
    sendLog({
      log_name: CLOSE_POPUP_SHOW_LOG_NAME,
      log_type: "popup",
      params: {
        ...baseParams,
        schema_id: CLOSE_POPUP_SHOW_SCHEMA_ID
      }
    });
  };
  const logClosePopupCtaClick = (confirm) => {
    sendLog({
      log_name: CLOSE_POPUP_CTA_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        close_yn: confirm ? "Y" : "N",
        schema_id: CLOSE_POPUP_CTA_CLICK_SCHEMA_ID,
        event_type: "click"
      }
    });
  };
  return {
    navBarImpression: logNavBarImpression,
    closePopupShow: logClosePopupShow,
    closePopupCtaClick: logClosePopupCtaClick,
    closeButtonClick: logCloseButtonClick,
    homeButtonClick: logHomeButtonClick
  };
}

// src/components/NavigationBar/RNNavigationBar/hooks/useNavigationEvent.ts
function useNavigationEvent() {
  const backEventContext = useBackEventContext();
  const logging = useNavigationBarLogging();
  const navigation = useNavigation();
  const closeConfirm = useCloseConfirm();
  return useMemo2(() => {
    const close = async () => {
      const hasConfirmed = await closeConfirm({
        onEntered: logging.closePopupShow
      });
      logging.closePopupCtaClick(hasConfirmed);
      if (hasConfirmed) {
        closeView3();
      }
    };
    return {
      handleBack: () => {
        if (backEventContext.hasBackEvent) {
          backEventContext.onBack();
        } else if (navigation.canGoBack()) {
          navigation.goBack();
        } else {
          close();
        }
      },
      handleHomeButtonClick: () => {
        logging.homeButtonClick();
        if (homeEvent.hasSubscriptions()) {
          homeEvent.emit({});
        } else {
          navigation.navigate("/");
        }
      },
      handleCloseButtonClick: () => {
        logging.closeButtonClick();
        close();
      }
    };
  }, [backEventContext, navigation, closeConfirm, logging]);
}

// src/core/hooks/useMoreButtonBottomSheet/index.tsx
import { INTERNAL__appBridgeHandler as INTERNAL__appBridgeHandler2, isMinVersionSupported as isMinVersionSupported2 } from "@apps-in-toss/native-modules";
import { openURL as openURL5 } from "@granite-js/react-native";
import { BottomSheet, List, ListHeader } from "@toss/tds-react-native";
import { useAdaptive as useAdaptive2, useOverlay as useOverlay2 } from "@toss/tds-react-native/private";
import { useEffect as useEffect5, useState as useState4 } from "react";

// src/core/hooks/useMoreButtonBottomSheet/AppShareListMenu.tsx
import { getOperationalEnvironment, shareWithScheme } from "@apps-in-toss/native-modules";
import { getSchemeUri as getSchemeUri4 } from "@granite-js/react-native";
import { useDialog as useDialog4 } from "@toss/tds-react-native";

// src/core/hooks/useMoreButtonBottomSheet/Menu.tsx
import { ListRow, Asset as Asset2 } from "@toss/tds-react-native";
import { useAdaptive } from "@toss/tds-react-native/private";
import { View as View2 } from "react-native";

// src/core/hooks/useMoreButtonBottomSheet/useMoreButtonBottomSheetLogging.tsx
import { Granite as Granite4 } from "@granite-js/react-native";
var BOTTOM_SHEET_SCHEMA_ID = 1596825;
var BOTTOM_SHEET_LOG_NAME = "appsintoss_app_visit__common_module::bottomsheet__more";
var BOTTOM_SHEET_OPEN_SCHEMA_ID = 1596829;
var BOTTOM_SHEET_OPEN_LOG_NAME = "appsintoss_app_visit__common_module::click__icon_more";
var BOTTOM_SHEET_CLOSE_CLICK_SCHEMA_ID = 1596843;
var BOTTOM_SHEET_CLOSE_CLICK_LOG_NAME = "appsintoss_app_visit__common_module::bottomsheet__more::click__close";
var BOTTOM_SHEET_MENU_CLICK_SCHEMA_ID = 1596841;
var BOTTOM_SHEET_MENU_CLICK_LOG_NAME = "appsintoss_app_visit__common_module::bottomsheet__more::click__menu";
function useMoreButtonBottomSheetLogging() {
  const referrer = useReferrer();
  const baseParams = {
    referrer,
    app_name: Granite4.appName
  };
  const logBottomSheetShow = () => {
    sendLog({
      log_name: BOTTOM_SHEET_LOG_NAME,
      log_type: "popup",
      params: {
        ...baseParams,
        schema_id: BOTTOM_SHEET_SCHEMA_ID
      }
    });
  };
  const logBottomSheetOpen = () => {
    sendLog({
      log_name: BOTTOM_SHEET_OPEN_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        schema_id: BOTTOM_SHEET_OPEN_SCHEMA_ID,
        event_type: "click"
      }
    });
  };
  const logBottomSheetCloseClick = () => {
    sendLog({
      log_name: BOTTOM_SHEET_CLOSE_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        schema_id: BOTTOM_SHEET_CLOSE_CLICK_SCHEMA_ID,
        event_type: "click"
      }
    });
  };
  const logBottomSheetMenuClick = ({ title }) => {
    sendLog({
      log_name: BOTTOM_SHEET_MENU_CLICK_LOG_NAME,
      log_type: "event",
      params: {
        ...baseParams,
        schema_id: BOTTOM_SHEET_MENU_CLICK_SCHEMA_ID,
        event_type: "click",
        item_title: title
      }
    });
  };
  return {
    show: logBottomSheetShow,
    open: logBottomSheetOpen,
    close: logBottomSheetCloseClick,
    menuClick: logBottomSheetMenuClick
  };
}

// src/core/hooks/useMoreButtonBottomSheet/Menu.tsx
import { jsx as jsx5 } from "react/jsx-runtime";
function Menu({ title, iconURL, onPress }) {
  const globals = getAppsInTossGlobals();
  const adaptive = useAdaptive();
  const iconName = getIconName(iconURL);
  const logging = useMoreButtonBottomSheetLogging();
  const brandPrimaryColorRGB = hexToRGB(globals.brandPrimaryColor);
  const iconBackgroundColor = brandPrimaryColorRGB ? `rgba(${brandPrimaryColorRGB.join(",")},0.1)` : adaptive.grey100;
  return /* @__PURE__ */ jsx5(
    ListRow,
    {
      left: /* @__PURE__ */ jsx5(
        View2,
        {
          style: {
            width: 30,
            height: 30,
            borderRadius: 10,
            flexDirection: "row",
            alignItems: "center",
            justifyContent: "center",
            marginRight: 16,
            backgroundColor: iconBackgroundColor
          },
          children: iconName ? /* @__PURE__ */ jsx5(Asset2.Icon, { frameShape: { width: 20, height: 20 }, color: globals.brandPrimaryColor, name: iconName }) : /* @__PURE__ */ jsx5(Asset2.Image, { frameShape: { width: 20, height: 20 }, source: { uri: iconURL } })
        }
      ),
      contents: /* @__PURE__ */ jsx5(ListRow.Texts, { type: "1RowTypeA", top: title, topProps: { color: adaptive.grey700 } }),
      verticalPadding: "extraSmall",
      onPress: () => {
        logging.menuClick({ title });
        onPress?.();
      }
    }
  );
}
var iconNameRegExp = /^(icon-|icn-(?!(car|sec|bank-(fill|square|horizontal))-))(.+)/i;
var iconURLRegExp = /^https:\/\/static\.toss\.im\/icons\/(png\/\dx|svg|pdf)\/((icon-|icn-(?!(car|sec|bank-(fill|square|horizontal))-))(.+))\.(png|svg|pdf)$/i;
function isIconUrl(url) {
  return iconNameRegExp.test(url) || iconURLRegExp.test(url);
}
function getIconName(url) {
  if (!isIconUrl(url)) {
    return null;
  }
  if (iconNameRegExp.test(url)) {
    return url;
  }
  return url.match(iconURLRegExp)?.[2];
}
function hexToRGB(hex) {
  const rgb = hex.toLowerCase().match(/[0-9,a-f]{2}/g)?.map((str) => parseInt(str, 16));
  return rgb?.length === 3 ? rgb : void 0;
}

// src/core/utils/url.ts
function addParamsToUrl(url, params) {
  const urlObj = new URL(url);
  for (const [key, value] of Object.entries(params)) {
    urlObj.searchParams.set(key, value);
  }
  return urlObj.toString();
}

// src/core/hooks/useMoreButtonBottomSheet/AppShareListMenu.tsx
import { jsx as jsx6 } from "react/jsx-runtime";
var SHARE_SCHEME_REFERRER = "appsintoss.common_module_share";
var APP_SHARE_MENU_INFO = {
  contactItemName: "\uACF5\uC720\uD558\uAE30",
  contactIconUrl: "https://static.toss.im/icons/png/4x/icon-share-dots-mono.png"
};
function AppShareListMenu() {
  const initialScheme = getSchemeUri4();
  const isSandbox = getOperationalEnvironment() === "sandbox";
  const { openConfirm } = useDialog4();
  const schemeForShare = addParamsToUrl(initialScheme, {
    referrer: SHARE_SCHEME_REFERRER
  });
  return /* @__PURE__ */ jsx6(
    Menu,
    {
      title: APP_SHARE_MENU_INFO.contactItemName,
      iconURL: APP_SHARE_MENU_INFO.contactIconUrl,
      onPress: () => {
        if (isSandbox) {
          openConfirm({
            title: "\uACF5\uC720\uD558\uAE30 \uAE30\uB2A5 \uBBF8\uC9C0\uC6D0",
            description: "\uC0CC\uB4DC\uBC15\uC2A4 \uD658\uACBD\uC5D0\uC11C\uB294 \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC5B4\uC694. \uCF58\uC194\uC744 \uD1B5\uD574 \uD1A0\uC2A4\uC571\uC5D0\uC11C \uD14C\uC2A4\uD2B8\uD574 \uC8FC\uC138\uC694.",
            rightButton: "\uD655\uC778",
            closeOnDimmerClick: true
          });
          return;
        }
        shareWithScheme({ schemeURL: schemeForShare });
      }
    }
  );
}

// src/core/hooks/useMoreButtonBottomSheet/HomeShortcutMenu.tsx
import { getOperationalEnvironment as getOperationalEnvironment2, INTERNAL__appBridgeHandler } from "@apps-in-toss/native-modules";
import { jsx as jsx7 } from "react/jsx-runtime";
function HomeShortcutMenu() {
  const isSandbox = getOperationalEnvironment2() === "sandbox";
  const globals = getAppsInTossGlobals();
  if (isSandbox) {
    return null;
  }
  return /* @__PURE__ */ jsx7(
    Menu,
    {
      title: "\uD734\uB300\uD3F0 \uD648 \uD654\uBA74\uC5D0 \uCD94\uAC00",
      iconURL: "https://static.toss.im/icons/png/4x/icon-plus-circle-mono.png",
      onPress: () => {
        INTERNAL__appBridgeHandler.invokeAppBridgeMethod(
          "addMiniAppShortcut",
          {
            title: globals.brandDisplayName,
            appName: global.__granite.app.name,
            iconUrl: globals.brandIcon,
            guideUrl: "https://service.toss.im/app-mini-home/shortcut"
          },
          { onSuccess: () => {
          }, onError: () => {
          } }
        );
      }
    }
  );
}

// src/core/hooks/useMoreButtonBottomSheet/SettingsMenu.tsx
import { getOperationalEnvironment as getOperationalEnvironment3 } from "@apps-in-toss/native-modules";
import { Granite as Granite5, openURL as openURL4 } from "@granite-js/react-native";
import { jsx as jsx8 } from "react/jsx-runtime";
function SettingsMenu() {
  const globals = getAppsInTossGlobals();
  if (getOperationalEnvironment3() === "sandbox") {
    return null;
  }
  return /* @__PURE__ */ jsx8(
    Menu,
    {
      title: "\uC124\uC815",
      iconURL: "https://static.toss.im/icons/png/4x/icon-setting-mono.png",
      onPress: () => openURL4(
        `servicetoss://apps-in-toss-menu/settings?appName=${Granite5.appName}&displayAppName=${globals.brandDisplayName}`
      )
    }
  );
}

// src/core/utils/ensureValue.ts
function ensureValue(value, name) {
  if (value === void 0) {
    throw new Error(`${name} is required`);
  }
  return value;
}

// src/core/hooks/useMoreButtonBottomSheet/index.tsx
import { Fragment as Fragment3, jsx as jsx9, jsxs as jsxs2 } from "react/jsx-runtime";
var APP_BRIDGE_METHOD_NAME = "getMiniAppsSupportContact";
var MIN_VERSION = {
  BOTTOM_SHEET: {
    android: "5.226.0",
    ios: "5.226.0"
  },
  SHARE_LIST_MENU: {
    android: "5.230.0",
    ios: "5.230.0"
  },
  HOME_SHORTCUT_MENU: {
    android: "5.246.0",
    ios: "5.246.0"
  },
  SETTINGS_MENU: {
    android: "5.238.0",
    ios: "5.237.0"
  }
};
function useMoreButtonBottomSheet() {
  const globals = getAppsInTossGlobals();
  const adaptive = useAdaptive2();
  const [itemList, setItemList] = useState4([]);
  const appUpdateDialog = useAppUpdateDialog();
  const logging = useMoreButtonBottomSheetLogging();
  const overlay = useOverlay2();
  const title = ensureValue(globals.brandDisplayName, "displayName");
  const isHomeShortcutSupported = isMinVersionSupported2(MIN_VERSION.HOME_SHORTCUT_MENU);
  const isBottomSheetSupported = isMinVersionSupported2(MIN_VERSION.BOTTOM_SHEET);
  const isShareListMenuSupported = isMinVersionSupported2(MIN_VERSION.SHARE_LIST_MENU);
  const isSettingsMenuSupported = isMinVersionSupported2(MIN_VERSION.SETTINGS_MENU);
  useEffect5(() => {
    if (!isBottomSheetSupported) {
      return;
    }
    INTERNAL__appBridgeHandler2.invokeAppBridgeMethod(
      APP_BRIDGE_METHOD_NAME,
      {},
      {
        onSuccess: ({ items }) => setItemList(items),
        onError: (error) => console.error("\uBA54\uB274 \uBAA9\uB85D\uC744 \uAC00\uC838\uC624\uB294 \uB370 \uC2E4\uD328\uD588\uC5B4\uC694:", error)
      }
    );
  }, [isBottomSheetSupported]);
  const onClickHandler = async () => {
    logging.open();
    if (!isBottomSheetSupported) {
      await appUpdateDialog.open({
        title: `\uC774 \uAE30\uB2A5\uC744 \uC4F0\uB824\uBA74 \uC571 \uC5C5\uB370\uC774\uD2B8\uAC00 \uD544\uC694\uD574\uC694`,
        description: `\uBB38\uC758, \uAD8C\uD55C \uC124\uC815, \uC2E0\uACE0 \uB4F1 \uAE30\uB2A5\uC744 \uC4F8 \uC218 \uC788\uC5B4\uC694`
      });
      return;
    }
    overlay.open(({ isOpen, exit, close }) => {
      const handleClose = () => {
        logging.close();
        close();
      };
      return /* @__PURE__ */ jsx9(BottomSheetImpressionArea, { children: /* @__PURE__ */ jsx9(
        BottomSheet.Root,
        {
          header: /* @__PURE__ */ jsx9(
            ListHeader,
            {
              title: /* @__PURE__ */ jsx9(ListHeader.TitleParagraph, { color: adaptive.grey800, fontWeight: "bold", typography: "t5", children: title })
            }
          ),
          open: isOpen,
          cta: /* @__PURE__ */ jsx9(
            BottomSheet.CTA,
            {
              size: "large",
              type: "dark",
              style: "weak",
              onPress: () => {
                handleClose();
              },
              children: "\uB2EB\uAE30"
            }
          ),
          onClose: handleClose,
          onExited: exit,
          children: /* @__PURE__ */ jsxs2(List, { rowSeparator: "none", children: [
            itemList.map((item) => /* @__PURE__ */ jsx9(
              Menu,
              {
                title: item.contactItemName,
                iconURL: item.contactIconUrl,
                onPress: () => openURL5(item.contactUri)
              },
              item.contactItemName
            )),
            isHomeShortcutSupported && /* @__PURE__ */ jsx9(HomeShortcutMenu, {}),
            isShareListMenuSupported && /* @__PURE__ */ jsx9(AppShareListMenu, {}),
            isSettingsMenuSupported && /* @__PURE__ */ jsx9(SettingsMenu, {})
          ] })
        }
      ) });
    });
  };
  return { open: onClickHandler };
}
function BottomSheetImpressionArea({ children }) {
  const logging = useMoreButtonBottomSheetLogging();
  useEffect5(() => {
    logging.show();
  }, [logging]);
  return /* @__PURE__ */ jsx9(Fragment3, { children });
}

// src/core/utils/toIcon.ts
function toIcon(source) {
  return source.startsWith("http") ? { source: { uri: source } } : { name: source };
}

// src/components/NavigationBar/common/NavigationBarImpressionArea.tsx
import { useEffect as useEffect6, useRef as useRef2 } from "react";
import { Fragment as Fragment4, jsx as jsx10 } from "react/jsx-runtime";
function NavigationBarImpressionArea({
  children,
  withHomeButton
}) {
  const hasLogged = useRef2(false);
  const logging = useNavigationBarLogging();
  useEffect6(() => {
    if (hasLogged.current === false) {
      logging.navBarImpression({ home_icon_yn: withHomeButton ? "Y" : "N" });
      hasLogged.current = true;
    }
  }, [logging, withHomeButton]);
  return /* @__PURE__ */ jsx10(Fragment4, { children });
}

// src/components/NavigationBar/RNNavigationBar/Default.tsx
import { jsx as jsx11 } from "react/jsx-runtime";
function DefaultNavigationBar() {
  const globals = getAppsInTossGlobals();
  const { open: openMoreButtonBottomSheet } = useMoreButtonBottomSheet();
  const parsedNavigationBar = globals.navigationBar != null ? safeParseNavigationBar(globals.navigationBar) : null;
  const withHomeButton = parsedNavigationBar?.withHomeButton ?? false;
  const withBackButton = parsedNavigationBar?.withBackButton ?? true;
  const { navigationRightButton } = useNavigationBarContext();
  const navigationEvent = useNavigationEvent();
  useHardwareBackPress(navigationEvent.handleBack);
  return /* @__PURE__ */ jsx11(NavigationBarImpressionArea, { withHomeButton, children: /* @__PURE__ */ jsx11(
    TopNavigation,
    {
      title: globals.brandDisplayName,
      icon: toIcon(globals.brandIcon),
      onPressDots: openMoreButtonBottomSheet,
      contentVisible: true,
      onPressTitle: withHomeButton ? navigationEvent.handleHomeButtonClick : void 0,
      onPressClose: navigationEvent.handleCloseButtonClick,
      withHome: withHomeButton,
      fixedRightButton: navigationRightButton,
      leftVisible: withBackButton,
      children: /* @__PURE__ */ jsx11(NavigationLeft, { visible: withBackButton, children: /* @__PURE__ */ jsx11(NavigationBackButton, { onPress: navigationEvent.handleBack, canGoBack: false }) })
    }
  ) });
}

// src/components/NavigationBar/RNNavigationBar/Game.tsx
import { useSafeAreaInsets } from "@granite-js/native/react-native-safe-area-context";
import { PageNavbar } from "@toss/tds-react-native";
import { NavigationRightContent, useSafeAreaTop } from "@toss/tds-react-native/private";
import { Platform as Platform2, View as View3 } from "react-native";
import { Fragment as Fragment5, jsx as jsx12, jsxs as jsxs3 } from "react/jsx-runtime";
function GameNavigationBar() {
  const safeAreaTop = useSafeAreaTop();
  const { open: openMoreButtonBottomSheet } = useMoreButtonBottomSheet();
  const navigationEvent = useNavigationEvent();
  const { navigationRightButton } = useNavigationBarContext();
  const { right: safeAreaRight } = useSafeAreaInsets();
  useHardwareBackPress(navigationEvent.handleBack);
  return /* @__PURE__ */ jsxs3(Fragment5, { children: [
    /* @__PURE__ */ jsx12(PageNavbar, { preference: { type: "none" } }),
    /* @__PURE__ */ jsx12(
      View3,
      {
        style: {
          width: "100%",
          // TODO: UI관련 스타일 설정은 TDS로 이관
          height: Platform2.OS === "ios" ? 44 : 54,
          flexDirection: "row",
          alignItems: "center",
          justifyContent: "flex-end",
          position: "absolute",
          zIndex: Z_INDEX.CLOSE_BUTTON,
          marginTop: safeAreaTop,
          paddingRight: safeAreaRight + 10
        },
        pointerEvents: "box-none",
        children: /* @__PURE__ */ jsx12(
          NavigationRightContent,
          {
            fixedRightButton: navigationRightButton,
            onPressDots: openMoreButtonBottomSheet,
            onPressClose: navigationEvent.handleCloseButtonClick,
            theme: "dark"
          }
        )
      }
    )
  ] });
}

// src/components/NavigationBar/RNNavigationBar/index.tsx
var RNNavigationBar = {
  Default: DefaultNavigationBar,
  Game: GameNavigationBar
};

// src/bridge-entry.ts
var bridge_entry_exports = {};
__reExport(bridge_entry_exports, async_bridges_star);
__reExport(bridge_entry_exports, constant_bridges_star);
__reExport(bridge_entry_exports, event_bridges_star);
import * as async_bridges_star from "@apps-in-toss/native-modules/async-bridges";
import * as constant_bridges_star from "@apps-in-toss/native-modules/constant-bridges";
import * as event_bridges_star from "@apps-in-toss/native-modules/event-bridges";

// src/components/RNAppContainer.tsx
import { Fragment as Fragment6, jsx as jsx13, jsxs as jsxs4 } from "react/jsx-runtime";
function RNAppContainer({ children }) {
  const global2 = getAppsInTossGlobals();
  switch (global2.appType) {
    case "game":
      return /* @__PURE__ */ jsx13(GameAppContainer, { children });
    case "general":
    default:
      return /* @__PURE__ */ jsx13(GeneralAppContainer, { children });
  }
}
function GameAppContainer({ children }) {
  useEffect7(() => {
    if (Platform3.OS === "ios") {
      setIosSwipeGestureEnabled({ isEnabled: false });
      return () => {
        setIosSwipeGestureEnabled({ isEnabled: true });
      };
    }
    return;
  }, []);
  return /* @__PURE__ */ jsxs4(Fragment6, { children: [
    /* @__PURE__ */ jsx13(RNNavigationBar.Game, {}),
    (0, bridge_entry_exports.getOperationalEnvironment)() === "toss" ? /* @__PURE__ */ jsx13(GameInitializer, { children }) : children
  ] });
}
function GeneralAppContainer({ children }) {
  return /* @__PURE__ */ jsxs4(Fragment6, { children: [
    /* @__PURE__ */ jsx13(RNNavigationBar.Default, {}),
    children
  ] });
}

// src/core/registerApp.tsx
import { Fragment as Fragment7, jsx as jsx14, jsxs as jsxs5 } from "react/jsx-runtime";
function AppsInTossContainer(Container, { children, ...initialProps }) {
  if (!isMinVersionSupported3({
    android: "5.220.0",
    ios: "5.221.0"
  })) {
    return /* @__PURE__ */ jsxs5(Fragment7, { children: [
      /* @__PURE__ */ jsx14(AppEvent.System, { ...initialProps }),
      /* @__PURE__ */ jsx14(AppUpdate, {})
    ] });
  }
  return /* @__PURE__ */ jsxs5(Fragment7, { children: [
    /* @__PURE__ */ jsx14(AppEvent.System, { ...initialProps }),
    /* @__PURE__ */ jsx14(Container, { ...initialProps, children: /* @__PURE__ */ jsx14(TDSProvider, { colorPreference: "light", token: { color: { primary: getAppsInTossGlobals().brandPrimaryColor } }, children }) })
  ] });
}
function registerApp(container, { context, analytics }) {
  const appName = getAppName();
  const isRegistered = AppRegistry.getAppKeys().includes(appName);
  if (!isRegistered) {
    Analytics.init({
      logger: (params) => void eventLog2(params),
      debug: analytics?.debug ?? __DEV__
    });
    const App = Granite6.registerApp(AppsInTossContainer.bind(null, container), {
      appName,
      context,
      initialScheme: getSchemeUri5(),
      setIosSwipeGestureEnabled: setIosSwipeGestureEnabled2,
      router: {
        screenContainer: AppsInTossScreenContainer,
        defaultScreenOption: {
          statusBarStyle: "dark"
        }
      }
    });
    global.Page = App;
  }
  return global.Page;
}
function AppsInTossScreenContainer({ children }) {
  const isRNApp = getAppsInTossGlobals().webViewType == null;
  return /* @__PURE__ */ jsx14(NavigationBarContextProvider, { children: isRNApp ? /* @__PURE__ */ jsx14(Analytics.Screen, { children: /* @__PURE__ */ jsx14(RNAppContainer, { children }) }) : children });
}
function getAppName() {
  try {
    return global.__granite.app.name;
  } catch (error) {
    console.error("unexpected error occurred while getting app name");
    throw error;
  }
}

// src/core/index.ts
var AppsInToss = {
  registerApp
};

// src/components/WebView.tsx
import {
  GoogleAdMob as GoogleAdMob2,
  IAP,
  Storage,
  appsInTossEvent,
  getServerTime,
  requestReview,
  iapCreateOneTimePurchaseOrder,
  processProductGrant,
  requestOneTimePurchase,
  safeSyncPostMessage
} from "@apps-in-toss/native-modules";
import * as appsInTossAsyncBridges from "@apps-in-toss/native-modules/async-bridges";
import * as appsInTossConstantBridges from "@apps-in-toss/native-modules/constant-bridges";
import * as appsInTossEventBridges from "@apps-in-toss/native-modules/event-bridges";
import { afterDocumentLoad, beforeDocumentLoad } from "@apps-in-toss/user-scripts";
import { useSafeAreaInsets as useSafeAreaInsets4 } from "@granite-js/native/react-native-safe-area-context";
import { getSchemeUri as getSchemeUri9 } from "@granite-js/react-native";
import { ExternalWebViewScreen, tdsEvent } from "@toss/tds-react-native";
import { useSafeAreaBottom, useSafeAreaTop as useSafeAreaTop3 } from "@toss/tds-react-native/private";
import { useEffect as useEffect12, useMemo as useMemo6, useRef as useRef5, useState as useState6 } from "react";
import { BackHandler as BackHandler2, Linking, Platform as Platform6 } from "react-native";

// src/components/GameWebView.tsx
import { setIosSwipeGestureEnabled as setIosSwipeGestureEnabled3, getOperationalEnvironment as getOperationalEnvironment5 } from "@apps-in-toss/native-modules";
import {
  WebView as PlainWebView
} from "@granite-js/native/react-native-webview";
import { forwardRef, useEffect as useEffect8 } from "react";
import { Platform as Platform5 } from "react-native";

// src/components/NavigationBar/GameWebviewNavigationBar.tsx
import { useSafeAreaInsets as useSafeAreaInsets2 } from "@granite-js/native/react-native-safe-area-context";
import { closeView as closeView4 } from "@granite-js/react-native";
import { PageNavbar as PageNavbar2, useDialog as useDialog5 } from "@toss/tds-react-native";
import { NavigationRightContent as NavigationRightContent2, useSafeAreaTop as useSafeAreaTop2 } from "@toss/tds-react-native/private";
import { josa as josa3 } from "es-hangul";
import { useCallback as useCallback7 } from "react";
import { Platform as Platform4, View as View4 } from "react-native";
import { Fragment as Fragment8, jsx as jsx15, jsxs as jsxs6 } from "react/jsx-runtime";
function GameWebviewNavigationBar() {
  const safeAreaTop = useSafeAreaTop2();
  const { openConfirm } = useDialog5();
  const global2 = getAppsInTossGlobals();
  const logging = useNavigationBarLogging();
  const { open: openMoreButtonBottomSheet } = useMoreButtonBottomSheet();
  const { navigationRightButton } = useNavigationBarContext();
  const { right: safeAreaRight } = useSafeAreaInsets2();
  const handleGameWebviewClose = useCallback7(async () => {
    logging.closeButtonClick();
    const isConfirmed = await openConfirm({
      title: `${josa3(global2.brandDisplayName, "\uC744/\uB97C")} \uC885\uB8CC\uD560\uAE4C\uC694?`,
      leftButton: "\uB2EB\uAE30",
      rightButton: "\uC885\uB8CC\uD558\uAE30",
      closeOnDimmerClick: true,
      onEntered: logging.closePopupShow
    });
    logging.closePopupCtaClick(isConfirmed);
    if (isConfirmed) {
      closeView4();
    }
  }, [global2.brandDisplayName, logging, openConfirm]);
  return /* @__PURE__ */ jsxs6(Fragment8, { children: [
    /* @__PURE__ */ jsx15(PageNavbar2, { preference: { type: "none" } }),
    /* @__PURE__ */ jsx15(
      View4,
      {
        style: {
          width: "100%",
          // TODO: UI관련 스타일 설정은 TDS로 이관
          height: Platform4.OS === "ios" ? 44 : 54,
          flexDirection: "row",
          alignItems: "center",
          justifyContent: "flex-end",
          position: "absolute",
          zIndex: Z_INDEX.CLOSE_BUTTON,
          marginTop: safeAreaTop,
          paddingRight: safeAreaRight + 10
        },
        pointerEvents: "box-none",
        children: /* @__PURE__ */ jsx15(
          NavigationRightContent2,
          {
            fixedRightButton: navigationRightButton,
            onPressDots: openMoreButtonBottomSheet,
            onPressClose: () => {
              handleGameWebviewClose();
            },
            theme: "dark"
          }
        )
      }
    )
  ] });
}

// src/components/GameWebView.tsx
import { Fragment as Fragment9, jsx as jsx16, jsxs as jsxs7 } from "react/jsx-runtime";
var GameWebView = forwardRef(function GameWebView2(props, ref) {
  useEffect8(() => {
    if (Platform5.OS === "ios") {
      setIosSwipeGestureEnabled3({ isEnabled: false });
      return () => {
        setIosSwipeGestureEnabled3({ isEnabled: true });
      };
    }
    return;
  }, []);
  return /* @__PURE__ */ jsxs7(Fragment9, { children: [
    /* @__PURE__ */ jsx16(GameWebviewNavigationBar, {}),
    getOperationalEnvironment5() === "toss" ? /* @__PURE__ */ jsx16(GameInitializer, { children: /* @__PURE__ */ jsx16(PlainWebView, { ref, ...props }) }) : /* @__PURE__ */ jsx16(PlainWebView, { ref, ...props })
  ] });
});

// src/components/PartnerWebView.tsx
import {
  WebView as PlainWebView2
} from "@granite-js/native/react-native-webview";
import { forwardRef as forwardRef2 } from "react";

// src/components/NavigationBar/PartnerWebviewNavigationBar.tsx
import { closeView as closeView5 } from "@granite-js/react-native";
import { useDialog as useDialog6 } from "@toss/tds-react-native";
import { NavigationBackButton as NavigationBackButton2, NavigationLeft as NavigationLeft2, TopNavigation as TopNavigation2 } from "@toss/tds-react-native/private";
import { josa as josa4 } from "es-hangul";
import { useCallback as useCallback8 } from "react";
import { jsx as jsx17 } from "react/jsx-runtime";
function PartnerWebviewNavigationBar({ onBackButtonClick, onHomeButtonClick }) {
  const globals = getAppsInTossGlobals();
  const logging = useNavigationBarLogging();
  const { openConfirm } = useDialog6();
  const { open: openMoreButtonBottomSheet } = useMoreButtonBottomSheet();
  const parsedNavigationBar = globals.navigationBar != null ? safeParseNavigationBar(globals.navigationBar) : null;
  const withHomeButton = parsedNavigationBar?.withHomeButton ?? false;
  const withBackButton = parsedNavigationBar?.withBackButton ?? true;
  const { navigationRightButton } = useNavigationBarContext();
  const handleClose = useCallback8(async () => {
    logging.closeButtonClick();
    const isConfirmed = await openConfirm({
      title: `${josa4(globals.brandDisplayName, "\uC744/\uB97C")} \uC885\uB8CC\uD560\uAE4C\uC694?`,
      leftButton: "\uB2EB\uAE30",
      rightButton: "\uC885\uB8CC\uD558\uAE30",
      closeOnDimmerClick: true,
      onEntered: logging.closePopupShow
    });
    logging.closePopupCtaClick(isConfirmed);
    if (isConfirmed) {
      closeView5();
    }
  }, [globals.brandDisplayName, logging, openConfirm]);
  return /* @__PURE__ */ jsx17(NavigationBarImpressionArea, { withHomeButton, children: /* @__PURE__ */ jsx17(
    TopNavigation2,
    {
      title: globals.brandDisplayName,
      icon: toIcon(globals.brandIcon),
      onPressDots: openMoreButtonBottomSheet,
      contentVisible: true,
      onPressTitle: withHomeButton ? onHomeButtonClick : void 0,
      onPressClose: handleClose,
      withHome: withHomeButton,
      fixedRightButton: navigationRightButton,
      leftVisible: withBackButton,
      children: /* @__PURE__ */ jsx17(NavigationLeft2, { visible: withBackButton, children: /* @__PURE__ */ jsx17(NavigationBackButton2, { onPress: onBackButtonClick, canGoBack: false }) })
    }
  ) });
}

// src/components/PartnerWebView.tsx
import { Fragment as Fragment10, jsx as jsx18, jsxs as jsxs8 } from "react/jsx-runtime";
var PartnerWebView = forwardRef2(function PartnerWebViewScreen({ onBackButtonClick, onHomeButtonClick, ...webViewProps }, ref) {
  return /* @__PURE__ */ jsxs8(Fragment10, { children: [
    /* @__PURE__ */ jsx18(PartnerWebviewNavigationBar, { onBackButtonClick, onHomeButtonClick }),
    /* @__PURE__ */ jsx18(PlainWebView2, { ref, ...webViewProps, style: { flex: 1 } })
  ] });
});

// src/components/utils/url.ts
function convertIntentURL(url) {
  if (url.protocol !== "intent:") {
    return null;
  }
  const components = url.hash.replace("#Intent;", "").split(";").map((component) => component.split("="));
  const scheme = components.find(([key]) => key === "scheme")?.[1];
  if (!scheme) {
    return null;
  }
  return `${scheme}://${url.hostname}${url.pathname}${url.search}`;
}

// src/ads/fetchTossAd.ts
import {
  getOperationalEnvironment as getOperationalEnvironment6,
  INTERNAL__appBridgeHandler as INTERNAL__appBridgeHandler3,
  isMinVersionSupported as isMinVersionSupported4
} from "@apps-in-toss/native-modules";

// ../../../.yarn/cache/es-toolkit-npm-1.34.1-4cd6371dcb-aab6d07be3.zip/node_modules/es-toolkit/dist/function/noop.mjs
function noop() {
}

// src/ads/adGroupId.ts
var INVALID_AD_GROUP_ID_ERROR_MESSAGE = "\uC798\uBABB\uB41C \uC694\uCCAD\uC774\uC5D0\uC694. \uD544\uC694\uD55C \uAC12\uC774 \uBE44\uC5B4 \uC788\uC5B4\uC694.";
function normalizeAdGroupId(adGroupId) {
  return adGroupId.trim();
}
function createInvalidAdGroupIdError() {
  return new Error(INVALID_AD_GROUP_ID_ERROR_MESSAGE);
}

// src/ads/generateSessionId.ts
function generateSessionId() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = Math.random() * 16 | 0;
    const v = c === "x" ? r : r & 3 | 8;
    return v.toString(16);
  });
}

// src/ads/constants.ts
var sessionId = generateSessionId();

// src/ads/fetchTossAd.ts
var ANDROID_FETCH_TOSS_AD_SUPPORTED_VERSION = "5.241.0";
var IOS_FETCH_TOSS_AD_SUPPORTED_VERSION = "5.241.0";
var UNSUPPORTED_ERROR_MESSAGE = "This feature is not supported in the current environment";
var ENVIRONMENT = getOperationalEnvironment6();
function fetchTossAd(params) {
  if (!fetchTossAd.isSupported()) {
    params.onError(new Error(UNSUPPORTED_ERROR_MESSAGE));
    return noop;
  }
  const { onEvent, onError, options } = params;
  const adGroupId = normalizeAdGroupId(options.adGroupId);
  if (adGroupId.length === 0) {
    params.onError(createInvalidAdGroupIdError());
    return noop;
  }
  const unregisterCallbacks = INTERNAL__appBridgeHandler3.invokeAppBridgeMethod(
    "fetchTossAd",
    { spaceUnitId: adGroupId, sessionId, sdkId: options.sdkId, availableStyleIds: options.availableStyleIds },
    {
      onSuccess: (response) => {
        onEvent(response);
      },
      onError: (error) => {
        onError(error);
      }
    }
  );
  return unregisterCallbacks;
}
fetchTossAd.isSupported = () => {
  if (ENVIRONMENT !== "toss") {
    return false;
  }
  return isMinVersionSupported4({
    android: ANDROID_FETCH_TOSS_AD_SUPPORTED_VERSION,
    ios: IOS_FETCH_TOSS_AD_SUPPORTED_VERSION
  });
};

// src/ads/integratedAd.ts
import { GoogleAdMob, INTERNAL__appBridgeHandler as INTERNAL__appBridgeHandler4, isMinVersionSupported as isMinVersionSupported5 } from "@apps-in-toss/native-modules";

// src/ads/generateRequestId.ts
function generateRequestId() {
  const timestamp = Date.now().toString(36).toUpperCase();
  const random = Math.random().toString(36).substring(2, 7).toUpperCase();
  return `ait-${timestamp}${random}`;
}

// src/ads/getReferrer.ts
import { getSchemeUri as getSchemeUri6 } from "@granite-js/react-native";
function getReferrer() {
  try {
    return new URL(getSchemeUri6()).searchParams.get("referrer");
  } catch {
    return null;
  }
}

// src/ads/remotes.ts
import { getOperationalEnvironment as getOperationalEnvironment7, getPlatformOS, getTossAppVersion } from "@apps-in-toss/native-modules";

// src/ads/getIsDev.ts
import { getSchemeUri as getSchemeUri7 } from "@granite-js/react-native";
function getIsDev() {
  try {
    return new URL(getSchemeUri7()).searchParams.get("isDev") === "true";
  } catch {
    return false;
  }
}

// src/ads/remotes.ts
var OPERATIONAL_ENVIRONMENT = getOperationalEnvironment7();
var OS = getPlatformOS();
var APP_VER = getTossAppVersion();
var ALPHA_EVENT_TRACKER_HTTP_ENDPOINT = "https://alpha-trillion.toss.im/trk/sdk-mediation/event";
var LIVE_EVENT_TRACKER_HTTP_ENDPOINT = "https://trillion.toss.im/trk/sdk-mediation/event";
function getPostEventTrackingUrl() {
  const isDev = getIsDev();
  if (OPERATIONAL_ENVIRONMENT === "sandbox" || isDev) {
    return ALPHA_EVENT_TRACKER_HTTP_ENDPOINT;
  }
  return LIVE_EVENT_TRACKER_HTTP_ENDPOINT;
}
function postEventTracking(params) {
  const endpoint = getPostEventTrackingUrl();
  const bodyJSON = JSON.stringify({
    ...params,
    os: OS,
    appVer: APP_VER,
    deviceIdType: "NONE",
    platform: "RN"
  });
  return fetch(endpoint, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: bodyJSON
  }).catch(noop);
}

// src/ads/integratedAd.ts
var INTEGRATED_AD_SDK_VERSION = "0.0.0";
var ANDROID_INTEGRATED_AD_SUPPORTED_VERSION = "5.247.0";
var IOS_INTEGRATED_AD_SUPPORTED_VERSION = "5.247.0";
var UNSUPPORTED_ERROR_MESSAGE2 = "This feature is not supported in the current environment";
var INTG_AD_ADM_FALLBACK_RID_MAP = {};
function integratedAdIsSupported() {
  return isMinVersionSupported5({
    android: ANDROID_INTEGRATED_AD_SUPPORTED_VERSION,
    ios: IOS_INTEGRATED_AD_SUPPORTED_VERSION
  });
}
function generateLoadFullScreenAd(sdkId) {
  const fn = (params) => {
    if (!GoogleAdMob.loadAppsInTossAdMob.isSupported()) {
      params.onError(new Error(UNSUPPORTED_ERROR_MESSAGE2));
      return noop;
    }
    const normalizedAdGroupId = normalizeAdGroupId(params.options.adGroupId);
    if (normalizedAdGroupId.length === 0) {
      params.onError(createInvalidAdGroupIdError());
      return noop;
    }
    const normalizedParams = {
      ...params,
      options: {
        ...params.options,
        adGroupId: normalizedAdGroupId
      }
    };
    if (!integratedAdIsSupported()) {
      return GoogleAdMob.loadAppsInTossAdMob({
        ...normalizedParams,
        onEvent: (event) => {
          if (event.type === "loaded") {
            const requestId = generateRequestId();
            INTG_AD_ADM_FALLBACK_RID_MAP[normalizedAdGroupId] = requestId;
            postEventTracking({
              eventName: "LOAD",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: event.data.responseInfo.responseId,
              mediationType: "ADMOB",
              format: "",
              adSourceName: event.data.responseInfo.loadedAdNetworkInfo?.adSourceName ?? "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
          }
          return normalizedParams.onEvent(event);
        },
        onError: (error) => {
          const requestId = INTG_AD_ADM_FALLBACK_RID_MAP[normalizedAdGroupId] ?? "";
          postEventTracking({
            eventName: "FAILED_TO_LOAD",
            spaceUnitId: normalizedAdGroupId,
            requestId,
            responseId: "",
            mediationType: "ADMOB",
            format: "",
            adSourceName: "",
            sdkVer: INTEGRATED_AD_SDK_VERSION
          });
          return normalizedParams.onError(error);
        }
      });
    }
    const { onEvent, onError, options } = normalizedParams;
    const { adGroupId } = options;
    const referrer = getReferrer();
    const unregisterCallbacks = INTERNAL__appBridgeHandler4.invokeAppBridgeMethod(
      "loadTossAdOrAdmob",
      { spaceUnitId: adGroupId, referrer, sessionId, sdkId },
      {
        onSuccess: () => {
          onEvent({ type: "loaded" });
        },
        onError: (error) => {
          onError(error);
        }
      }
    );
    return unregisterCallbacks;
  };
  fn.isSupported = GoogleAdMob.loadAppsInTossAdMob.isSupported;
  return fn;
}
var loadFullScreenAd = generateLoadFullScreenAd("107");
var loadFullScreenAdForWeb = generateLoadFullScreenAd("107");
function showFullScreenAd(params) {
  if (!GoogleAdMob.showAppsInTossAdMob.isSupported()) {
    params.onError(new Error(UNSUPPORTED_ERROR_MESSAGE2));
    return noop;
  }
  const normalizedAdGroupId = normalizeAdGroupId(params.options.adGroupId);
  if (normalizedAdGroupId.length === 0) {
    params.onError(createInvalidAdGroupIdError());
    return noop;
  }
  const normalizedParams = {
    ...params,
    options: {
      ...params.options,
      adGroupId: normalizedAdGroupId
    }
  };
  if (!integratedAdIsSupported()) {
    return GoogleAdMob.showAppsInTossAdMob({
      ...normalizedParams,
      onEvent: (event) => {
        const requestId = INTG_AD_ADM_FALLBACK_RID_MAP[normalizedAdGroupId] ?? "";
        switch (event.type) {
          case "show":
            postEventTracking({
              eventName: "SHOW",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: "",
              mediationType: "ADMOB",
              format: "",
              adSourceName: "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
            break;
          case "failedToShow":
            postEventTracking({
              eventName: "FAILED_TO_SHOW",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: "",
              mediationType: "ADMOB",
              format: "",
              adSourceName: "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
            break;
          case "impression":
            postEventTracking({
              eventName: "IMP",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: "",
              mediationType: "ADMOB",
              format: "",
              adSourceName: "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
            break;
          case "clicked":
            postEventTracking({
              eventName: "CLICK",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: "",
              mediationType: "ADMOB",
              format: "",
              adSourceName: "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
            break;
          case "dismissed":
            postEventTracking({
              eventName: "DISMISS",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: "",
              mediationType: "ADMOB",
              format: "",
              adSourceName: "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
            break;
          case "userEarnedReward":
            postEventTracking({
              eventName: "USER_EARNED_REWARD",
              spaceUnitId: normalizedAdGroupId,
              requestId,
              responseId: "",
              mediationType: "ADMOB",
              format: "",
              adSourceName: "",
              sdkVer: INTEGRATED_AD_SDK_VERSION
            });
            break;
        }
        return normalizedParams.onEvent(event);
      },
      onError: (error) => {
        const requestId = INTG_AD_ADM_FALLBACK_RID_MAP[normalizedAdGroupId] ?? "";
        postEventTracking({
          eventName: "FAILED_TO_SHOW",
          spaceUnitId: normalizedAdGroupId,
          requestId,
          responseId: "",
          mediationType: "ADMOB",
          format: "",
          adSourceName: "",
          sdkVer: INTEGRATED_AD_SDK_VERSION
        });
        return normalizedParams.onError(error);
      }
    });
  }
  const { onEvent, onError, options } = normalizedParams;
  const { adGroupId } = options;
  const referrer = getReferrer();
  const unregisterCallbacks = INTERNAL__appBridgeHandler4.invokeAppBridgeMethod(
    "showTossAdOrAdmob",
    { spaceUnitId: adGroupId, referrer, sessionId },
    {
      onAdClicked: () => {
        onEvent({ type: "clicked" });
      },
      onAdDismissed: () => {
        onEvent({ type: "dismissed" });
      },
      onAdFailedToShow: () => {
        onEvent({ type: "failedToShow" });
      },
      onAdImpression: () => {
        onEvent({ type: "impression" });
      },
      onAdShow: () => {
        onEvent({ type: "show" });
      },
      onUserEarnedReward: (data) => {
        onEvent({ type: "userEarnedReward", data });
      },
      onSuccess: () => {
        onEvent({ type: "requested" });
      },
      onError: (error) => {
        onError(error);
      }
    }
  );
  return unregisterCallbacks;
}
showFullScreenAd.isSupported = GoogleAdMob.showAppsInTossAdMob.isSupported;

// src/ads/tossAdEventLog.ts
import { Granite as Granite7 } from "@granite-js/react-native";
async function tossAdEventLog(params) {
  const referrer = getReferrer();
  const appName = Granite7.appName;
  const eventLogParams = {
    ...params,
    params: {
      ...params.params,
      referrer,
      app_name: appName
    }
  };
  return sendLog(eventLogParams, { pipelines: ["core"] });
}

// src/bridge-handler/useBridgeHandler.tsx
import { useCallback as useCallback9, useEffect as useEffect9, useMemo as useMemo3, useRef as useRef3 } from "react";
function serializeError(error) {
  return JSON.stringify(error, (_, value) => {
    if (value instanceof Error) {
      return {
        ...Object.entries(value).reduce((acc, [key, value2]) => {
          acc[key] = value2;
          return acc;
        }, {}),
        name: value.name,
        message: value.message,
        stack: value.stack,
        __isError: true
      };
    }
    return value;
  });
}
function methodHandler({
  args,
  eventId,
  functionName,
  handlerMap,
  injectJavaScript
}) {
  const func = handlerMap[functionName];
  if (!func) {
    console.error(`${functionName} is not a function`);
    return;
  }
  const wrappedFunc = async (...args2) => {
    const result = await func(...args2);
    return result;
  };
  wrappedFunc(...args).then((result) => {
    injectJavaScript?.(`
      window.__GRANITE_NATIVE_EMITTER.emit('${functionName}/resolve/${eventId}', ${JSON.stringify(result, null, 0)});
    `);
  }).catch((error) => {
    const serializedError = serializeError(error);
    injectJavaScript?.(`
      window.__GRANITE_NATIVE_EMITTER.emit('${functionName}/reject/${eventId}', ${serializedError});
    `);
  });
}
var globalEventListenerMap = /* @__PURE__ */ new Map();
function useBridgeHandler({
  onMessage,
  constantHandlerMap,
  asyncHandlerMap,
  eventListenerMap
}) {
  const ref = useRef3(null);
  const injectedJavaScript = useMemo3(
    () => `window.__CONSTANT_HANDLER_MAP = ${JSON.stringify(
      Object.entries(constantHandlerMap).reduce(
        (acc, [key, value]) => {
          acc[key] = typeof value === "function" ? value() : value;
          return acc;
        },
        {}
      )
    )};`,
    [constantHandlerMap]
  );
  useEffect9(() => {
    ref.current?.injectJavaScript(injectedJavaScript);
  }, [injectedJavaScript]);
  const createHandleOnEvent = (functionName, eventId) => (response) => {
    ref.current?.injectJavaScript(`
      window.__GRANITE_NATIVE_EMITTER.emit('${functionName}/onEvent/${eventId}', ${JSON.stringify(response, null, 0)});
    `);
  };
  const createHandleOnError = (functionName, eventId) => (error) => {
    const serializedError = serializeError(error);
    ref.current?.injectJavaScript(`
      window.__GRANITE_NATIVE_EMITTER.emit('${functionName}/onError/${eventId}', ${serializedError});
    `);
  };
  const $onMessage = useCallback9(
    async (e) => {
      onMessage?.(e);
      const data = parseNativeEventData(e.nativeEvent.data);
      if (typeof data !== "object" || data === null || typeof data.functionName !== "string" || typeof data.eventId !== "string" || typeof data.type !== "string" || !["addEventListener", "removeEventListener", "method"].includes(data.type)) {
        return;
      }
      switch (data.type) {
        case "addEventListener": {
          const handleOnEvent = createHandleOnEvent(data.functionName, data.eventId);
          const handleOnError = createHandleOnError(data.functionName, data.eventId);
          const remove = eventListenerMap[data.functionName]?.({
            onEvent: handleOnEvent,
            onError: handleOnError,
            options: data.args
          });
          if (remove) {
            globalEventListenerMap.set(`${data.functionName}/${data.eventId}`, remove);
          }
          break;
        }
        case "removeEventListener": {
          const key = `${data.functionName}/${data.eventId}`;
          const remove = globalEventListenerMap.get(key);
          remove?.();
          globalEventListenerMap.delete(key);
          break;
        }
        case "method": {
          methodHandler({
            args: data.args,
            eventId: data.eventId,
            functionName: data.functionName,
            handlerMap: asyncHandlerMap,
            injectJavaScript: ref.current?.injectJavaScript
          });
          break;
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [onMessage]
  );
  return {
    ref,
    injectedJavaScript,
    onMessage: $onMessage
  };
}
function parseNativeEventData(data) {
  try {
    return JSON.parse(data);
  } catch (error) {
    console.error(error);
    return null;
  }
}

// src/core/hooks/useSafeAreaInsetsEvent.tsx
import { useSafeAreaInsets as useSafeAreaInsets3 } from "@granite-js/native/react-native-safe-area-context";
import { useEffect as useEffect10 } from "react";
function useSafeAreaInsetsEvent() {
  const insets = useSafeAreaInsets3();
  useEffect10(() => {
    safeAreaInsetsChange.emit(insets);
  }, [insets]);
  useEffect10(() => {
    return () => safeAreaInsetsChange.clearSubscriptions();
  }, []);
  return safeAreaInsetsChange;
}

// src/core/hooks/useWebBackHandler.tsx
import { closeView as closeView6, useBackEventState } from "@granite-js/react-native";
import { useDialog as useDialog7 } from "@toss/tds-react-native";
import { josa as josa5 } from "es-hangul";
import { useCallback as useCallback11, useMemo as useMemo5 } from "react";

// src/hooks/useWebviewHistoryStack.tsx
import { useCallback as useCallback10, useMemo as useMemo4, useReducer } from "react";
var INITIAL_STATE = { stack: [], index: -1 };
function reducer(state, action) {
  switch (action.type) {
    case "NAVIGATION_CHANGE": {
      const { url, canGoForward } = action;
      if (state.stack.length === 0) {
        return { stack: [url], index: 0 };
      }
      const { stack, index } = state;
      const cur = stack[index];
      if (url === cur) {
        return state;
      }
      const prev = index > 0 ? stack[index - 1] : void 0;
      const next = index < stack.length - 1 ? stack[index + 1] : void 0;
      if (prev && url === prev && canGoForward) {
        return { ...state, index: index - 1 };
      }
      if (next && url === next) {
        return { ...state, index: index + 1 };
      }
      const base = stack.slice(0, index + 1);
      const nextStack = [...base, url];
      return { stack: nextStack, index: nextStack.length - 1 };
    }
    default:
      return state;
  }
}
function useWebViewHistory() {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const onNavigationStateChange = useCallback10(({ url, canGoForward: canGoForward2 }) => {
    dispatch({ type: "NAVIGATION_CHANGE", url, canGoForward: canGoForward2 });
  }, []);
  const { canGoBack, canGoForward } = useMemo4(() => {
    const canBack = state.index > 0;
    const canFwd = state.index >= 0 && state.index < state.stack.length - 1;
    return { canGoBack: canBack, canGoForward: canFwd };
  }, [state.index, state.stack.length]);
  return { onNavigationStateChange, hasHistory: canGoBack, canGoForward };
}

// src/core/hooks/useWebBackHandler.tsx
function useWebBackHandler(webViewInitialURL, webViewRef) {
  const { hasHistory, onNavigationStateChange } = useWebViewHistory();
  const {
    handlersRef: webBackHandlersRef,
    hasBackEvent: hasWebBackEvent,
    addEventListener: addWebBackEventListener,
    removeEventListener: removeWebBackEventListener
  } = useBackEventState();
  const logging = useNavigationBarLogging();
  const { openConfirm } = useDialog7();
  const global2 = getAppsInTossGlobals();
  const addEventListener = useCallback11(
    (handler) => {
      addWebBackEventListener(handler);
    },
    [addWebBackEventListener]
  );
  const removeEventListener = useCallback11(
    (handler) => {
      removeWebBackEventListener(handler);
    },
    [removeWebBackEventListener]
  );
  const handleWebBack = useCallback11(async () => {
    if (hasWebBackEvent) {
      for (const handler of webBackHandlersRef) {
        handler();
      }
      return;
    }
    if (hasHistory) {
      webViewRef.current?.injectJavaScript("window.history.back();");
    } else {
      const isConfirmed = await openConfirm({
        title: `${josa5(global2.brandDisplayName, "\uC744/\uB97C")} \uC885\uB8CC\uD560\uAE4C\uC694?`,
        leftButton: "\uB2EB\uAE30",
        rightButton: "\uC885\uB8CC\uD558\uAE30",
        closeOnDimmerClick: true,
        onEntered: logging.closePopupShow
      });
      logging.closePopupCtaClick(isConfirmed);
      if (isConfirmed) {
        closeView6();
      }
    }
  }, [global2.brandDisplayName, hasHistory, hasWebBackEvent, webBackHandlersRef, logging, openConfirm, webViewRef]);
  const handleWebHome = useCallback11(() => {
    logging.homeButtonClick();
    if (homeEvent.hasSubscriptions()) {
      homeEvent.emit({});
      return;
    }
    webViewRef.current?.injectJavaScript(`
    (function() {
      const webBundleOrigin = '${webViewInitialURL.origin}';
      window.location.href = window.location.origin === webBundleOrigin ? '/' : webBundleOrigin;
    })();
    `);
  }, [hasWebBackEvent, webBackHandlersRef, logging, webViewInitialURL, webViewRef]);
  return useMemo5(
    () => ({ addEventListener, removeEventListener, handleWebBack, handleWebHome, onNavigationStateChange }),
    [addEventListener, removeEventListener, handleWebBack, handleWebHome, onNavigationStateChange]
  );
}

// src/core/utils/mergeRefs.ts
function mergeRefs(...refs) {
  return (value) => {
    refs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref != null) {
        ref.current = value;
      }
    });
  };
}

// src/hooks/useCreateUserAgent.ts
import { getPlatformOS as getPlatformOS2, getTossAppVersion as getTossAppVersion2 } from "@apps-in-toss/native-modules";
import { useWindowDimensions } from "react-native";
var FontA11yCategory = {
  Large: "Large",
  xLarge: "xLarge",
  xxLarge: "xxLarge",
  xxxLarge: "xxxLarge",
  A11y_Medium: "A11y_Medium",
  A11y_Large: "A11y_Large",
  A11y_xLarge: "A11y_xLarge",
  A11y_xxLarge: "A11y_xxLarge",
  A11y_xxxLarge: "A11y_xxxLarge"
};
var androidFontScaleMap = {
  100: FontA11yCategory.Large,
  110: FontA11yCategory.xLarge,
  120: FontA11yCategory.xxLarge,
  135: FontA11yCategory.xxxLarge,
  160: FontA11yCategory.A11y_Medium,
  190: FontA11yCategory.A11y_Large,
  235: FontA11yCategory.A11y_xLarge,
  275: FontA11yCategory.A11y_xxLarge,
  310: FontA11yCategory.A11y_xxxLarge
};
var iosScaleToAndroidScale = {
  0.823: 100,
  0.882: 100,
  0.941: 100,
  1: 100,
  1.118: 110,
  1.235: 120,
  1.353: 135,
  1.786: 160,
  2.143: 190,
  2.643: 235,
  3.143: 275,
  3.571: 310
};
function convertToAndroidStyleScale(fontScale, platform) {
  if (platform === "android") {
    if (fontScale <= 1) {
      return 100;
    }
    const scaledValue = Math.round(fontScale * 100);
    const keys = Object.keys(androidFontScaleMap).map(Number).sort((a, b) => a - b);
    let closestKey = keys[0];
    let minDiff = Math.abs(scaledValue - closestKey);
    for (const key of keys) {
      const diff = Math.abs(scaledValue - key);
      if (diff < minDiff) {
        minDiff = diff;
        closestKey = key;
      }
    }
    return closestKey;
  } else {
    const iosScales = Object.keys(iosScaleToAndroidScale).map(Number);
    let closestScale = iosScales[0];
    let minDiff = Math.abs(fontScale - closestScale);
    for (const scale of iosScales) {
      const diff = Math.abs(fontScale - scale);
      if (diff < minDiff) {
        minDiff = diff;
        closestScale = scale;
      }
    }
    return iosScaleToAndroidScale[closestScale];
  }
}
function mapIOSFontScaleToCategory(fontScale) {
  if (fontScale < 1) {
    return FontA11yCategory.Large;
  }
  if (Math.abs(fontScale - 1) < 0.05) {
    return FontA11yCategory.Large;
  }
  if (Math.abs(fontScale - 1.118) < 0.05) {
    return FontA11yCategory.xLarge;
  }
  if (Math.abs(fontScale - 1.235) < 0.05) {
    return FontA11yCategory.xxLarge;
  }
  if (Math.abs(fontScale - 1.353) < 0.05) {
    return FontA11yCategory.xxxLarge;
  }
  if (Math.abs(fontScale - 1.786) < 0.05) {
    return FontA11yCategory.A11y_Medium;
  }
  if (Math.abs(fontScale - 2.143) < 0.05) {
    return FontA11yCategory.A11y_Large;
  }
  if (Math.abs(fontScale - 2.643) < 0.05) {
    return FontA11yCategory.A11y_xLarge;
  }
  if (Math.abs(fontScale - 3.143) < 0.05) {
    return FontA11yCategory.A11y_xxLarge;
  }
  if (Math.abs(fontScale - 3.571) < 0.05) {
    return FontA11yCategory.A11y_xxxLarge;
  }
  return FontA11yCategory.Large;
}
function mapAndroidFontScaleToCategory(fontScale) {
  if (fontScale <= 1) {
    return androidFontScaleMap[100];
  }
  const scaledValue = Math.round(fontScale * 100);
  const keys = Object.keys(androidFontScaleMap).map(Number).sort((a, b) => a - b);
  if (keys.length === 0) {
    return androidFontScaleMap[100];
  }
  let closestKey = keys[0];
  let minDiff = Math.abs(scaledValue - closestKey);
  for (const key of keys) {
    const diff = Math.abs(scaledValue - key);
    if (diff < minDiff) {
      minDiff = diff;
      closestKey = key;
    }
  }
  return androidFontScaleMap[closestKey];
}
function mapFontScaleToCategory(fontScale, platform) {
  return platform === "ios" ? mapIOSFontScaleToCategory(fontScale) : mapAndroidFontScaleToCategory(fontScale);
}
function useCreateUserAgent({
  batteryModePreference,
  colorPreference,
  locale,
  navbarPreference,
  pureSafeArea,
  safeArea,
  safeAreaBottomTransparency
}) {
  const platform = getPlatformOS2();
  const appVersion = getTossAppVersion2();
  const { fontScale } = useWindowDimensions();
  const platformString = platform === "ios" ? "iPhone" : "Android phone";
  const fontA11y = mapFontScaleToCategory(fontScale, platform);
  const normalizedFontScale = convertToAndroidStyleScale(fontScale, platform);
  return [
    "AppsInToss",
    `TossApp/${appVersion}`,
    batteryModePreference && `TossBatteryModePreference/${batteryModePreference}`,
    colorPreference && `TossColorPreference/${colorPreference}`,
    `TossFontAccessibility/${fontA11y}`,
    `TossFontScale/${normalizedFontScale}`,
    locale && `TossLocale/${locale}`,
    navbarPreference && `TossNavbarPreference/${navbarPreference}`,
    pureSafeArea && `TossPureSafeArea/${pureSafeArea}`,
    safeArea && `TossSafeArea/${safeArea}`,
    safeAreaBottomTransparency && `TossSafeAreaBottomTransparency/${safeAreaBottomTransparency}`,
    platformString
  ].filter(Boolean).join(" ");
}

// src/hooks/useGeolocation.ts
import { startUpdateLocation } from "@apps-in-toss/native-modules";
import { useVisibility } from "@granite-js/react-native";
import { useEffect as useEffect11, useState as useState5 } from "react";
function useGeolocation({ accuracy, distanceInterval, timeInterval }) {
  const isVisible = useVisibility();
  const [location, setLocation] = useState5(null);
  useEffect11(() => {
    if (!isVisible) {
      return;
    }
    return startUpdateLocation({
      options: {
        accuracy,
        distanceInterval,
        timeInterval
      },
      onEvent: setLocation,
      onError: console.error
    });
  }, [accuracy, distanceInterval, timeInterval, isVisible]);
  return location;
}

// src/hooks/useWaitForReturnNavigator.tsx
import { useNavigation as useNavigation2, useVisibilityChange } from "@granite-js/react-native";
import { useCallback as useCallback12, useRef as useRef4 } from "react";
function useWaitForReturnNavigator() {
  const callbacks = useRef4([]).current;
  const navigation = useNavigation2();
  const startNavigating = useCallback12(
    (route, params) => {
      return new Promise((resolve) => {
        callbacks.push(resolve);
        navigation.navigate(route, params);
      });
    },
    [callbacks, navigation]
  );
  const handleVisibilityChange = useCallback12(
    (state) => {
      if (state === "visible" && callbacks.length > 0) {
        for (const callback of callbacks) {
          callback();
        }
        callbacks.splice(0, callbacks.length);
      }
    },
    [callbacks]
  );
  useVisibilityChange(handleVisibilityChange);
  return startNavigating;
}

// src/hooks/useTopNavigation.tsx
function useTopNavigation() {
  const { addNavigationRightButton, removeNavigationRightButton } = useNavigationBarContext();
  return {
    addAccessoryButton: addNavigationRightButton,
    removeAccessoryButton: removeNavigationRightButton
  };
}

// src/hooks/useWebViewMemoryDebug.tsx
import { getOperationalEnvironment as getOperationalEnvironment8 } from "@apps-in-toss/native-modules";
import { Granite as Granite8 } from "@granite-js/react-native";
import { useToast } from "@toss/tds-react-native";
function useHandleWebViewProcessDidTerminate(webViewRef) {
  const toast = useToast();
  return () => {
    webViewDebugLog("AppsInTossWebViewProcessDidTerminate", {});
    toast.open("\uBB38\uC81C\uAC00 \uC0DD\uACA8\uC11C \uBBF8\uB2C8\uC571\uC744 \uB2E4\uC2DC \uC2DC\uC791\uD588\uC5B4\uC694.", {
      type: "bottom",
      lottie: `https://static.toss.im/lotties-common/check-green-spot.json`
    });
    webViewRef.current?.reload();
  };
}
async function webViewMemoryDebugLog(params) {
  webViewDebugLog("AppsInTossWebViewMemoryDebug", params);
}
function webViewDebugLog(logName, params) {
  const global2 = getAppsInTossGlobals();
  const event = {
    log_name: logName,
    log_type: "debug",
    params: {
      ...params,
      app_name: Granite8.appName,
      deployment_id: global2.deploymentId
    }
  };
  if (getOperationalEnvironment8() === "toss") {
    sendLog(event, { pipelines: ["core"] });
  } else {
    console.log("[webViewDebugLog]", event);
  }
}

// src/utils/log.ts
import { eventLog as eventLogNative } from "@apps-in-toss/native-modules";
import { getSchemeUri as getSchemeUri8 } from "@granite-js/react-native";

// src/utils/extractDateFromUUIDv7.ts
var extractDateFromUUIDv7 = (uuid) => {
  const timestampHex = uuid.split("-").join("").slice(0, 12);
  const timestamp = Number.parseInt(timestampHex, 16);
  return new Date(timestamp);
};

// src/utils/log.ts
var getGroupId = (url) => {
  try {
    const urlObject = new URL(url);
    return {
      groupId: urlObject.pathname,
      search: urlObject.search.startsWith("?") ? urlObject.search.substring(1) : urlObject.search
    };
  } catch {
    return {
      groupId: "unknown",
      search: "unknown"
    };
  }
};
var getReferrer2 = () => {
  try {
    const referrer = new URL(getSchemeUri8());
    return referrer.searchParams.get("referrer");
  } catch {
    return "";
  }
};
var trackScreen = (url, title) => {
  const { groupId, search } = getGroupId(url);
  const log = {
    log_type: "screen",
    log_name: `${groupId}::screen`,
    params: {
      search,
      referrer: getReferrer2(),
      document_title: title,
      deployment_id: env.getDeploymentId(),
      deployment_timestamp: extractDateFromUUIDv7(env.getDeploymentId()).getTime()
    }
  };
  return eventLogNative(log);
};

// src/components/WebView.tsx
import { jsx as jsx19 } from "react/jsx-runtime";
var operationalEnvironment = appsInTossConstantBridges.getOperationalEnvironment();
var TYPES = ["partner", "external", "game"];
var WEBVIEW_TYPES = {
  partner: PartnerWebView,
  external: ExternalWebViewScreen,
  game: GameWebView
};
function mergeSchemeQueryParamsInto(url) {
  const baseUrl = new URL(url);
  const schemeUrl = new URL(getSchemeUri9());
  const resultUrl = new URL(schemeUrl.pathname, baseUrl.origin);
  for (const [key, value] of baseUrl.searchParams.entries()) {
    resultUrl.searchParams.set(key, value);
  }
  for (const [key, value] of schemeUrl.searchParams.entries()) {
    resultUrl.searchParams.set(key, value);
  }
  return resultUrl;
}
function getWebViewURL(local) {
  if (__DEV__) {
    const devUrl = `http://${local.host}:${local.port}`;
    return mergeSchemeQueryParamsInto(devUrl);
  }
  const { url: rawUrl } = safeSyncPostMessage("getWebBundleURL", {});
  const url = mergeSchemeQueryParamsInto(rawUrl);
  const deploymentId = env.getDeploymentId();
  if (deploymentId) {
    url.searchParams.set("_deploymentId", deploymentId);
  }
  return url;
}
function WebView({ type, local, onMessage, ...props }) {
  if (!TYPES.includes(type)) {
    throw new Error(`Invalid WebView type: '${type}'`);
  }
  const webViewRef = useRef5(null);
  const url = useMemo6(() => getWebViewURL(local), [local]);
  const webBackHandler = useWebBackHandler(url, webViewRef);
  const top = useSafeAreaTop3();
  const bottom = useSafeAreaBottom();
  const insets = useSafeAreaInsets4();
  const global2 = getAppsInTossGlobals();
  const navigationBarContext = useNavigationBarContext();
  const safeAreaInsetsEvent = useSafeAreaInsetsEvent();
  const [allowsBackForwardNavigationGestures, setAllowsBackForwardNavigationGestures] = useState6(
    props.allowsBackForwardNavigationGestures
  );
  const handler = useBridgeHandler({
    onMessage,
    eventListenerMap: {
      ...appsInTossEventBridges,
      navigationAccessoryEvent: ({ onEvent, onError }) => tdsEvent.addEventListener("navigationAccessoryEvent", { onEvent, onError }),
      backEvent: ({ onEvent }) => {
        webBackHandler.addEventListener(onEvent);
        return () => {
          webBackHandler.removeEventListener(onEvent);
        };
      },
      homeEvent: ({ onEvent }) => homeEvent.subscribe(onEvent),
      updateLocationEvent: ({ onEvent, onError, options }) => appsInTossEvent.addEventListener("updateLocationEvent", { onEvent, onError, options }),
      safeAreaInsetsChange: ({ onEvent }) => safeAreaInsetsEvent.subscribe(onEvent),
      /** @internal */
      appBridgeCallbackEvent: ({ onEvent, onError, options }) => appsInTossEvent.addEventListener("appBridgeCallbackEvent", { onEvent, onError, options }),
      /** AdMobV2 */
      loadAppsInTossAdMob: GoogleAdMob2.loadAppsInTossAdMob,
      showAppsInTossAdMob: GoogleAdMob2.showAppsInTossAdMob,
      /** IntegratedAd */
      loadFullScreenAd: loadFullScreenAdForWeb,
      showFullScreenAd,
      /** TossAd */
      fetchTossAd,
      /** IAP */
      iapCreateOneTimePurchaseOrder: IAP.createOneTimePurchaseOrder,
      requestSubscriptionPurchase: IAP.createSubscriptionPurchaseOrder,
      requestOneTimePurchase
    },
    constantHandlerMap: {
      ...appsInTossConstantBridges,
      getSafeAreaInsets: () => ({
        top,
        bottom,
        left: insets.left,
        right: insets.right
      }),
      ...Object.fromEntries(Object.entries(global2).map(([key, value]) => [key, () => value])),
      getServerTime_isSupported: getServerTime.isSupported,
      requestReview_isSupported: requestReview.isSupported,
      /** AdMobV2 */
      loadAppsInTossAdMob_isSupported: GoogleAdMob2.loadAppsInTossAdMob.isSupported,
      showAppsInTossAdMob_isSupported: GoogleAdMob2.showAppsInTossAdMob.isSupported,
      isAppsInTossAdMobLoaded_isSupported: GoogleAdMob2.isAppsInTossAdMobLoaded.isSupported,
      /** IntegratedAd */
      loadFullScreenAd_isSupported: loadFullScreenAdForWeb.isSupported,
      showFullScreenAd_isSupported: showFullScreenAd.isSupported,
      /** TossAd */
      fetchTossAd_isSupported: fetchTossAd.isSupported,
      /** env */
      getDeploymentId: env.getDeploymentId,
      getWebViewType: env.getWebViewType,
      getAppName: env.getAppName
    },
    asyncHandlerMap: {
      ...appsInTossAsyncBridges,
      getServerTime,
      setIosSwipeGestureEnabled: (options) => {
        setAllowsBackForwardNavigationGestures(options.isEnabled);
        return appsInTossAsyncBridges.setIosSwipeGestureEnabled(options);
      },
      addAccessoryButton: async (params) => navigationBarContext.addNavigationRightButton(params),
      removeAccessoryButton: async () => navigationBarContext.removeNavigationRightButton(),
      /** permissions */
      requestPermission: appsInTossAsyncBridges.requestPermission,
      openPermissionDialog: appsInTossAsyncBridges.openPermissionDialog,
      getPermission: appsInTossAsyncBridges.getPermission,
      getClipboardText: appsInTossAsyncBridges.getClipboardText,
      setClipboardText: appsInTossAsyncBridges.setClipboardText,
      fetchAlbumPhotos: appsInTossAsyncBridges.fetchAlbumPhotos,
      fetchContacts: appsInTossAsyncBridges.fetchContacts,
      getCurrentLocation: appsInTossAsyncBridges.getCurrentLocation,
      openCamera: appsInTossAsyncBridges.openCamera,
      /** Storage */
      getStorageItem: Storage.getItem,
      setStorageItem: Storage.setItem,
      removeStorageItem: Storage.removeItem,
      clearItems: Storage.clearItems,
      /** IAP */
      iapGetProductItemList: IAP.getProductItemList,
      iapCreateOneTimePurchaseOrder,
      processProductGrant,
      getPendingOrders: IAP.getPendingOrders,
      getCompletedOrRefundedOrders: IAP.getCompletedOrRefundedOrders,
      completeProductGrant: IAP.completeProductGrant,
      getSubscriptionInfo: IAP.getSubscriptionInfo,
      /** AdMobV2 */
      isAppsInTossAdMobLoaded: GoogleAdMob2.isAppsInTossAdMobLoaded,
      /** Toss Ads */
      tossAdEventLog,
      /** Private */
      memoryDebugLog: webViewMemoryDebugLog,
      debugLog: async (event) => {
        sendLog(event);
      }
    }
  });
  const headerPropForExternalWebView = useMemo6(() => {
    const parsedNavigationBar = global2.navigationBar != null ? safeParseNavigationBar(global2.navigationBar) : null;
    const initialAccessoryButton = parsedNavigationBar?.initialAccessoryButton;
    const withBackButton = parsedNavigationBar?.withBackButton ?? true;
    if (type === "external" && initialAccessoryButton != null) {
      return {
        header: {
          withBackButton,
          title: initialAccessoryButton.title,
          icon: initialAccessoryButton.icon
        }
      };
    }
    return {};
  }, [global2.navigationBar, type]);
  const BaseWebView = WEBVIEW_TYPES[type];
  const webViewDebuggingEnabled = operationalEnvironment === "sandbox";
  const userAgent = useCreateUserAgent({
    colorPreference: "light"
  });
  const refs = mergeRefs(handler.ref, webViewRef);
  useEffect12(() => {
    const callback = () => {
      webBackHandler.handleWebBack();
      return true;
    };
    const subscription = BackHandler2.addEventListener("hardwareBackPress", callback);
    return () => subscription.remove();
  }, [webBackHandler]);
  const handleWebViewProcessDidTerminate = useHandleWebViewProcessDidTerminate(webViewRef);
  return /* @__PURE__ */ jsx19(
    BaseWebView,
    {
      ref: refs,
      ...props,
      ...headerPropForExternalWebView,
      source: {
        uri: url.href,
        // NOTE: https://github.com/react-native-webview/react-native-webview/pull/3133
        headers: {
          "User-Agent": userAgent
        }
      },
      onHomeButtonClick: webBackHandler.handleWebHome,
      onBackButtonClick: webBackHandler.handleWebBack,
      onNavigationStateChange: (event) => {
        if (event.url) {
          trackScreen(event.url, event.title);
        }
        props.onNavigationStateChange?.(event);
        webBackHandler.onNavigationStateChange(event);
      },
      textZoom: 100,
      userAgent: Platform6.OS === "ios" ? userAgent : void 0,
      sharedCookiesEnabled: true,
      webviewDebuggingEnabled: webViewDebuggingEnabled,
      thirdPartyCookiesEnabled: true,
      onMessage: handler.onMessage,
      injectedJavaScript: afterDocumentLoad,
      injectedJavaScriptBeforeContentLoaded: [handler.injectedJavaScript, beforeDocumentLoad].join("\n"),
      decelerationRate: Platform6.OS === "ios" ? 1 : void 0,
      allowsBackForwardNavigationGestures,
      onShouldStartLoadWithRequest: (event) => {
        try {
          const url2 = new URL(event.url);
          if (["https:", "http:"].includes(url2.protocol) || url2.href === "about:blank") {
            return true;
          } else {
            Linking.openURL(convertIntentURL(url2) ?? url2.href);
            return false;
          }
        } catch (error) {
          console.error(error);
          return false;
        }
      },
      originWhitelist: [
        "https://*",
        "http://*",
        "intoss://*",
        "intoss-private://*",
        "servicetoss://*",
        "supertoss://*",
        "intent://*"
      ],
      onContentProcessDidTerminate: handleWebViewProcessDidTerminate,
      onRenderProcessGone: handleWebViewProcessDidTerminate
    }
  );
}

// src/index.ts
export * from "@apps-in-toss/analytics";
import { useOverlay as useOverlay3, OverlayProvider } from "@toss/tds-react-native/private";
export * from "@apps-in-toss/native-modules";
export * from "@apps-in-toss/types";
import { StartUpdateLocationPermissionError } from "@apps-in-toss/types";

// src/ads/inlineAd/InlineAd.tsx
import { ImpressionArea, useVisibilityChange as useVisibilityChange2 } from "@granite-js/react-native";
import { useEffect as useEffect13, useRef as useRef8, useState as useState7 } from "react";
import { StyleSheet as StyleSheet3, View as View7, useColorScheme } from "react-native";

// src/ads/inlineAd/constants.ts
var SDK_ID = "106";
var LIST_BANNER_STYLE_ID = "1";
var NATIVE_IMAGE_STYLE_ID = "2";
var FEED_BANNER_STYLE_ID = NATIVE_IMAGE_STYLE_ID;
var AVAILABLE_STYLE_IDS = [LIST_BANNER_STYLE_ID, NATIVE_IMAGE_STYLE_ID];
var DEFAULT_INLINE_AD_THEME = "auto";
var DEFAULT_INLINE_AD_TONE = "blackAndWhite";
var DEFAULT_INLINE_AD_VARIANT = "expanded";
var DEFAULT_INLINE_AD_PADDING_BANNER = "16px 20px";
var DEFAULT_INLINE_AD_PADDING_NATIVE_IMAGE = "20px";
var ERROR_CODES = {
  NETWORK_ERROR: 1001,
  INVALID_REQUEST: 1002,
  NO_AD: 1003,
  ERROR: 1004,
  INVALID_SPACE: 1005,
  SDK_NOT_INITIALIZED: 1007,
  HTTP_TIMEOUT: 1008,
  EXECUTION_FAIL: 1009,
  INTERRUPTED: 1010,
  FAIL: 1011,
  BLOCKED: 1012,
  TIMEOUT: 1013,
  LIMITED_AD: 1014,
  CONSENT_REQUIRED: 1015,
  TEST_MODE: 1016,
  INTERNAL_ERROR: 1017
};
var ERROR_MESSAGES = {
  [ERROR_CODES.NETWORK_ERROR]: "Network error occurred",
  [ERROR_CODES.INVALID_REQUEST]: "Invalid request parameters",
  [ERROR_CODES.NO_AD]: "No ads available",
  [ERROR_CODES.ERROR]: "Server error occurred",
  [ERROR_CODES.INVALID_SPACE]: "Invalid or disabled inventory",
  [ERROR_CODES.SDK_NOT_INITIALIZED]: "SDK is not initialized",
  [ERROR_CODES.HTTP_TIMEOUT]: "Request timed out",
  [ERROR_CODES.EXECUTION_FAIL]: "Request execution failed",
  [ERROR_CODES.INTERRUPTED]: "Request interrupted",
  [ERROR_CODES.FAIL]: "Request failed",
  [ERROR_CODES.BLOCKED]: "Ad blocked by policy",
  [ERROR_CODES.TIMEOUT]: "SSP processing timed out",
  [ERROR_CODES.LIMITED_AD]: "Ad limited by privacy",
  [ERROR_CODES.CONSENT_REQUIRED]: "Consent required to show ads",
  [ERROR_CODES.TEST_MODE]: "Test mode only",
  [ERROR_CODES.INTERNAL_ERROR]: "Unknown error occurred"
};
var API_RESULT_ERROR_MAP = {
  HTTP_TIMEOUT: {
    code: ERROR_CODES.HTTP_TIMEOUT,
    message: ERROR_MESSAGES[ERROR_CODES.HTTP_TIMEOUT]
  },
  NETWORK_ERROR: {
    code: ERROR_CODES.NETWORK_ERROR,
    message: ERROR_MESSAGES[ERROR_CODES.NETWORK_ERROR]
  },
  EXECUTION_FAIL: {
    code: ERROR_CODES.EXECUTION_FAIL,
    message: ERROR_MESSAGES[ERROR_CODES.EXECUTION_FAIL]
  },
  INTERRUPTED: {
    code: ERROR_CODES.INTERRUPTED,
    message: ERROR_MESSAGES[ERROR_CODES.INTERRUPTED]
  },
  INTERNAL_ERROR: {
    code: ERROR_CODES.ERROR,
    message: ERROR_MESSAGES[ERROR_CODES.ERROR]
  },
  FAIL: {
    code: ERROR_CODES.FAIL,
    message: ERROR_MESSAGES[ERROR_CODES.FAIL]
  }
};
var AD_STATUS_ERROR_MAP = {
  NO_AD: {
    code: ERROR_CODES.NO_AD,
    message: ERROR_MESSAGES[ERROR_CODES.NO_AD]
  },
  BLOCKED: {
    code: ERROR_CODES.BLOCKED,
    message: ERROR_MESSAGES[ERROR_CODES.BLOCKED]
  },
  ERROR: {
    code: ERROR_CODES.ERROR,
    message: ERROR_MESSAGES[ERROR_CODES.ERROR]
  },
  TIMEOUT: {
    code: ERROR_CODES.TIMEOUT,
    message: ERROR_MESSAGES[ERROR_CODES.TIMEOUT]
  },
  INVALID_SPACE: {
    code: ERROR_CODES.INVALID_SPACE,
    message: ERROR_MESSAGES[ERROR_CODES.INVALID_SPACE]
  },
  INVALID_REQUEST: {
    code: ERROR_CODES.INVALID_REQUEST,
    message: ERROR_MESSAGES[ERROR_CODES.INVALID_REQUEST]
  },
  LIMITED_AD: {
    code: ERROR_CODES.LIMITED_AD,
    message: ERROR_MESSAGES[ERROR_CODES.LIMITED_AD]
  },
  CONSENT_REQUIRED: {
    code: ERROR_CODES.CONSENT_REQUIRED,
    message: ERROR_MESSAGES[ERROR_CODES.CONSENT_REQUIRED]
  },
  TEST_MODE: {
    code: ERROR_CODES.TEST_MODE,
    message: ERROR_MESSAGES[ERROR_CODES.TEST_MODE]
  }
};

// src/ads/inlineAd/loadAd.ts
function createError(code, message) {
  return {
    code,
    message: message || ERROR_MESSAGES[code] || "Unknown error",
    domain: "@apps-in-toss/framework"
  };
}
function isAdError(error) {
  return Boolean(
    error && typeof error === "object" && "code" in error && typeof error.code === "number" && "message" in error && typeof error.message === "string"
  );
}
function fetchTossAdPromise(options) {
  return new Promise((resolve, reject) => {
    if (!fetchTossAd.isSupported()) {
      reject(new Error("fetchTossAd is not supported in this environment."));
      return;
    }
    fetchTossAd({
      options,
      onEvent: resolve,
      onError: reject
    });
  });
}
function isApiResponse(payload) {
  return Boolean(payload && typeof payload === "object" && "resultType" in payload);
}
function isAdResponse(payload) {
  return Boolean(payload && typeof payload === "object" && "ads" in payload);
}
function normalizeAdResponse(adResponse) {
  const ads = Array.isArray(adResponse.ads) ? adResponse.ads.map((ad) => ({
    ...ad,
    requestId: ad.requestId ?? adResponse.requestId
  })).filter((ad) => AVAILABLE_STYLE_IDS.includes(String(ad.styleId))) : [];
  return {
    requestId: adResponse.requestId ?? "",
    status: adResponse.status ?? "OK",
    ads,
    ext: adResponse.ext
  };
}
function normalizeApiResponse(raw) {
  if (isApiResponse(raw)) {
    if (raw.resultType !== "SUCCESS") {
      return raw;
    }
    if (!raw.success) {
      return {
        resultType: "FAIL",
        error: { reason: "fetchTossAd returned SUCCESS without payload" }
      };
    }
    return { ...raw, success: normalizeAdResponse(raw.success) };
  }
  if (isAdResponse(raw)) {
    return {
      resultType: "SUCCESS",
      success: normalizeAdResponse(raw)
    };
  }
  return {
    resultType: "FAIL",
    error: { reason: "Invalid response from fetchTossAd" }
  };
}
function isStructurallyValidAd(ad) {
  return Boolean(ad && ad.creative && typeof ad.creative === "object");
}
async function loadAd(adGroupId) {
  const normalizedAdGroupId = normalizeAdGroupId(adGroupId);
  if (normalizedAdGroupId.length === 0) {
    return {
      type: "error",
      error: createError(ERROR_CODES.INVALID_SPACE, INVALID_AD_GROUP_ID_ERROR_MESSAGE)
    };
  }
  if (!fetchTossAd.isSupported()) {
    return {
      type: "error",
      error: createError(ERROR_CODES.SDK_NOT_INITIALIZED, "This feature is not supported in the current environment")
    };
  }
  let response;
  try {
    const raw = await fetchTossAdPromise({
      adGroupId: normalizedAdGroupId,
      sdkId: SDK_ID,
      availableStyleIds: AVAILABLE_STYLE_IDS
    });
    response = normalizeApiResponse(raw);
  } catch (error) {
    if (isAdError(error)) {
      return { type: "error", error };
    }
    const message = error instanceof Error ? error.message : "Network error";
    return {
      type: "error",
      error: createError(ERROR_CODES.NETWORK_ERROR, message)
    };
  }
  if (response.resultType !== "SUCCESS") {
    const mappedError = API_RESULT_ERROR_MAP[response.resultType];
    const errorMessage = response.error?.reason ?? mappedError?.message ?? ERROR_MESSAGES[ERROR_CODES.INTERNAL_ERROR];
    const errorCode = mappedError?.code ?? ERROR_CODES.INTERNAL_ERROR;
    return {
      type: "error",
      error: createError(errorCode, errorMessage)
    };
  }
  if (!response.success) {
    return {
      type: "error",
      error: createError(ERROR_CODES.INTERNAL_ERROR, "Missing ad response payload")
    };
  }
  const adResponse = response.success;
  if (adResponse.status !== "OK" && adResponse.status !== "TEST_MODE") {
    const statusError = AD_STATUS_ERROR_MAP[adResponse.status];
    const errorMessage = statusError?.message ?? `Ad response status: ${adResponse.status}`;
    const errorCode = statusError?.code ?? ERROR_CODES.INTERNAL_ERROR;
    const error = createError(errorCode, errorMessage);
    if (error.code === ERROR_CODES.NO_AD) {
      return { type: "noFill", error, response: adResponse };
    }
    return { type: "error", error, response: adResponse };
  }
  const topPriorityAd = adResponse.ads[0];
  const selectedAd = isStructurallyValidAd(topPriorityAd) ? topPriorityAd : adResponse.ads.find(isStructurallyValidAd) ?? null;
  if (!selectedAd) {
    return {
      type: "error",
      error: createError(ERROR_CODES.INTERNAL_ERROR, "Unsupported styleId in response"),
      response: adResponse
    };
  }
  return {
    type: "success",
    ad: selectedAd,
    responseExt: adResponse.ext ?? null,
    requestId: adResponse.requestId,
    response: adResponse
  };
}

// src/ads/inlineAd/opener.ts
import { openURL as openURL6 } from "@granite-js/react-native";
function isSafeUrl(url) {
  try {
    const parsed = new URL(url);
    const dangerous = [
      "javascript:",
      "data:",
      "vbscript:",
      "file:",
      "about:",
      "blob:",
      "filesystem:",
      "chrome:",
      "chrome-extension:",
      "ftp:",
      "ftps:",
      "sftp:",
      "telnet:",
      "ssh:",
      "intent:"
    ];
    return !dangerous.includes(parsed.protocol.toLowerCase());
  } catch {
    return false;
  }
}
function openLandingUrl(url) {
  if (!isSafeUrl(url)) {
    console.warn("[InlineAd] Blocked unsafe landing URL:", url);
    return;
  }
  try {
    openURL6(getWebSchemeOrUri(url));
  } catch (error) {
    console.error("[InlineAd] Failed to open landing URL:", error);
  }
}
function getWebSchemeOrUri(uri) {
  const isHttp = ["http://", "https://"].some((protocol) => uri.startsWith(protocol));
  return isHttp ? supertossWeb(uri) : uri;
}
function supertossWeb(uri) {
  return `supertoss://web?url=${encodeURIComponent(uri)}&external=true`;
}

// src/ads/inlineAd/tracking.ts
var GlobalEventDeduplicator = class _GlobalEventDeduplicator {
  static MAX_EVENTS = 1e4;
  trackedEvents = /* @__PURE__ */ new Set();
  hasTracked(requestId, creativeId, eventType) {
    return this.trackedEvents.has(this.buildKey(requestId, creativeId, eventType));
  }
  markTracked(requestId, creativeId, eventType) {
    if (this.trackedEvents.size >= _GlobalEventDeduplicator.MAX_EVENTS) {
      const firstKey = this.trackedEvents.keys().next().value;
      if (firstKey) {
        this.trackedEvents.delete(firstKey);
      }
    }
    this.trackedEvents.add(this.buildKey(requestId, creativeId, eventType));
  }
  buildKey(requestId, creativeId, eventType) {
    return `${requestId}:${creativeId}:${eventType}`;
  }
};
var globalEventDeduplicator = new GlobalEventDeduplicator();
var EventTracker = class {
  constructor(eventTrackingUrls = [], eventTypes = [], eventPayload = "", requestId, creativeId, requestHeaders) {
    this.eventTrackingUrls = eventTrackingUrls;
    this.eventPayload = eventPayload;
    this.requestId = requestId;
    this.creativeId = creativeId;
    this.requestHeaders = requestHeaders;
    this.eventTypes = new Set(eventTypes);
    this.repeatableEvents = /* @__PURE__ */ new Set(["VIEW_MUTE", "VIEW_UNMUTE"]);
  }
  eventTypes;
  repeatableEvents;
  async track(eventType) {
    if (!this.eventTypes.has(eventType)) {
      return;
    }
    if (!this.repeatableEvents.has(eventType)) {
      if (this.requestId && this.creativeId) {
        if (globalEventDeduplicator.hasTracked(this.requestId, this.creativeId, eventType)) {
          return;
        }
      }
    }
    if (this.eventTrackingUrls.length === 0) {
      return;
    }
    const payload = {
      type: eventType,
      data: this.eventPayload
    };
    await Promise.allSettled(this.eventTrackingUrls.map((url) => this.sendTrackingRequest(url, payload)));
    if (!this.repeatableEvents.has(eventType) && this.requestId && this.creativeId) {
      globalEventDeduplicator.markTracked(this.requestId, this.creativeId, eventType);
    }
  }
  async sendTrackingRequest(url, payload) {
    const headers = {
      "Content-Type": "application/json",
      ...this.requestHeaders ?? {}
    };
    if (this.requestId) {
      headers["X-Toss-RequestId"] = this.requestId;
    }
    await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(payload)
    });
  }
};

// src/ads/inlineAd/ui/FeedBannerAdView.tsx
import { Path, Svg } from "@granite-js/native/react-native-svg";
import { useRef as useRef6 } from "react";
import {
  Animated,
  Easing,
  Image,
  Pressable as Pressable2,
  StyleSheet,
  Text,
  View as View5
} from "react-native";
import { jsx as jsx20, jsxs as jsxs9 } from "react/jsx-runtime";
var ARROW_PATH = "m7.5 20.4c-.5-.5-.5-1.2 0-1.7l6.7-6.7-6.8-6.7c-.5-.5-.5-1.2 0-1.7s1.2-.5 1.7 0l7.5 7.5c.5.5.5 1.2 0 1.7l-7.5 7.5c-.2.3-.5.4-.8.4s-.6-.1-.8-.3z";
function FeedBannerAdView({
  brandName,
  brandLogoUri,
  title,
  subtitle,
  mainImageUri,
  ctaText,
  ctaTextColor,
  ctaBackgroundColor,
  adClearanceText,
  colors,
  isAdBadgeEnabled,
  paddingStyle,
  onPress
}) {
  const scale = useRef6(new Animated.Value(1)).current;
  const animateScale = (toValue) => {
    Animated.timing(scale, {
      toValue,
      duration: 100,
      easing: Easing.inOut(Easing.ease),
      useNativeDriver: true
    }).start();
  };
  const handlePressIn = () => {
    animateScale(0.98);
  };
  const handlePressOut = () => {
    animateScale(1);
  };
  const resolvedCtaBackground = ctaBackgroundColor ?? "#3081F9";
  const resolvedCtaTextColor = ctaTextColor ?? "#ffffff";
  return /* @__PURE__ */ jsx20(
    Pressable2,
    {
      accessibilityRole: "button",
      onPress: () => onPress(null),
      onPressIn: handlePressIn,
      onPressOut: handlePressOut,
      style: styles.pressable,
      children: /* @__PURE__ */ jsxs9(Animated.View, { style: [styles.container, paddingStyle, { transform: [{ scale }] }], children: [
        /* @__PURE__ */ jsxs9(View5, { style: styles.profileContainer, children: [
          /* @__PURE__ */ jsxs9(View5, { style: styles.brandArea, children: [
            /* @__PURE__ */ jsx20(
              Pressable2,
              {
                accessibilityRole: "button",
                onPress: () => onPress("202"),
                onPressIn: handlePressIn,
                onPressOut: handlePressOut,
                style: styles.logoContainer,
                children: /* @__PURE__ */ jsxs9(View5, { style: [styles.logoWrapper, { backgroundColor: colors.brandLogoBg }], children: [
                  brandLogoUri ? /* @__PURE__ */ jsx20(Image, { source: { uri: brandLogoUri }, style: styles.logoImage, resizeMode: "cover" }) : null,
                  /* @__PURE__ */ jsx20(View5, { style: [styles.logoOverlay, { borderColor: colors.brandLogoBorder }] })
                ] })
              }
            ),
            /* @__PURE__ */ jsxs9(View5, { style: styles.brandTextContainer, children: [
              /* @__PURE__ */ jsx20(
                Pressable2,
                {
                  accessibilityRole: "button",
                  onPress: () => onPress("103"),
                  onPressIn: handlePressIn,
                  onPressOut: handlePressOut,
                  children: /* @__PURE__ */ jsx20(
                    Text,
                    {
                      allowFontScaling: false,
                      numberOfLines: 1,
                      style: [styles.brandName, { color: colors.brandName }],
                      children: brandName
                    }
                  )
                }
              ),
              isAdBadgeEnabled ? /* @__PURE__ */ jsx20(Text, { allowFontScaling: false, numberOfLines: 1, style: [styles.adBadge, { color: colors.adBadge }], children: "\uAD11\uACE0" }) : null
            ] })
          ] }),
          /* @__PURE__ */ jsxs9(View5, { style: styles.textArea, children: [
            /* @__PURE__ */ jsx20(
              Pressable2,
              {
                accessibilityRole: "button",
                onPress: () => onPress("101"),
                onPressIn: handlePressIn,
                onPressOut: handlePressOut,
                children: /* @__PURE__ */ jsx20(Text, { allowFontScaling: false, style: [styles.title, { color: colors.title }], children: title })
              }
            ),
            /* @__PURE__ */ jsx20(
              Pressable2,
              {
                accessibilityRole: "button",
                onPress: () => onPress("102"),
                onPressIn: handlePressIn,
                onPressOut: handlePressOut,
                children: /* @__PURE__ */ jsx20(Text, { allowFontScaling: false, style: [styles.subtitle, { color: colors.subtitle }], children: subtitle })
              }
            )
          ] })
        ] }),
        /* @__PURE__ */ jsxs9(View5, { style: styles.card, children: [
          /* @__PURE__ */ jsx20(
            Pressable2,
            {
              accessibilityRole: "button",
              onPress: () => onPress("201"),
              onPressIn: handlePressIn,
              onPressOut: handlePressOut,
              style: styles.imageButton,
              children: /* @__PURE__ */ jsx20(View5, { style: styles.imageContainer, children: mainImageUri ? /* @__PURE__ */ jsx20(Image, { source: { uri: mainImageUri }, style: styles.mainImage, resizeMode: "cover" }) : null })
            }
          ),
          /* @__PURE__ */ jsxs9(
            Pressable2,
            {
              accessibilityRole: "button",
              onPress: () => onPress("0"),
              onPressIn: handlePressIn,
              onPressOut: handlePressOut,
              style: [styles.cta, { backgroundColor: resolvedCtaBackground }],
              children: [
                /* @__PURE__ */ jsx20(Text, { allowFontScaling: false, style: [styles.ctaText, { color: resolvedCtaTextColor }], children: ctaText }),
                /* @__PURE__ */ jsx20(Svg, { width: 20, height: 20, viewBox: "0 0 24 24", children: /* @__PURE__ */ jsx20(Path, { d: ARROW_PATH, fill: resolvedCtaTextColor }) })
              ]
            }
          ),
          /* @__PURE__ */ jsx20(View5, { style: [styles.cardOverlay, { borderColor: colors.imageOverlayBorder }] })
        ] }),
        adClearanceText ? /* @__PURE__ */ jsx20(Text, { allowFontScaling: false, style: [styles.adClearance, { color: colors.adClearance }], children: adClearanceText }) : null
      ] })
    }
  );
}
var styles = StyleSheet.create({
  pressable: {
    width: "100%"
  },
  container: {
    flexDirection: "column",
    padding: 0,
    width: "100%"
  },
  profileContainer: {
    flexDirection: "column",
    paddingBottom: 8,
    gap: 12
  },
  brandArea: {
    flexDirection: "row",
    alignItems: "center",
    gap: 8
  },
  logoContainer: {
    width: 40,
    height: 40,
    alignItems: "center",
    justifyContent: "center"
  },
  logoWrapper: {
    width: 36,
    height: 36,
    borderRadius: 18,
    overflow: "hidden",
    alignItems: "center",
    justifyContent: "center"
  },
  logoImage: {
    width: 36,
    height: 36,
    borderRadius: 18
  },
  logoOverlay: {
    ...StyleSheet.absoluteFillObject,
    borderWidth: 1,
    borderRadius: 18
  },
  brandTextContainer: {
    flexDirection: "column"
  },
  brandName: {
    fontSize: 15,
    fontWeight: "600",
    lineHeight: 22.5
  },
  adBadge: {
    fontSize: 13,
    fontWeight: "400",
    lineHeight: 19.5
  },
  textArea: {
    flexDirection: "column",
    alignItems: "flex-start"
  },
  title: {
    fontSize: 15,
    fontWeight: "600",
    lineHeight: 22.5
  },
  subtitle: {
    fontSize: 15,
    fontWeight: "400",
    lineHeight: 22.5
  },
  card: {
    borderRadius: 12,
    overflow: "hidden",
    position: "relative",
    width: "100%"
  },
  imageButton: {
    width: "100%"
  },
  imageContainer: {
    width: "100%",
    aspectRatio: 16 / 9,
    overflow: "hidden"
  },
  mainImage: {
    width: "100%",
    height: "100%"
  },
  cta: {
    height: 50,
    paddingLeft: 20,
    paddingRight: 16,
    alignItems: "center",
    justifyContent: "space-between",
    flexDirection: "row"
  },
  ctaText: {
    fontSize: 15,
    fontWeight: "600",
    lineHeight: 22.5
  },
  cardOverlay: {
    ...StyleSheet.absoluteFillObject,
    borderWidth: 2,
    borderRadius: 12,
    pointerEvents: "none"
  },
  adClearance: {
    marginTop: 10,
    fontSize: 8,
    fontWeight: "400"
  }
});

// src/ads/inlineAd/ui/ListBannerAdView.tsx
import { ClipPath, Defs, Image as SvgImage, Path as Path2, Svg as Svg2 } from "@granite-js/native/react-native-svg";
import { useRef as useRef7 } from "react";
import { Animated as Animated2, Easing as Easing2, Pressable as Pressable3, StyleSheet as StyleSheet2, Text as Text2, View as View6 } from "react-native";
import { jsx as jsx21, jsxs as jsxs10 } from "react/jsx-runtime";
var SQUIRCLE_PATH = "M0 17.352C0 3.564 3.564 0 17.352 0H18.648C32.436 0 36 3.564 36 17.352V18.648C36 32.436 32.436 36 18.648 36H17.352C3.564 36 0 32.436 0 18.648V17.352Z";
function ListBannerAdView({
  title,
  subtitle,
  adClearanceText,
  adClearanceFontSize,
  imageUri,
  paddingStyle,
  colors,
  onPress
}) {
  const scale = useRef7(new Animated2.Value(1)).current;
  const clipIdRef = useRef7(`clip-${Math.random().toString(36).slice(2)}`);
  const animateScale = (toValue) => {
    Animated2.timing(scale, {
      toValue,
      duration: 100,
      easing: Easing2.inOut(Easing2.ease),
      useNativeDriver: true
    }).start();
  };
  const handlePressIn = () => {
    animateScale(0.96);
  };
  const handlePressOut = () => {
    animateScale(1);
  };
  return /* @__PURE__ */ jsx21(
    Pressable3,
    {
      accessibilityRole: "button",
      onPress: () => onPress(null),
      onPressIn: handlePressIn,
      onPressOut: handlePressOut,
      style: styles2.pressable,
      children: /* @__PURE__ */ jsxs10(Animated2.View, { style: [styles2.container, paddingStyle, { transform: [{ scale }] }], children: [
        /* @__PURE__ */ jsxs10(View6, { style: styles2.titleRow, children: [
          /* @__PURE__ */ jsx21(
            Pressable3,
            {
              accessibilityRole: "button",
              onPress: () => onPress("202"),
              onPressIn: handlePressIn,
              onPressOut: handlePressOut,
              style: styles2.iconWrapper,
              children: /* @__PURE__ */ jsxs10(Svg2, { width: 36, height: 36, viewBox: "0 0 36 36", children: [
                /* @__PURE__ */ jsx21(Defs, { children: /* @__PURE__ */ jsx21(ClipPath, { id: clipIdRef.current, children: /* @__PURE__ */ jsx21(Path2, { d: SQUIRCLE_PATH }) }) }),
                /* @__PURE__ */ jsx21(Path2, { d: SQUIRCLE_PATH, fill: colors.iconBg }),
                imageUri ? /* @__PURE__ */ jsx21(
                  SvgImage,
                  {
                    href: { uri: imageUri },
                    width: 36,
                    height: 36,
                    preserveAspectRatio: "xMidYMid slice",
                    clipPath: `url(#${clipIdRef.current})`
                  }
                ) : null,
                /* @__PURE__ */ jsx21(
                  Path2,
                  {
                    d: SQUIRCLE_PATH,
                    fill: "none",
                    stroke: colors.iconBorder,
                    strokeWidth: 2,
                    clipPath: `url(#${clipIdRef.current})`
                  }
                )
              ] })
            }
          ),
          /* @__PURE__ */ jsxs10(View6, { style: styles2.textWrapper, children: [
            /* @__PURE__ */ jsx21(
              Pressable3,
              {
                accessibilityRole: "button",
                onPress: () => onPress("101"),
                onPressIn: handlePressIn,
                onPressOut: handlePressOut,
                children: /* @__PURE__ */ jsx21(Text2, { allowFontScaling: false, style: [styles2.title, { color: colors.title }], children: title })
              }
            ),
            /* @__PURE__ */ jsx21(
              Pressable3,
              {
                accessibilityRole: "button",
                onPress: () => onPress("102"),
                onPressIn: handlePressIn,
                onPressOut: handlePressOut,
                children: /* @__PURE__ */ jsx21(Text2, { allowFontScaling: false, style: [styles2.subtitle, { color: colors.subtitle }], children: subtitle })
              }
            )
          ] })
        ] }),
        adClearanceText ? /* @__PURE__ */ jsx21(
          Text2,
          {
            allowFontScaling: false,
            style: [
              styles2.adClearance,
              {
                color: colors.adClearance,
                fontSize: adClearanceFontSize,
                lineHeight: adClearanceFontSize * 1.2
              }
            ],
            children: adClearanceText
          }
        ) : null
      ] })
    }
  );
}
var styles2 = StyleSheet2.create({
  pressable: {
    width: "100%"
  },
  container: {
    flexDirection: "column",
    padding: 0,
    width: "100%"
  },
  titleRow: {
    flexDirection: "row",
    alignItems: "center"
  },
  iconWrapper: {
    width: 36,
    height: 36,
    overflow: "hidden",
    alignItems: "center",
    justifyContent: "center",
    marginRight: 12
  },
  textWrapper: {
    flex: 1,
    flexDirection: "column",
    alignItems: "flex-start",
    justifyContent: "center",
    width: "100%"
  },
  title: {
    fontSize: 17,
    fontWeight: "700",
    lineHeight: 25.5
  },
  subtitle: {
    fontSize: 13,
    fontWeight: "500",
    flexShrink: 1,
    lineHeight: 15.6
  },
  adClearance: {
    marginTop: 8,
    paddingLeft: 48,
    width: "100%"
  }
});

// src/ads/inlineAd/utils/ids.ts
var slotCounter = 0;
function createSlotId() {
  slotCounter += 1;
  return `banner-slot-${Date.now()}-${slotCounter}`;
}

// src/ads/inlineAd/utils/padding.ts
function resolvePaddingStyle(padding) {
  if (padding === void 0) {
    return void 0;
  }
  if (typeof padding === "number") {
    return { padding };
  }
  const trimmed = padding.trim();
  if (!trimmed) {
    return void 0;
  }
  const parts = trimmed.split(/\s+/);
  if (parts.length < 1 || parts.length > 4) {
    return void 0;
  }
  const values = parts.map((part) => {
    const normalized = part.endsWith("px") ? part.slice(0, -2).trim() : part;
    const numeric = Number(normalized);
    return Number.isFinite(numeric) ? numeric : null;
  });
  if (values.some((value) => value === null)) {
    return void 0;
  }
  const [top, right, bottom, left] = normalizePaddingValues(values);
  return {
    paddingTop: top,
    paddingRight: right,
    paddingBottom: bottom,
    paddingLeft: left
  };
}
function normalizePaddingValues(values) {
  const [top = 0, right = top, bottom = top, left = right] = values;
  return [top, right, bottom, left];
}

// src/ads/inlineAd/utils/payload.ts
function buildBannerEventPayload(slotId, adGroupId, ad) {
  return {
    slotId,
    adGroupId,
    adMetadata: {
      creativeId: ad.creative?.id ?? "",
      requestId: ad.requestId ?? "",
      styleId: ad.styleId ?? LIST_BANNER_STYLE_ID
    }
  };
}
function buildBannerErrorPayload(slotId, adGroupId, error) {
  return {
    slotId,
    adGroupId,
    adMetadata: {},
    error
  };
}
function buildNoFillPayload(slotId, adGroupId) {
  return {
    slotId,
    adGroupId,
    adMetadata: {}
  };
}

// src/ads/inlineAd/InlineAd.tsx
import { jsx as jsx22, jsxs as jsxs11 } from "react/jsx-runtime";
var themePalette = {
  light: {
    title: "#4e5968",
    subtitle: "#6b7684",
    adClearance: "#b0b8c1",
    iconBg: "rgba(2,32,71,0.05)",
    iconBorder: "rgba(2,32,71,0.05)"
  },
  dark: {
    title: "rgba(253,253,255,0.75)",
    subtitle: "rgba(248,248,255,0.6)",
    adClearance: "rgba(242,242,255,0.47)",
    iconBg: "#ffffff",
    iconBorder: "rgba(217,217,255,0.11)"
  }
};
var feedThemePalette = {
  light: {
    brandLogoBg: "rgba(2,32,71,0.05)",
    brandLogoBorder: "rgba(2,32,71,0.05)",
    brandName: "#6b7684",
    adBadge: "#8b95a1",
    title: "#333d4b",
    subtitle: "#4e5968",
    adClearance: "#8b95a1",
    imageOverlayBorder: "rgba(2,32,71,0.05)"
  },
  dark: {
    brandLogoBg: "#ffffff",
    brandLogoBorder: "rgba(217,217,255,0.11)",
    brandName: "#9e9ea4",
    adBadge: "#7e7e87",
    title: "#e4e4e5",
    subtitle: "#c3c3c6",
    adClearance: "#7e7e87",
    imageOverlayBorder: "rgba(217,217,255,0.11)"
  }
};
function createError2(code, message) {
  return {
    code,
    message: message || ERROR_MESSAGES[code] || "Unknown error",
    domain: "@apps-in-toss/framework"
  };
}
function isListBannerCreative(creative) {
  return Boolean(
    creative && typeof creative.title === "string" && typeof creative.subTitle === "string" && typeof creative.landingUrl === "string" && typeof creative.imageUrl === "string"
  );
}
function isFeedBannerCreative(creative) {
  if (!creative || typeof creative !== "object") {
    return false;
  }
  if (!("brandName" in creative)) {
    return false;
  }
  const candidate = creative;
  return Boolean(
    typeof candidate.brandName === "string" && typeof candidate.brandLogoUrl === "string" && typeof candidate.title === "string" && typeof candidate.subTitle === "string" && typeof candidate.mainImageUrl === "string" && typeof candidate.ctaText === "string" && typeof candidate.landingUrl === "string"
  );
}
function isValidCreative(ad) {
  const creative = ad.creative;
  const styleId = String(ad.styleId);
  if (styleId === LIST_BANNER_STYLE_ID) {
    return isListBannerCreative(creative);
  }
  if (styleId === FEED_BANNER_STYLE_ID) {
    return isFeedBannerCreative(creative);
  }
  return false;
}
function InlineAd(props) {
  const {
    adGroupId,
    theme,
    tone,
    variant,
    impressFallbackOnMount,
    onAdRendered,
    onAdViewable,
    onAdClicked,
    onAdImpression,
    onAdFailedToRender,
    onNoFill
  } = props;
  const slotIdRef = useRef8(createSlotId());
  const [ad, setAd] = useState7(null);
  const [isAdBadgeEnabled, setIsAdBadgeEnabled] = useState7(true);
  const eventTrackerRef = useRef8(null);
  const eventPayloadRef = useRef8(null);
  const hasRenderedRef = useRef8(false);
  const hasLoggedImp1pxRef = useRef8(false);
  const hasLoggedImp100pRef = useRef8(false);
  const hasNotifiedViewableRef = useRef8(false);
  const viewableTimerRef = useRef8(null);
  const refetchIntervalMsRef = useRef8(null);
  const lastImp1pxAtRef = useRef8(null);
  const loadingRef = useRef8(false);
  const loadRef = useRef8(null);
  const callbacksRef = useRef8({
    onAdRendered,
    onAdViewable,
    onAdClicked,
    onAdImpression,
    onAdFailedToRender,
    onNoFill
  });
  const isMountedRef = useRef8(false);
  const colorScheme = useColorScheme();
  const selectedTheme = theme ?? DEFAULT_INLINE_AD_THEME;
  const resolvedTheme = selectedTheme === "auto" ? colorScheme === "dark" ? "dark" : "light" : selectedTheme;
  const resolvedTone = tone ?? DEFAULT_INLINE_AD_TONE;
  const resolvedVariant = variant ?? DEFAULT_INLINE_AD_VARIANT;
  const colors = themePalette[resolvedTheme];
  const feedColors = feedThemePalette[resolvedTheme];
  const backgroundColor = resolvedTone === "grey" ? resolvedTheme === "dark" ? "#101013" : "#f2f4f7" : resolvedTheme === "dark" ? "#17171c" : "#ffffff";
  callbacksRef.current = {
    onAdRendered,
    onAdViewable,
    onAdClicked,
    onAdImpression,
    onAdFailedToRender,
    onNoFill
  };
  useEffect13(() => {
    isMountedRef.current = true;
    return () => {
      isMountedRef.current = false;
      if (viewableTimerRef.current) {
        clearTimeout(viewableTimerRef.current);
        viewableTimerRef.current = null;
      }
    };
  }, []);
  useEffect13(() => {
    const normalizedAdGroupId = normalizeAdGroupId(adGroupId);
    const loadAdRequest = () => {
      if (loadingRef.current) {
        return;
      }
      if (normalizedAdGroupId.length === 0) {
        const error = createError2(ERROR_CODES.INVALID_SPACE, INVALID_AD_GROUP_ID_ERROR_MESSAGE);
        callbacksRef.current.onAdFailedToRender?.(
          buildBannerErrorPayload(slotIdRef.current, normalizedAdGroupId, error)
        );
        return;
      }
      loadingRef.current = true;
      loadAd(normalizedAdGroupId).then((result) => {
        if (!isMountedRef.current) return;
        if (result.type === "noFill") {
          callbacksRef.current.onNoFill?.(buildNoFillPayload(slotIdRef.current, normalizedAdGroupId));
          return;
        }
        if (result.type === "error") {
          callbacksRef.current.onAdFailedToRender?.(
            buildBannerErrorPayload(slotIdRef.current, normalizedAdGroupId, result.error)
          );
          return;
        }
        if (!isValidCreative(result.ad)) {
          const invalidError = createError2(ERROR_CODES.INTERNAL_ERROR, "Invalid creative payload");
          callbacksRef.current.onAdFailedToRender?.(
            buildBannerErrorPayload(slotIdRef.current, normalizedAdGroupId, invalidError)
          );
          return;
        }
        if (result.responseExt?.refetchSeconds !== void 0) {
          refetchIntervalMsRef.current = typeof result.responseExt.refetchSeconds === "number" && result.responseExt.refetchSeconds > 0 ? result.responseExt.refetchSeconds * 1e3 : null;
        }
        hasRenderedRef.current = false;
        hasLoggedImp1pxRef.current = false;
        hasLoggedImp100pRef.current = false;
        hasNotifiedViewableRef.current = false;
        lastImp1pxAtRef.current = null;
        if (viewableTimerRef.current) {
          clearTimeout(viewableTimerRef.current);
          viewableTimerRef.current = null;
        }
        const payload = buildBannerEventPayload(slotIdRef.current, normalizedAdGroupId, result.ad);
        eventPayloadRef.current = payload;
        eventTrackerRef.current = new EventTracker(
          result.ad.eventTrackingUrls ?? [],
          result.ad.eventTypes ?? [],
          result.ad.eventPayload ?? "",
          result.requestId,
          result.ad.creative?.id
        );
        setIsAdBadgeEnabled(result.responseExt?.isAdBadgeEnabled !== false);
        setAd(result.ad);
      }).catch((error) => {
        if (!isMountedRef.current) return;
        const adError = error instanceof Error ? createError2(ERROR_CODES.INTERNAL_ERROR, error.message) : createError2(ERROR_CODES.INTERNAL_ERROR, "Unknown error");
        callbacksRef.current.onAdFailedToRender?.(
          buildBannerErrorPayload(slotIdRef.current, normalizedAdGroupId, adError)
        );
      }).finally(() => {
        loadingRef.current = false;
      });
    };
    loadRef.current = loadAdRequest;
    loadAdRequest();
    return () => {
      loadRef.current = null;
    };
  }, [adGroupId]);
  useVisibilityChange2((documentVisibility) => {
    if (documentVisibility !== "visible") {
      return;
    }
    const interval = refetchIntervalMsRef.current;
    const lastImp1pxAt = lastImp1pxAtRef.current;
    if (!interval || !lastImp1pxAt || loadingRef.current) {
      return;
    }
    if (Date.now() - lastImp1pxAt >= interval) {
      loadRef.current?.();
    }
  });
  useEffect13(() => {
    if (!ad || hasRenderedRef.current) {
      return;
    }
    hasRenderedRef.current = true;
    const payload = eventPayloadRef.current;
    if (payload) {
      callbacksRef.current.onAdRendered?.(payload);
    }
  }, [ad]);
  if (!ad) {
    return null;
  }
  const impressionKey = `${ad.requestId ?? "request"}:${ad.creative?.id ?? "creative"}`;
  const resolvedStyleId = ad.styleId || LIST_BANNER_STYLE_ID;
  const defaultPaddingValue = resolvedStyleId === LIST_BANNER_STYLE_ID ? DEFAULT_INLINE_AD_PADDING_BANNER : DEFAULT_INLINE_AD_PADDING_NATIVE_IMAGE;
  const paddingStyle = resolvePaddingStyle(defaultPaddingValue);
  const isListBanner = String(resolvedStyleId) === LIST_BANNER_STYLE_ID;
  const isFeedBanner = String(resolvedStyleId) === FEED_BANNER_STYLE_ID;
  const listCreative = isListBanner ? ad.creative : null;
  const feedCreative = isFeedBanner ? ad.creative : null;
  const subtitleText = listCreative && isAdBadgeEnabled ? `${listCreative.subTitle} \u2022 AD` : listCreative?.subTitle ?? "";
  const adClearanceText = listCreative?.adClearanceText ?? feedCreative?.adClearanceText ?? void 0;
  const adClearanceFontSize = adClearanceText && adClearanceText.length >= 60 ? 6 : 8;
  const imageUri = listCreative?.imageUrl && isSafeUrl(listCreative.imageUrl) ? listCreative.imageUrl : null;
  const brandLogoUri = feedCreative?.brandLogoUrl && isSafeUrl(feedCreative.brandLogoUrl) ? feedCreative.brandLogoUrl : null;
  const mainImageUri = feedCreative?.mainImageUrl && isSafeUrl(feedCreative.mainImageUrl) ? feedCreative.mainImageUrl : null;
  const basePayload = eventPayloadRef.current;
  const trackEvent = (eventType) => {
    eventTrackerRef.current?.track(eventType).catch((error) => {
      console.error("[InlineAd] Failed to track event:", error);
    });
  };
  const handleImpression1px = () => {
    if (hasLoggedImp1pxRef.current) return;
    hasLoggedImp1pxRef.current = true;
    lastImp1pxAtRef.current = Date.now();
    trackEvent("IMP_1PX");
    if (basePayload) {
      void tossAdEventLog({
        log_name: "display_ads_all::impression__1px_banner",
        log_type: "event",
        params: {
          event_type: "impression",
          schema_id: 1812034,
          request_id: basePayload.adMetadata.requestId ?? ""
        }
      });
      callbacksRef.current.onAdImpression?.(basePayload);
    }
  };
  const handleImpression100p = () => {
    if (hasLoggedImp100pRef.current) return;
    hasLoggedImp100pRef.current = true;
    trackEvent("IMP_100P");
  };
  const handleViewableStart = () => {
    if (hasNotifiedViewableRef.current || viewableTimerRef.current) return;
    viewableTimerRef.current = setTimeout(() => {
      hasNotifiedViewableRef.current = true;
      viewableTimerRef.current = null;
      trackEvent("VIMP");
      if (basePayload) {
        callbacksRef.current.onAdViewable?.(basePayload);
      }
    }, 1e3);
  };
  const handleViewableEnd = () => {
    if (viewableTimerRef.current) {
      clearTimeout(viewableTimerRef.current);
      viewableTimerRef.current = null;
    }
  };
  const handleClick = (slotKey) => {
    trackEvent("CLICK");
    if (slotKey != null) {
      trackEvent(`CLICK_${slotKey}`);
    }
    if (isSafeUrl(ad.creative.landingUrl)) {
      openLandingUrl(ad.creative.landingUrl);
    }
    if (basePayload) {
      callbacksRef.current.onAdClicked?.(basePayload);
    }
  };
  return /* @__PURE__ */ jsx22(
    ImpressionArea,
    {
      style: styles3.impressionArea,
      onImpressionStart: handleImpression1px,
      areaThreshold: 0,
      UNSAFE__impressFallbackOnMount: Boolean(impressFallbackOnMount),
      children: /* @__PURE__ */ jsx22(
        ImpressionArea,
        {
          onImpressionStart: handleImpression100p,
          areaThreshold: 1,
          UNSAFE__impressFallbackOnMount: Boolean(impressFallbackOnMount),
          children: /* @__PURE__ */ jsx22(
            ImpressionArea,
            {
              onImpressionStart: handleViewableStart,
              onImpressionEnd: handleViewableEnd,
              areaThreshold: 0.5,
              UNSAFE__impressFallbackOnMount: Boolean(impressFallbackOnMount),
              children: /* @__PURE__ */ jsx22(View7, { style: [styles3.wrapper, resolvedVariant === "card" && styles3.cardWrapper], children: /* @__PURE__ */ jsxs11(View7, { style: [styles3.surface, { backgroundColor }, resolvedVariant === "card" && styles3.cardSurface], children: [
                isListBanner && listCreative ? /* @__PURE__ */ jsx22(
                  ListBannerAdView,
                  {
                    title: listCreative.title,
                    subtitle: subtitleText,
                    adClearanceText,
                    adClearanceFontSize,
                    imageUri,
                    paddingStyle,
                    colors,
                    onPress: handleClick
                  }
                ) : null,
                isFeedBanner && feedCreative ? /* @__PURE__ */ jsx22(
                  FeedBannerAdView,
                  {
                    brandName: feedCreative.brandName,
                    brandLogoUri,
                    title: feedCreative.title,
                    subtitle: feedCreative.subTitle,
                    mainImageUri,
                    ctaText: feedCreative.ctaText,
                    ctaTextColor: feedCreative.ctaTextColor,
                    ctaBackgroundColor: feedCreative.ctaBackgroundColor,
                    adClearanceText,
                    colors: feedColors,
                    isAdBadgeEnabled,
                    paddingStyle,
                    onPress: handleClick
                  }
                ) : null
              ] }) })
            }
          )
        }
      )
    },
    impressionKey
  );
}
var styles3 = StyleSheet3.create({
  impressionArea: {
    width: "100%"
  },
  wrapper: {
    width: "100%"
  },
  cardWrapper: {
    paddingHorizontal: 10
  },
  surface: {
    width: "100%",
    overflow: "hidden"
  },
  cardSurface: {
    borderRadius: 16
  }
});

// src/index.ts
var Analytics2 = {
  init: InternalAnalytics.init,
  Impression: InternalAnalytics.Impression,
  Press: InternalAnalytics.Press,
  Area: InternalAnalytics.Area
};
export {
  Analytics2 as Analytics,
  AppsInToss,
  INTERNAL__onVisibilityChangedByTransparentServiceWeb,
  InlineAd,
  OverlayProvider,
  StartUpdateLocationPermissionError,
  WebView,
  env,
  homeEvent,
  loadFullScreenAd,
  safeAreaInsetsChange,
  showFullScreenAd,
  useCreateUserAgent,
  useGeolocation,
  useOverlay3 as useOverlay,
  useTopNavigation,
  useWaitForReturnNavigator
};
