// @apps-in-toss/web-framework/dist/config/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// ../../../.yarn/__virtual__/tsup-virtual-5fd798e2dc/0/cache/tsup-npm-8.3.5-ed25596739-7794953cbc.zip/node_modules/tsup/assets/esm_shims.js
var init_esm_shims = __esm({
  "../../../.yarn/__virtual__/tsup-virtual-5fd798e2dc/0/cache/tsup-npm-8.3.5-ed25596739-7794953cbc.zip/node_modules/tsup/assets/esm_shims.js"() {
    "use strict";
  }
});

// ../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_validateReport.js
var require_validateReport = __commonJS({
  "../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_validateReport.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports._validateReport = void 0;
    var _validateReport2 = (array) => {
      const reportable = (path) => {
        if (array.length === 0)
          return true;
        const last = array[array.length - 1].path;
        return path.length > last.length || last.substring(0, path.length) !== path;
      };
      return (exceptable, error) => {
        var _a;
        if (exceptable && reportable(error.path)) {
          if (error.value === void 0)
            (_a = error.description) !== null && _a !== void 0 ? _a : error.description = [
              "The value at this path is `undefined`.",
              "",
              `Please fill the \`${error.expected}\` typed value next time.`
            ].join("\n");
          array.push(error);
        }
        return false;
      };
    };
    exports._validateReport = _validateReport2;
  }
});

// ../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_createStandardSchema.js
var require_createStandardSchema = __commonJS({
  "../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_createStandardSchema.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports._createStandardSchema = void 0;
    var _createStandardSchema2 = (fn) => Object.assign(fn, {
      "~standard": {
        version: 1,
        vendor: "typia",
        validate: (input) => {
          const result = fn(input);
          if (result.success) {
            return {
              value: result.data
            };
          } else {
            return {
              issues: result.errors.map((error) => ({
                message: `expected ${error.expected}, got ${error.value}`,
                path: typiaPathToStandardSchemaPath(error.path)
              }))
            };
          }
        }
      }
    });
    exports._createStandardSchema = _createStandardSchema2;
    var PathParserState;
    (function(PathParserState2) {
      PathParserState2[PathParserState2["Start"] = 0] = "Start";
      PathParserState2[PathParserState2["Property"] = 1] = "Property";
      PathParserState2[PathParserState2["StringKey"] = 2] = "StringKey";
      PathParserState2[PathParserState2["NumberKey"] = 3] = "NumberKey";
    })(PathParserState || (PathParserState = {}));
    var typiaPathToStandardSchemaPath = (path) => {
      if (!path.startsWith("$input")) {
        throw new Error(`Invalid path: ${JSON.stringify(path)}`);
      }
      const segments = [];
      let currentSegment = "";
      let state = PathParserState.Start;
      let index = "$input".length - 1;
      while (index < path.length - 1) {
        index++;
        const char = path[index];
        if (state === PathParserState.Property) {
          if (char === "." || char === "[") {
            segments.push({
              key: currentSegment
            });
            state = PathParserState.Start;
          } else if (index === path.length - 1) {
            currentSegment += char;
            segments.push({
              key: currentSegment
            });
            index++;
            state = PathParserState.Start;
          } else {
            currentSegment += char;
          }
        } else if (state === PathParserState.StringKey) {
          if (char === '"') {
            segments.push({
              key: JSON.parse(currentSegment + char)
            });
            index += 2;
            state = PathParserState.Start;
          } else if (char === "\\") {
            currentSegment += path[index];
            index++;
            currentSegment += path[index];
          } else {
            currentSegment += char;
          }
        } else if (state === PathParserState.NumberKey) {
          if (char === "]") {
            segments.push({
              key: Number.parseInt(currentSegment)
            });
            index++;
            state = PathParserState.Start;
          } else {
            currentSegment += char;
          }
        }
        if (state === PathParserState.Start && index < path.length - 1) {
          const newChar = path[index];
          currentSegment = "";
          if (newChar === "[") {
            if (path[index + 1] === '"') {
              state = PathParserState.StringKey;
              index++;
              currentSegment = '"';
            } else {
              state = PathParserState.NumberKey;
            }
          } else if (newChar === ".") {
            state = PathParserState.Property;
          } else {
            throw new Error("Unreachable: pointer points invalid character");
          }
        }
      }
      if (state !== PathParserState.Start) {
        throw new Error(`Failed to parse path: ${JSON.stringify(path)}`);
      }
      return segments;
    };
  }
});

// src/config/index.ts
init_esm_shims();

// src/config/defineConfig.ts
init_esm_shims();

// ../../../.yarn/cache/es-toolkit-npm-1.32.0-019fc7c155-56ba965570.zip/node_modules/es-toolkit/dist/index.mjs
init_esm_shims();

// ../../../.yarn/cache/es-toolkit-npm-1.32.0-019fc7c155-56ba965570.zip/node_modules/es-toolkit/dist/predicate/isPlainObject.mjs
init_esm_shims();
function isPlainObject(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  const proto = Object.getPrototypeOf(value);
  const hasObjectPrototype = proto === null || proto === Object.prototype || Object.getPrototypeOf(proto) === null;
  if (!hasObjectPrototype) {
    return false;
  }
  return Object.prototype.toString.call(value) === "[object Object]";
}

// ../../../.yarn/cache/es-toolkit-npm-1.32.0-019fc7c155-56ba965570.zip/node_modules/es-toolkit/dist/object/merge.mjs
init_esm_shims();
function merge(target, source) {
  const sourceKeys = Object.keys(source);
  for (let i = 0; i < sourceKeys.length; i++) {
    const key = sourceKeys[i];
    const sourceValue = source[key];
    const targetValue = target[key];
    if (Array.isArray(sourceValue)) {
      if (Array.isArray(targetValue)) {
        target[key] = merge(targetValue, sourceValue);
      } else {
        target[key] = merge([], sourceValue);
      }
    } else if (isPlainObject(sourceValue)) {
      if (isPlainObject(targetValue)) {
        target[key] = merge(targetValue, sourceValue);
      } else {
        target[key] = merge({}, sourceValue);
      }
    } else if (targetValue === void 0 || sourceValue !== void 0) {
      target[key] = sourceValue;
    }
  }
  return target;
}

// src/plugins/appsInTossWeb.ts
init_esm_shims();
var __typia_transform__validateReport = __toESM(require_validateReport(), 1);
var __typia_transform__createStandardSchema = __toESM(require_createStandardSchema(), 1);
var validateAppInTossWebPluginOptions = (() => {
  const _io0 = (input) => "object" === typeof input.commands && null !== input.commands && _io1(input.commands) && ("object" === typeof input.brand && null !== input.brand && _io2(input.brand)) && (void 0 === input.navigationBar || "object" === typeof input.navigationBar && null !== input.navigationBar && false === Array.isArray(input.navigationBar) && _io3(input.navigationBar)) && (void 0 === input.webViewProps || "object" === typeof input.webViewProps && null !== input.webViewProps && false === Array.isArray(input.webViewProps) && _io8(input.webViewProps)) && (Array.isArray(input.permissions) && input.permissions.every((elem) => "object" === typeof elem && null !== elem && _iu0(elem)));
  const _io1 = (input) => "string" === typeof input.build;
  const _io2 = (input) => "string" === typeof input.displayName && "string" === typeof input.primaryColor && "string" === typeof input.icon;
  const _io3 = (input) => (void 0 === input.withBackButton || "boolean" === typeof input.withBackButton) && (void 0 === input.withHomeButton || "boolean" === typeof input.withHomeButton) && (void 0 === input.initialAccessoryButton || "object" === typeof input.initialAccessoryButton && null !== input.initialAccessoryButton && _io4(input.initialAccessoryButton));
  const _io4 = (input) => "string" === typeof input.id && "string" === typeof input.title && ("object" === typeof input.icon && null !== input.icon && _iu1(input.icon));
  const _io5 = (input) => "object" === typeof input.source && null !== input.source && _io6(input.source) && (null !== input.name && void 0 === input.name);
  const _io6 = (input) => "string" === typeof input.uri;
  const _io7 = (input) => "string" === typeof input.name && (null !== input.source && void 0 === input.source);
  const _io8 = (input) => (void 0 === input.type || "partner" === input.type || "external" === input.type || "game" === input.type) && (void 0 === input.allowsInlineMediaPlayback || "boolean" === typeof input.allowsInlineMediaPlayback) && (void 0 === input.bounces || "boolean" === typeof input.bounces) && (void 0 === input.pullToRefreshEnabled || "boolean" === typeof input.pullToRefreshEnabled) && (void 0 === input.overScrollMode || "always" === input.overScrollMode || "content" === input.overScrollMode || "never" === input.overScrollMode) && (void 0 === input.mediaPlaybackRequiresUserAction || "boolean" === typeof input.mediaPlaybackRequiresUserAction) && (void 0 === input.allowsBackForwardNavigationGestures || "boolean" === typeof input.allowsBackForwardNavigationGestures);
  const _io9 = (input) => "clipboard" === input.name && ("read" === input.access || "write" === input.access);
  const _io10 = (input) => "geolocation" === input.name && "access" === input.access;
  const _io11 = (input) => "contacts" === input.name && ("read" === input.access || "write" === input.access);
  const _io12 = (input) => "photos" === input.name && ("read" === input.access || "write" === input.access);
  const _io13 = (input) => "camera" === input.name && "access" === input.access;
  const _io14 = (input) => "microphone" === input.name && "access" === input.access;
  const _iu0 = (input) => (() => {
    if ("microphone" === input.name)
      return _io14(input);
    else if ("camera" === input.name)
      return _io13(input);
    else if ("photos" === input.name)
      return _io12(input);
    else if ("contacts" === input.name)
      return _io11(input);
    else if ("geolocation" === input.name)
      return _io10(input);
    else if ("clipboard" === input.name)
      return _io9(input);
    else
      return false;
  })();
  const _iu1 = (input) => (() => {
    if ("string" === typeof input.name)
      return _io7(input);
    else if ("object" === typeof input.source && null !== input.source && _io6(input.source))
      return _io5(input);
    else
      return false;
  })();
  const _vo0 = (input, _path, _exceptionable = true) => [("object" === typeof input.commands && null !== input.commands || _report(_exceptionable, {
    path: _path + ".commands",
    expected: "__type",
    value: input.commands
  })) && _vo1(input.commands, _path + ".commands", _exceptionable) || _report(_exceptionable, {
    path: _path + ".commands",
    expected: "__type",
    value: input.commands
  }), ("object" === typeof input.brand && null !== input.brand || _report(_exceptionable, {
    path: _path + ".brand",
    expected: "__type.o1",
    value: input.brand
  })) && _vo2(input.brand, _path + ".brand", _exceptionable) || _report(_exceptionable, {
    path: _path + ".brand",
    expected: "__type.o1",
    value: input.brand
  }), void 0 === input.navigationBar || ("object" === typeof input.navigationBar && null !== input.navigationBar && false === Array.isArray(input.navigationBar) || _report(_exceptionable, {
    path: _path + ".navigationBar",
    expected: "(__type.o2 | undefined)",
    value: input.navigationBar
  })) && _vo3(input.navigationBar, _path + ".navigationBar", _exceptionable) || _report(_exceptionable, {
    path: _path + ".navigationBar",
    expected: "(__type.o2 | undefined)",
    value: input.navigationBar
  }), void 0 === input.webViewProps || ("object" === typeof input.webViewProps && null !== input.webViewProps && false === Array.isArray(input.webViewProps) || _report(_exceptionable, {
    path: _path + ".webViewProps",
    expected: "(__type.o6 | undefined)",
    value: input.webViewProps
  })) && _vo8(input.webViewProps, _path + ".webViewProps", _exceptionable) || _report(_exceptionable, {
    path: _path + ".webViewProps",
    expected: "(__type.o6 | undefined)",
    value: input.webViewProps
  }), (Array.isArray(input.permissions) || _report(_exceptionable, {
    path: _path + ".permissions",
    expected: "Array<Permission>",
    value: input.permissions
  })) && input.permissions.map((elem, _index2) => ("object" === typeof elem && null !== elem || _report(_exceptionable, {
    path: _path + ".permissions[" + _index2 + "]",
    expected: "(CameraPermission | ClipboardPermission | ContactsPermission | GeolocationPermission | MicrophonePermission | PhotosPermission)",
    value: elem
  })) && _vu0(elem, _path + ".permissions[" + _index2 + "]", _exceptionable) || _report(_exceptionable, {
    path: _path + ".permissions[" + _index2 + "]",
    expected: "(CameraPermission | ClipboardPermission | ContactsPermission | GeolocationPermission | MicrophonePermission | PhotosPermission)",
    value: elem
  })).every((flag) => flag) || _report(_exceptionable, {
    path: _path + ".permissions",
    expected: "Array<Permission>",
    value: input.permissions
  })].every((flag) => flag);
  const _vo1 = (input, _path, _exceptionable = true) => ["string" === typeof input.build || _report(_exceptionable, {
    path: _path + ".build",
    expected: "string",
    value: input.build
  })].every((flag) => flag);
  const _vo2 = (input, _path, _exceptionable = true) => ["string" === typeof input.displayName || _report(_exceptionable, {
    path: _path + ".displayName",
    expected: "string",
    value: input.displayName
  }), "string" === typeof input.primaryColor || _report(_exceptionable, {
    path: _path + ".primaryColor",
    expected: "string",
    value: input.primaryColor
  }), "string" === typeof input.icon || _report(_exceptionable, {
    path: _path + ".icon",
    expected: "string",
    value: input.icon
  })].every((flag) => flag);
  const _vo3 = (input, _path, _exceptionable = true) => [void 0 === input.withBackButton || "boolean" === typeof input.withBackButton || _report(_exceptionable, {
    path: _path + ".withBackButton",
    expected: "(boolean | undefined)",
    value: input.withBackButton
  }), void 0 === input.withHomeButton || "boolean" === typeof input.withHomeButton || _report(_exceptionable, {
    path: _path + ".withHomeButton",
    expected: "(boolean | undefined)",
    value: input.withHomeButton
  }), void 0 === input.initialAccessoryButton || ("object" === typeof input.initialAccessoryButton && null !== input.initialAccessoryButton || _report(_exceptionable, {
    path: _path + ".initialAccessoryButton",
    expected: "(InitialAccessoryButton | undefined)",
    value: input.initialAccessoryButton
  })) && _vo4(input.initialAccessoryButton, _path + ".initialAccessoryButton", _exceptionable) || _report(_exceptionable, {
    path: _path + ".initialAccessoryButton",
    expected: "(InitialAccessoryButton | undefined)",
    value: input.initialAccessoryButton
  })].every((flag) => flag);
  const _vo4 = (input, _path, _exceptionable = true) => ["string" === typeof input.id || _report(_exceptionable, {
    path: _path + ".id",
    expected: "string",
    value: input.id
  }), "string" === typeof input.title || _report(_exceptionable, {
    path: _path + ".title",
    expected: "string",
    value: input.title
  }), ("object" === typeof input.icon && null !== input.icon || _report(_exceptionable, {
    path: _path + ".icon",
    expected: "(__type.o3 | __type.o5)",
    value: input.icon
  })) && _vu1(input.icon, _path + ".icon", _exceptionable) || _report(_exceptionable, {
    path: _path + ".icon",
    expected: "(__type.o3 | __type.o5)",
    value: input.icon
  })].every((flag) => flag);
  const _vo5 = (input, _path, _exceptionable = true) => [("object" === typeof input.source && null !== input.source || _report(_exceptionable, {
    path: _path + ".source",
    expected: "__type.o4",
    value: input.source
  })) && _vo6(input.source, _path + ".source", _exceptionable) || _report(_exceptionable, {
    path: _path + ".source",
    expected: "__type.o4",
    value: input.source
  }), (null !== input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: "undefined",
    value: input.name
  })) && (void 0 === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: "undefined",
    value: input.name
  }))].every((flag) => flag);
  const _vo6 = (input, _path, _exceptionable = true) => ["string" === typeof input.uri || _report(_exceptionable, {
    path: _path + ".uri",
    expected: "string",
    value: input.uri
  })].every((flag) => flag);
  const _vo7 = (input, _path, _exceptionable = true) => ["string" === typeof input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: "string",
    value: input.name
  }), (null !== input.source || _report(_exceptionable, {
    path: _path + ".source",
    expected: "undefined",
    value: input.source
  })) && (void 0 === input.source || _report(_exceptionable, {
    path: _path + ".source",
    expected: "undefined",
    value: input.source
  }))].every((flag) => flag);
  const _vo8 = (input, _path, _exceptionable = true) => [void 0 === input.type || "partner" === input.type || "external" === input.type || "game" === input.type || _report(_exceptionable, {
    path: _path + ".type",
    expected: '("external" | "game" | "partner" | undefined)',
    value: input.type
  }), void 0 === input.allowsInlineMediaPlayback || "boolean" === typeof input.allowsInlineMediaPlayback || _report(_exceptionable, {
    path: _path + ".allowsInlineMediaPlayback",
    expected: "(boolean | undefined)",
    value: input.allowsInlineMediaPlayback
  }), void 0 === input.bounces || "boolean" === typeof input.bounces || _report(_exceptionable, {
    path: _path + ".bounces",
    expected: "(boolean | undefined)",
    value: input.bounces
  }), void 0 === input.pullToRefreshEnabled || "boolean" === typeof input.pullToRefreshEnabled || _report(_exceptionable, {
    path: _path + ".pullToRefreshEnabled",
    expected: "(boolean | undefined)",
    value: input.pullToRefreshEnabled
  }), void 0 === input.overScrollMode || "always" === input.overScrollMode || "content" === input.overScrollMode || "never" === input.overScrollMode || _report(_exceptionable, {
    path: _path + ".overScrollMode",
    expected: '("always" | "content" | "never" | undefined)',
    value: input.overScrollMode
  }), void 0 === input.mediaPlaybackRequiresUserAction || "boolean" === typeof input.mediaPlaybackRequiresUserAction || _report(_exceptionable, {
    path: _path + ".mediaPlaybackRequiresUserAction",
    expected: "(boolean | undefined)",
    value: input.mediaPlaybackRequiresUserAction
  }), void 0 === input.allowsBackForwardNavigationGestures || "boolean" === typeof input.allowsBackForwardNavigationGestures || _report(_exceptionable, {
    path: _path + ".allowsBackForwardNavigationGestures",
    expected: "(boolean | undefined)",
    value: input.allowsBackForwardNavigationGestures
  })].every((flag) => flag);
  const _vo9 = (input, _path, _exceptionable = true) => ["clipboard" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"clipboard"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo10 = (input, _path, _exceptionable = true) => ["geolocation" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"geolocation"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo11 = (input, _path, _exceptionable = true) => ["contacts" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"contacts"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo12 = (input, _path, _exceptionable = true) => ["photos" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"photos"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo13 = (input, _path, _exceptionable = true) => ["camera" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"camera"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo14 = (input, _path, _exceptionable = true) => ["microphone" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"microphone"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vu0 = (input, _path, _exceptionable = true) => (() => {
    if ("microphone" === input.name)
      return _vo14(input, _path, _exceptionable);
    else if ("camera" === input.name)
      return _vo13(input, _path, _exceptionable);
    else if ("photos" === input.name)
      return _vo12(input, _path, _exceptionable);
    else if ("contacts" === input.name)
      return _vo11(input, _path, _exceptionable);
    else if ("geolocation" === input.name)
      return _vo10(input, _path, _exceptionable);
    else if ("clipboard" === input.name)
      return _vo9(input, _path, _exceptionable);
    else
      return _report(_exceptionable, {
        path: _path,
        expected: "(MicrophonePermission | CameraPermission | PhotosPermission | ContactsPermission | GeolocationPermission | ClipboardPermission)",
        value: input
      });
  })();
  const _vu1 = (input, _path, _exceptionable = true) => (() => {
    if ("string" === typeof input.name)
      return _vo7(input, _path, _exceptionable);
    else if ("object" === typeof input.source && null !== input.source && _vo6(input.source, _path + ".source", false))
      return _vo5(input, _path, _exceptionable);
    else
      return _report(_exceptionable, {
        path: _path,
        expected: "(__type.o5 | __type.o3)",
        value: input
      });
  })();
  const __is = (input) => "object" === typeof input && null !== input && _io0(input);
  let errors;
  let _report;
  return __typia_transform__createStandardSchema._createStandardSchema((input) => {
    if (false === __is(input)) {
      errors = [];
      _report = __typia_transform__validateReport._validateReport(errors);
      ((input2, _path, _exceptionable = true) => ("object" === typeof input2 && null !== input2 || _report(true, {
        path: _path + "",
        expected: "AppsInTossWebPluginOptions",
        value: input2
      })) && _vo0(input2, _path + "", true) || _report(true, {
        path: _path + "",
        expected: "AppsInTossWebPluginOptions",
        value: input2
      }))(input, "$input", true);
      const success = 0 === errors.length;
      return success ? {
        success,
        data: input
      } : {
        success,
        errors,
        data: input
      };
    }
    return {
      success: true,
      data: input
    };
  });
})();

// src/config/defineConfig.ts
var defineConfig = (config) => {
  const defaultConfig = {
    scheme: "intoss",
    outdir: "dist",
    web: {
      host: "localhost"
    },
    webViewProps: {
      type: "partner",
      mediaPlaybackRequiresUserAction: true,
      allowsInlineMediaPlayback: false,
      bounces: true,
      allowsBackForwardNavigationGestures: true,
      pullToRefreshEnabled: true,
      overScrollMode: "always"
    },
    navigationBar: {
      withBackButton: true,
      withHomeButton: false
    }
  };
  const merged = merge(defaultConfig, config);
  const result = validateAppInTossWebPluginOptions({
    commands: merged.web.commands,
    permissions: merged.permissions,
    brand: merged.brand,
    navigationBar: merged.navigationBar,
    webViewProps: merged.webViewProps
  });
  if (!result.success) {
    console.error("[Apps In Toss Plugin] \uD50C\uB7EC\uADF8\uC778 \uC635\uC158\uC774 \uC62C\uBC14\uB974\uC9C0 \uC54A\uC2B5\uB2C8\uB2E4. granite.config.ts \uAD6C\uC131\uC744 \uD655\uC778\uD574\uC8FC\uC138\uC694.");
    process.exit(1);
  }
  return merged;
};
export {
  defineConfig
};
