// @apps-in-toss/web-config/dist/index.js
import { createRequire } from 'module'; const require = createRequire(import.meta.url);

// src/babelBuild.ts
import fs from "fs";
import path from "path";
import { transform } from "@babel/core";

// src/babelJsonReplacementPlugin.ts
import * as t from "@babel/types";
function babelJsonReplacementPlugin(_, opts) {
  const json = opts.json;
  return {
    name: "babel-json-replacement-plugin",
    visitor: {
      MemberExpression(path2) {
        const { node } = path2;
        if (t.isIdentifier(node.object, { name: opts.identifierName }) && t.isIdentifier(node.property) && node.property.name in json) {
          path2.replaceWith(t.valueToNode(json[node.property.name]));
        }
      }
    }
  };
}

// src/babelBuild.ts
async function babelBuild({ buildNumber, deploymentId, config, prebuiltPath }) {
  const bundle = await fs.promises.readFile(prebuiltPath, "utf-8");
  const result = await transform(bundle, {
    babelrc: false,
    configFile: false,
    sourceMaps: true,
    plugins: [
      [
        babelJsonReplacementPlugin,
        {
          json: {
            appName: config.appName,
            webPort: "NOOP",
            webHost: "NOOP",
            buildNumber,
            deploymentId,
            brandDisplayName: config.brand.displayName,
            brandIcon: config.brand.icon,
            brandPrimaryColor: config.brand.primaryColor,
            webViewType: config.webViewProps.type,
            navigationBar: JSON.stringify(config.navigationBar),
            ...config.webViewProps
          },
          identifierName: "Ait"
        }
      ]
    ],
    generatorOpts: {
      compact: false,
      comments: false
    }
  });
  return result;
}
async function writeBundle(result, { platform, config, cwd }) {
  if (!(result?.code && result?.map)) {
    throw new Error(`Failed to build ${platform}`);
  }
  await fs.promises.mkdir(path.join(cwd, config.outdir), { recursive: true });
  const basePath = path.join(cwd, config.outdir);
  const bundleName = `bundle.${platform}.js`;
  const bundlePath = path.join(basePath, bundleName);
  const sourcemapPath = `${bundlePath}.map`;
  await fs.promises.writeFile(bundlePath, result.code);
  await fs.promises.writeFile(sourcemapPath, JSON.stringify(result.map));
  return { bundlePath, sourcemapPath };
}

// src/getPackageManager.ts
function getPackageManager({ isExecutor = false } = {}) {
  const userAgent = process.env.npm_config_user_agent;
  const packageManagerCommands = {
    npm: isExecutor ? "npx" : "npm",
    pnpm: "pnpm",
    yarn: "yarn",
    bun: "bun"
  };
  if (!userAgent) {
    return {
      packageManager: packageManagerCommands["npm"],
      version: "0.0.0"
    };
  }
  const [packageManagerInfo] = userAgent.match(/(\w+)\/(\d+\.\d+\.\d+)/) || [];
  const [packageManagerRaw, version] = packageManagerInfo?.split("/") ?? ["npm", null];
  const packageManager = typeof packageManagerRaw === "string" && packageManagerRaw in packageManagerCommands ? packageManagerRaw : "npm";
  return {
    packageManager: packageManagerCommands[packageManager],
    version: version ?? "0.0.0"
  };
}

// src/loadConfig.ts
import { getPackageRoot } from "@granite-js/utils";
import { cosmiconfig } from "cosmiconfig";
import { TypeScriptLoader } from "cosmiconfig-typescript-loader";
var loadConfig = async () => {
  const explorer = cosmiconfig("granite", {
    searchPlaces: ["granite.config.ts", "granite.config.mts", "granite.config.js", "granite.config.cjs"],
    loaders: {
      ".ts": TypeScriptLoader(),
      ".mts": TypeScriptLoader()
    }
  });
  const result = await explorer.search(getPackageRoot());
  const config = await result?.config;
  return config;
};
export {
  babelBuild,
  babelJsonReplacementPlugin,
  getPackageManager,
  loadConfig,
  writeBundle
};
