// @apps-in-toss/web-config/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  babelBuild: () => babelBuild,
  babelJsonReplacementPlugin: () => babelJsonReplacementPlugin,
  getPackageManager: () => getPackageManager,
  loadConfig: () => loadConfig,
  writeBundle: () => writeBundle
});
module.exports = __toCommonJS(src_exports);

// src/babelBuild.ts
var import_fs = __toESM(require("fs"), 1);
var import_path = __toESM(require("path"), 1);
var import_core = require("@babel/core");

// src/babelJsonReplacementPlugin.ts
var t = __toESM(require("@babel/types"), 1);
function babelJsonReplacementPlugin(_, opts) {
  const json = opts.json;
  return {
    name: "babel-json-replacement-plugin",
    visitor: {
      MemberExpression(path2) {
        const { node } = path2;
        if (t.isIdentifier(node.object, { name: opts.identifierName }) && t.isIdentifier(node.property) && node.property.name in json) {
          path2.replaceWith(t.valueToNode(json[node.property.name]));
        }
      }
    }
  };
}

// src/babelBuild.ts
async function babelBuild({ buildNumber, deploymentId, config, prebuiltPath }) {
  const bundle = await import_fs.default.promises.readFile(prebuiltPath, "utf-8");
  const result = await (0, import_core.transform)(bundle, {
    babelrc: false,
    configFile: false,
    sourceMaps: true,
    plugins: [
      [
        babelJsonReplacementPlugin,
        {
          json: {
            appName: config.appName,
            webPort: "NOOP",
            webHost: "NOOP",
            buildNumber,
            deploymentId,
            brandDisplayName: config.brand.displayName,
            brandIcon: config.brand.icon,
            brandPrimaryColor: config.brand.primaryColor,
            webViewType: config.webViewProps.type,
            navigationBar: JSON.stringify(config.navigationBar),
            ...config.webViewProps
          },
          identifierName: "Ait"
        }
      ]
    ],
    generatorOpts: {
      compact: false,
      comments: false
    }
  });
  return result;
}
async function writeBundle(result, { platform, config, cwd }) {
  if (!(result?.code && result?.map)) {
    throw new Error(`Failed to build ${platform}`);
  }
  await import_fs.default.promises.mkdir(import_path.default.join(cwd, config.outdir), { recursive: true });
  const basePath = import_path.default.join(cwd, config.outdir);
  const bundleName = `bundle.${platform}.js`;
  const bundlePath = import_path.default.join(basePath, bundleName);
  const sourcemapPath = `${bundlePath}.map`;
  await import_fs.default.promises.writeFile(bundlePath, result.code);
  await import_fs.default.promises.writeFile(sourcemapPath, JSON.stringify(result.map));
  return { bundlePath, sourcemapPath };
}

// src/getPackageManager.ts
function getPackageManager({ isExecutor = false } = {}) {
  const userAgent = process.env.npm_config_user_agent;
  const packageManagerCommands = {
    npm: isExecutor ? "npx" : "npm",
    pnpm: "pnpm",
    yarn: "yarn",
    bun: "bun"
  };
  if (!userAgent) {
    return {
      packageManager: packageManagerCommands["npm"],
      version: "0.0.0"
    };
  }
  const [packageManagerInfo] = userAgent.match(/(\w+)\/(\d+\.\d+\.\d+)/) || [];
  const [packageManagerRaw, version] = packageManagerInfo?.split("/") ?? ["npm", null];
  const packageManager = typeof packageManagerRaw === "string" && packageManagerRaw in packageManagerCommands ? packageManagerRaw : "npm";
  return {
    packageManager: packageManagerCommands[packageManager],
    version: version ?? "0.0.0"
  };
}

// src/loadConfig.ts
var import_utils = require("@granite-js/utils");
var import_cosmiconfig = require("cosmiconfig");
var import_cosmiconfig_typescript_loader = require("cosmiconfig-typescript-loader");
var loadConfig = async () => {
  const explorer = (0, import_cosmiconfig.cosmiconfig)("granite", {
    searchPlaces: ["granite.config.ts", "granite.config.mts", "granite.config.js", "granite.config.cjs"],
    loaders: {
      ".ts": (0, import_cosmiconfig_typescript_loader.TypeScriptLoader)(),
      ".mts": (0, import_cosmiconfig_typescript_loader.TypeScriptLoader)()
    }
  });
  const result = await explorer.search((0, import_utils.getPackageRoot)());
  const config = await result?.config;
  return config;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  babelBuild,
  babelJsonReplacementPlugin,
  getPackageManager,
  loadConfig,
  writeBundle
});
