// @granite-js/video/dist/module/GraniteVideo.js
"use strict";

import React, { forwardRef, useRef, useImperativeHandle, useCallback } from 'react';
import { Platform, StyleSheet, View, Image, findNodeHandle, NativeModules } from 'react-native';
import NativeGraniteVideoView, { Commands } from './GraniteVideoNativeComponent';
import { jsx as _jsx } from "react/jsx-runtime";
const {
  GraniteVideoModule
} = NativeModules;
function toNativeStringMapEntries(value) {
  if (value === undefined) {
    return undefined;
  }
  return Object.entries(value).map(([name, value]) => ({
    name,
    value
  }));
}
function normalizeDrm(drm) {
  if (!drm) {
    return undefined;
  }
  return {
    ...drm,
    headers: toNativeStringMapEntries(drm.headers)
  };
}

// For Fabric (New Architecture), the component is always available through codegenNativeComponent
// We don't need to check UIManager.getViewManagerConfig which is Old Architecture only

// Convert CMCD config to native format.
function normalizeCmcd(cmcd) {
  if (cmcd === undefined || cmcd === null) {
    return undefined;
  }
  if (typeof cmcd === 'boolean') {
    if (!cmcd) {
      return undefined;
    }
    // When true, use MODE_QUERY_PARAMETER as default
    return {
      mode: 1
    };
  }
  return {
    mode: cmcd.mode ?? 1,
    request: toNativeStringMapEntries(cmcd.request),
    session: toNativeStringMapEntries(cmcd.session),
    object: toNativeStringMapEntries(cmcd.object),
    status: toNativeStringMapEntries(cmcd.status)
  };
}
function normalizeAd(ad) {
  if (!ad) {
    return undefined;
  }
  return {
    ...ad,
    adTagParameters: ad.type === 'ssai' ? toNativeStringMapEntries(ad.adTagParameters) : undefined
  };
}
function normalizeSource(source) {
  if (typeof source === 'number') {
    // require() - not yet supported in native
    return undefined;
  }
  const drm = normalizeDrm(source.drm);
  const cmcd = normalizeCmcd(source.cmcd);
  const ad = normalizeAd(source.ad);
  return {
    ...source,
    headers: toNativeStringMapEntries(source.headers),
    drm,
    cmcd,
    ad
  };
}
function normalizeSelectedTrack(track) {
  if (!track) {
    return undefined;
  }
  return {
    type: track.type,
    value: track.value?.toString()
  };
}
function normalizeSelectedVideoTrack(track) {
  if (!track) {
    return undefined;
  }
  return {
    type: track.type,
    value: track.value
  };
}
function getPosterUri(poster) {
  if (!poster) {
    return undefined;
  }
  if (typeof poster === 'string') {
    return poster;
  }
  if (typeof poster === 'object' && 'uri' in poster && poster.uri) {
    return poster.uri;
  }
  const resolved = Image.resolveAssetSource(poster);
  return resolved?.uri;
}
const VideoBase = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    // Test ID
    testID,
    // Style
    style,
    // Progress
    progressUpdateInterval,
    // Source
    source,
    // Poster
    poster,
    posterResizeMode = 'contain',
    // Playback Control
    paused = false,
    muted = false,
    volume = 1.0,
    rate = 1.0,
    repeat = false,
    playInBackground = false,
    playWhenInactive = false,
    automaticallyWaitsToMinimizeStalling = true,
    shutterColor,
    // Display
    resizeMode = 'contain',
    viewType = 'surface',
    useTextureView = false,
    useSecureView = false,
    // Buffering
    bufferConfig,
    minLoadRetryCount = 3,
    maxBitRate,
    preferredForwardBufferDuration,
    // Track Selection
    selectedAudioTrack,
    selectedTextTrack,
    selectedVideoTrack,
    textTracks,
    // DRM
    drm,
    localSourceEncryptionKeyScheme,
    // Ads
    adTagUrl,
    adLanguage,
    // Controls
    controls = false,
    showNotificationControls = false,
    disableFocus = false,
    disableDisconnectError = false,
    focusable = true,
    hideShutterView = false,
    preventsDisplaySleepDuringVideoPlayback = true,
    // Fullscreen
    fullscreen = false,
    fullscreenAutorotate = true,
    fullscreenOrientation = 'all',
    // Picture in Picture
    pictureInPicture = false,
    // Content
    contentStartTime,
    allowsExternalPlayback = true,
    audioOutput = 'speaker',
    ignoreSilentSwitch = 'inherit',
    mixWithOthers = 'inherit',
    // Debug
    debug,
    // Events
    onLoadStart,
    onLoad,
    onError,
    onProgress,
    onSeek,
    onEnd,
    onBuffer,
    onBandwidthUpdate,
    onPlaybackStateChanged,
    onPlaybackRateChange,
    onVolumeChange,
    onIdle,
    onReadyForDisplay,
    onAudioFocusChanged,
    onAudioBecomingNoisy,
    onFullscreenPlayerWillPresent,
    onFullscreenPlayerDidPresent,
    onFullscreenPlayerWillDismiss,
    onFullscreenPlayerDidDismiss,
    onPictureInPictureStatusChanged,
    onRestoreUserInterfaceForPictureInPictureStop,
    onControlsVisibilityChange,
    onExternalPlaybackChange,
    onAspectRatio,
    onTransferEnd
  } = props;
  const nativeRef = useRef(null);

  // === Imperative Handle ===
  useImperativeHandle(ref, () => ({
    seek: (time, tolerance) => {
      if (nativeRef.current) {
        Commands.seek(nativeRef.current, time, tolerance ?? 0);
      }
    },
    pause: () => {
      if (nativeRef.current) {
        Commands.pause(nativeRef.current);
      }
    },
    resume: () => {
      if (nativeRef.current) {
        Commands.resume(nativeRef.current);
      }
    },
    setVolume: vol => {
      if (nativeRef.current) {
        Commands.adjustVolume(nativeRef.current, vol);
      }
    },
    setFullScreen: fs => {
      if (nativeRef.current) {
        Commands.setFullScreen(nativeRef.current, fs);
      }
    },
    presentFullscreenPlayer: () => {
      if (nativeRef.current) {
        Commands.setFullScreen(nativeRef.current, true);
      }
    },
    dismissFullscreenPlayer: () => {
      if (nativeRef.current) {
        Commands.setFullScreen(nativeRef.current, false);
      }
    },
    enterPictureInPicture: () => {
      if (nativeRef.current) {
        Commands.enterPictureInPicture(nativeRef.current);
      }
    },
    exitPictureInPicture: () => {
      if (nativeRef.current) {
        Commands.exitPictureInPicture(nativeRef.current);
      }
    },
    setSource: newSource => {
      if (nativeRef.current && newSource.uri) {
        Commands.loadSource(nativeRef.current, newSource.uri);
      }
    },
    getCurrentPosition: async () => {
      const handle = findNodeHandle(nativeRef.current);
      if (handle && GraniteVideoModule?.getCurrentPosition) {
        return GraniteVideoModule.getCurrentPosition(handle);
      }
      return 0;
    },
    save: async options => {
      const handle = findNodeHandle(nativeRef.current);
      if (handle && GraniteVideoModule?.save) {
        return GraniteVideoModule.save(handle, options ?? {});
      }
      return {
        uri: ''
      };
    },
    restoreUserInterfaceForPictureInPictureStopCompleted: () => {
      // iOS specific - handled internally
    }
  }));

  // === Event Handlers ===
  const handleLoadStart = useCallback(event => {
    onLoadStart?.(event.nativeEvent);
  }, [onLoadStart]);
  const handleLoad = useCallback(event => {
    const nativeEvent = event.nativeEvent;
    const loadData = {
      ...nativeEvent,
      naturalSize: {
        ...nativeEvent.naturalSize,
        orientation: nativeEvent.naturalSize.orientation
      }
    };
    onLoad?.(loadData);
  }, [onLoad]);
  const handleError = useCallback(event => {
    onError?.(event.nativeEvent);
  }, [onError]);
  const handleProgress = useCallback(event => {
    onProgress?.(event.nativeEvent);
  }, [onProgress]);
  const handleSeek = useCallback(event => {
    onSeek?.(event.nativeEvent);
  }, [onSeek]);
  const handleEnd = useCallback(() => {
    onEnd?.();
  }, [onEnd]);
  const handleBuffer = useCallback(event => {
    onBuffer?.(event.nativeEvent);
  }, [onBuffer]);
  const handleBandwidthUpdate = useCallback(event => {
    onBandwidthUpdate?.(event.nativeEvent);
  }, [onBandwidthUpdate]);
  const handlePlaybackStateChanged = useCallback(event => {
    onPlaybackStateChanged?.(event.nativeEvent);
  }, [onPlaybackStateChanged]);
  const handlePlaybackRateChange = useCallback(event => {
    onPlaybackRateChange?.(event.nativeEvent);
  }, [onPlaybackRateChange]);
  const handleVolumeChange = useCallback(event => {
    onVolumeChange?.(event.nativeEvent);
  }, [onVolumeChange]);
  const handleIdle = useCallback(() => {
    onIdle?.();
  }, [onIdle]);
  const handleReadyForDisplay = useCallback(() => {
    onReadyForDisplay?.();
  }, [onReadyForDisplay]);
  const handleAudioFocusChanged = useCallback(event => {
    onAudioFocusChanged?.(event.nativeEvent);
  }, [onAudioFocusChanged]);
  const handleAudioBecomingNoisy = useCallback(() => {
    onAudioBecomingNoisy?.();
  }, [onAudioBecomingNoisy]);
  const handleFullscreenPlayerWillPresent = useCallback(() => {
    onFullscreenPlayerWillPresent?.();
  }, [onFullscreenPlayerWillPresent]);
  const handleFullscreenPlayerDidPresent = useCallback(() => {
    onFullscreenPlayerDidPresent?.();
  }, [onFullscreenPlayerDidPresent]);
  const handleFullscreenPlayerWillDismiss = useCallback(() => {
    onFullscreenPlayerWillDismiss?.();
  }, [onFullscreenPlayerWillDismiss]);
  const handleFullscreenPlayerDidDismiss = useCallback(() => {
    onFullscreenPlayerDidDismiss?.();
  }, [onFullscreenPlayerDidDismiss]);
  const handlePictureInPictureStatusChanged = useCallback(event => {
    onPictureInPictureStatusChanged?.(event.nativeEvent);
  }, [onPictureInPictureStatusChanged]);
  const handleRestoreUserInterfaceForPictureInPictureStop = useCallback(() => {
    onRestoreUserInterfaceForPictureInPictureStop?.();
  }, [onRestoreUserInterfaceForPictureInPictureStop]);
  const handleControlsVisibilityChange = useCallback(event => {
    onControlsVisibilityChange?.(event.nativeEvent);
  }, [onControlsVisibilityChange]);
  const handleExternalPlaybackChange = useCallback(event => {
    onExternalPlaybackChange?.(event.nativeEvent);
  }, [onExternalPlaybackChange]);
  const handleAspectRatio = useCallback(event => {
    onAspectRatio?.(event.nativeEvent);
  }, [onAspectRatio]);
  const handleTransferEnd = useCallback(event => {
    onTransferEnd?.(event.nativeEvent);
  }, [onTransferEnd]);

  // === Render ===
  const containerStyle = [styles.container, style];
  return /*#__PURE__*/_jsx(View, {
    style: containerStyle,
    testID: testID,
    children: /*#__PURE__*/_jsx(NativeGraniteVideoView, {
      ref: nativeRef,
      style: styles.video,
      progressUpdateInterval: progressUpdateInterval,
      source: normalizeSource(source),
      poster: getPosterUri(poster),
      posterResizeMode: posterResizeMode,
      paused: paused,
      muted: muted,
      volume: volume,
      rate: rate,
      repeat: repeat,
      playInBackground: playInBackground,
      playWhenInactive: playWhenInactive,
      automaticallyWaitsToMinimizeStalling: automaticallyWaitsToMinimizeStalling,
      shutterColor: shutterColor,
      resizeMode: resizeMode,
      viewType: viewType,
      useTextureView: useTextureView,
      useSecureView: useSecureView,
      bufferConfig: bufferConfig,
      minLoadRetryCount: minLoadRetryCount,
      maxBitRate: maxBitRate,
      preferredForwardBufferDuration: preferredForwardBufferDuration,
      selectedAudioTrack: normalizeSelectedTrack(selectedAudioTrack),
      selectedTextTrack: normalizeSelectedTrack(selectedTextTrack),
      selectedVideoTrack: normalizeSelectedVideoTrack(selectedVideoTrack),
      textTracks: textTracks,
      drm: normalizeDrm(drm),
      localSourceEncryptionKeyScheme: localSourceEncryptionKeyScheme,
      adTagUrl: adTagUrl,
      adLanguage: adLanguage,
      controls: controls,
      showNotificationControls: showNotificationControls,
      disableFocus: disableFocus,
      disableDisconnectError: disableDisconnectError,
      focusable: focusable,
      hideShutterView: hideShutterView,
      preventsDisplaySleepDuringVideoPlayback: preventsDisplaySleepDuringVideoPlayback,
      fullscreen: fullscreen,
      fullscreenAutorotate: fullscreenAutorotate,
      fullscreenOrientation: fullscreenOrientation,
      pictureInPicture: pictureInPicture,
      contentStartTime: contentStartTime,
      allowsExternalPlayback: allowsExternalPlayback,
      audioOutput: audioOutput,
      ignoreSilentSwitch: ignoreSilentSwitch,
      mixWithOthers: mixWithOthers,
      enableDebug: debug?.enable,
      enableDebugThread: debug?.thread,
      onVideoLoadStart: handleLoadStart,
      onVideoLoad: handleLoad,
      onVideoError: handleError,
      onVideoProgress: handleProgress,
      onVideoSeek: handleSeek,
      onVideoEnd: handleEnd,
      onVideoBuffer: handleBuffer,
      onVideoBandwidthUpdate: handleBandwidthUpdate,
      onVideoPlaybackStateChanged: handlePlaybackStateChanged,
      onVideoPlaybackRateChange: handlePlaybackRateChange,
      onVideoVolumeChange: handleVolumeChange,
      onVideoIdle: handleIdle,
      onVideoReadyForDisplay: handleReadyForDisplay,
      onVideoAudioFocusChanged: handleAudioFocusChanged,
      onVideoAudioBecomingNoisy: handleAudioBecomingNoisy,
      onVideoFullscreenPlayerWillPresent: handleFullscreenPlayerWillPresent,
      onVideoFullscreenPlayerDidPresent: handleFullscreenPlayerDidPresent,
      onVideoFullscreenPlayerWillDismiss: handleFullscreenPlayerWillDismiss,
      onVideoFullscreenPlayerDidDismiss: handleFullscreenPlayerDidDismiss,
      onVideoPictureInPictureStatusChanged: handlePictureInPictureStatusChanged,
      onVideoRestoreUserInterfaceForPictureInPictureStop: handleRestoreUserInterfaceForPictureInPictureStop,
      onVideoControlsVisibilityChange: handleControlsVisibilityChange,
      onVideoExternalPlaybackChange: handleExternalPlaybackChange,
      onVideoAspectRatio: handleAspectRatio,
      onTransferEnd: handleTransferEnd
    })
  });
});
VideoBase.displayName = 'Video';

// Type for GraniteVideo with static properties

// Static property to indicate availability
// For Fabric (New Architecture), the component is always available
VideoBase.isAvailable = true;
export const Video = VideoBase;
const styles = StyleSheet.create({
  container: {
    overflow: 'hidden'
  },
  video: {
    ...StyleSheet.absoluteFillObject
  }
});

// === Static Methods ===
export async function clearCache() {
  if (GraniteVideoModule?.clearCache) {
    return GraniteVideoModule.clearCache();
  }
}
export async function getWidevineLevel() {
  if (Platform.OS === 'android' && GraniteVideoModule?.getWidevineLevel) {
    return GraniteVideoModule.getWidevineLevel();
  }
  return 0;
}
export async function isCodecSupported(mimeType, width, height) {
  if (GraniteVideoModule?.isCodecSupported) {
    return GraniteVideoModule.isCodecSupported(mimeType, width, height);
  }
  return false;
}
export async function isHEVCSupported() {
  if (GraniteVideoModule?.isHEVCSupported) {
    return GraniteVideoModule.isHEVCSupported();
  }
  return false;
}
