// @apps-in-toss/ait-format-proto/dist/ait/v1/ait.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v1.181.2
//   protoc               v6.33.4
// source: ait/v1/ait.proto
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AITBundle = exports.AITMetadata = exports.ZipIndexEntry = exports.FileAttrs_ExtraEntry = exports.FileAttrs = exports.Permission = exports.PlatformType = exports.CompressionMethod = void 0;
exports.compressionMethodFromJSON = compressionMethodFromJSON;
exports.compressionMethodToJSON = compressionMethodToJSON;
exports.platformTypeFromJSON = platformTypeFromJSON;
exports.platformTypeToJSON = platformTypeToJSON;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const struct_1 = require("../../google/protobuf/struct");
/** Compression method (values chosen to match common zip codes where possible) */
var CompressionMethod;
(function (CompressionMethod) {
    CompressionMethod[CompressionMethod["COMPRESSION_UNSPECIFIED"] = 0] = "COMPRESSION_UNSPECIFIED";
    CompressionMethod[CompressionMethod["NONE"] = 1] = "NONE";
    CompressionMethod[CompressionMethod["DEFLATE"] = 8] = "DEFLATE";
    CompressionMethod[CompressionMethod["GZIP"] = 9] = "GZIP";
    CompressionMethod[CompressionMethod["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(CompressionMethod || (exports.CompressionMethod = CompressionMethod = {}));
function compressionMethodFromJSON(object) {
    switch (object) {
        case 0:
        case "COMPRESSION_UNSPECIFIED":
            return CompressionMethod.COMPRESSION_UNSPECIFIED;
        case 1:
        case "NONE":
            return CompressionMethod.NONE;
        case 8:
        case "DEFLATE":
            return CompressionMethod.DEFLATE;
        case 9:
        case "GZIP":
            return CompressionMethod.GZIP;
        case -1:
        case "UNRECOGNIZED":
        default:
            return CompressionMethod.UNRECOGNIZED;
    }
}
function compressionMethodToJSON(object) {
    switch (object) {
        case CompressionMethod.COMPRESSION_UNSPECIFIED:
            return "COMPRESSION_UNSPECIFIED";
        case CompressionMethod.NONE:
            return "NONE";
        case CompressionMethod.DEFLATE:
            return "DEFLATE";
        case CompressionMethod.GZIP:
            return "GZIP";
        case CompressionMethod.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
/** Platform type */
var PlatformType;
(function (PlatformType) {
    PlatformType[PlatformType["PLATFORM_UNSPECIFIED"] = 0] = "PLATFORM_UNSPECIFIED";
    PlatformType[PlatformType["WEB"] = 1] = "WEB";
    PlatformType[PlatformType["REACT_NATIVE"] = 2] = "REACT_NATIVE";
    PlatformType[PlatformType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(PlatformType || (exports.PlatformType = PlatformType = {}));
function platformTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "PLATFORM_UNSPECIFIED":
            return PlatformType.PLATFORM_UNSPECIFIED;
        case 1:
        case "WEB":
            return PlatformType.WEB;
        case 2:
        case "REACT_NATIVE":
            return PlatformType.REACT_NATIVE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return PlatformType.UNRECOGNIZED;
    }
}
function platformTypeToJSON(object) {
    switch (object) {
        case PlatformType.PLATFORM_UNSPECIFIED:
            return "PLATFORM_UNSPECIFIED";
        case PlatformType.WEB:
            return "WEB";
        case PlatformType.REACT_NATIVE:
            return "REACT_NATIVE";
        case PlatformType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePermission() {
    return { name: "", access: "" };
}
exports.Permission = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.access !== "") {
            writer.uint32(18).string(message.access);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePermission();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.access = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            access: isSet(object.access) ? globalThis.String(object.access) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.access !== "") {
            obj.access = message.access;
        }
        return obj;
    },
    create(base) {
        return exports.Permission.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePermission();
        message.name = object.name ?? "";
        message.access = object.access ?? "";
        return message;
    },
};
function createBaseFileAttrs() {
    return { modifiedTimeMs: 0n, createdTimeMs: 0n, mimeType: "", extra: {} };
}
exports.FileAttrs = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.modifiedTimeMs !== 0n) {
            if (BigInt.asIntN(64, message.modifiedTimeMs) !== message.modifiedTimeMs) {
                throw new globalThis.Error("value provided for field message.modifiedTimeMs of type int64 too large");
            }
            writer.uint32(8).int64(message.modifiedTimeMs.toString());
        }
        if (message.createdTimeMs !== 0n) {
            if (BigInt.asIntN(64, message.createdTimeMs) !== message.createdTimeMs) {
                throw new globalThis.Error("value provided for field message.createdTimeMs of type int64 too large");
            }
            writer.uint32(16).int64(message.createdTimeMs.toString());
        }
        if (message.mimeType !== "") {
            writer.uint32(26).string(message.mimeType);
        }
        Object.entries(message.extra).forEach(([key, value]) => {
            exports.FileAttrs_ExtraEntry.encode({ key: key, value }, writer.uint32(34).fork()).ldelim();
        });
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFileAttrs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.modifiedTimeMs = longToBigint(reader.int64());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.createdTimeMs = longToBigint(reader.int64());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.mimeType = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    const entry4 = exports.FileAttrs_ExtraEntry.decode(reader, reader.uint32());
                    if (entry4.value !== undefined) {
                        message.extra[entry4.key] = entry4.value;
                    }
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            modifiedTimeMs: isSet(object.modifiedTimeMs) ? BigInt(object.modifiedTimeMs) : 0n,
            createdTimeMs: isSet(object.createdTimeMs) ? BigInt(object.createdTimeMs) : 0n,
            mimeType: isSet(object.mimeType) ? globalThis.String(object.mimeType) : "",
            extra: isObject(object.extra)
                ? Object.entries(object.extra).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.modifiedTimeMs !== 0n) {
            obj.modifiedTimeMs = message.modifiedTimeMs.toString();
        }
        if (message.createdTimeMs !== 0n) {
            obj.createdTimeMs = message.createdTimeMs.toString();
        }
        if (message.mimeType !== "") {
            obj.mimeType = message.mimeType;
        }
        if (message.extra) {
            const entries = Object.entries(message.extra);
            if (entries.length > 0) {
                obj.extra = {};
                entries.forEach(([k, v]) => {
                    obj.extra[k] = v;
                });
            }
        }
        return obj;
    },
    create(base) {
        return exports.FileAttrs.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseFileAttrs();
        message.modifiedTimeMs = object.modifiedTimeMs ?? 0n;
        message.createdTimeMs = object.createdTimeMs ?? 0n;
        message.mimeType = object.mimeType ?? "";
        message.extra = Object.entries(object.extra ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = globalThis.String(value);
            }
            return acc;
        }, {});
        return message;
    },
};
function createBaseFileAttrs_ExtraEntry() {
    return { key: "", value: "" };
}
exports.FileAttrs_ExtraEntry = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFileAttrs_ExtraEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
    create(base) {
        return exports.FileAttrs_ExtraEntry.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseFileAttrs_ExtraEntry();
        message.key = object.key ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
function createBaseZipIndexEntry() {
    return {
        name: "",
        zipOffset: 0n,
        compressedSize: 0n,
        uncompressedSize: 0n,
        compMethod: 0,
        sha256Hex: "",
        attrs: undefined,
    };
}
exports.ZipIndexEntry = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.zipOffset !== 0n) {
            if (BigInt.asUintN(64, message.zipOffset) !== message.zipOffset) {
                throw new globalThis.Error("value provided for field message.zipOffset of type uint64 too large");
            }
            writer.uint32(16).uint64(message.zipOffset.toString());
        }
        if (message.compressedSize !== 0n) {
            if (BigInt.asUintN(64, message.compressedSize) !== message.compressedSize) {
                throw new globalThis.Error("value provided for field message.compressedSize of type uint64 too large");
            }
            writer.uint32(24).uint64(message.compressedSize.toString());
        }
        if (message.uncompressedSize !== 0n) {
            if (BigInt.asUintN(64, message.uncompressedSize) !== message.uncompressedSize) {
                throw new globalThis.Error("value provided for field message.uncompressedSize of type uint64 too large");
            }
            writer.uint32(32).uint64(message.uncompressedSize.toString());
        }
        if (message.compMethod !== 0) {
            writer.uint32(40).int32(message.compMethod);
        }
        if (message.sha256Hex !== "") {
            writer.uint32(50).string(message.sha256Hex);
        }
        if (message.attrs !== undefined) {
            exports.FileAttrs.encode(message.attrs, writer.uint32(58).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseZipIndexEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.zipOffset = longToBigint(reader.uint64());
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.compressedSize = longToBigint(reader.uint64());
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.uncompressedSize = longToBigint(reader.uint64());
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.compMethod = reader.int32();
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.sha256Hex = reader.string();
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.attrs = exports.FileAttrs.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            zipOffset: isSet(object.zipOffset) ? BigInt(object.zipOffset) : 0n,
            compressedSize: isSet(object.compressedSize) ? BigInt(object.compressedSize) : 0n,
            uncompressedSize: isSet(object.uncompressedSize) ? BigInt(object.uncompressedSize) : 0n,
            compMethod: isSet(object.compMethod) ? compressionMethodFromJSON(object.compMethod) : 0,
            sha256Hex: isSet(object.sha256Hex) ? globalThis.String(object.sha256Hex) : "",
            attrs: isSet(object.attrs) ? exports.FileAttrs.fromJSON(object.attrs) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.zipOffset !== 0n) {
            obj.zipOffset = message.zipOffset.toString();
        }
        if (message.compressedSize !== 0n) {
            obj.compressedSize = message.compressedSize.toString();
        }
        if (message.uncompressedSize !== 0n) {
            obj.uncompressedSize = message.uncompressedSize.toString();
        }
        if (message.compMethod !== 0) {
            obj.compMethod = compressionMethodToJSON(message.compMethod);
        }
        if (message.sha256Hex !== "") {
            obj.sha256Hex = message.sha256Hex;
        }
        if (message.attrs !== undefined) {
            obj.attrs = exports.FileAttrs.toJSON(message.attrs);
        }
        return obj;
    },
    create(base) {
        return exports.ZipIndexEntry.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseZipIndexEntry();
        message.name = object.name ?? "";
        message.zipOffset = object.zipOffset ?? 0n;
        message.compressedSize = object.compressedSize ?? 0n;
        message.uncompressedSize = object.uncompressedSize ?? 0n;
        message.compMethod = object.compMethod ?? 0;
        message.sha256Hex = object.sha256Hex ?? "";
        message.attrs = (object.attrs !== undefined && object.attrs !== null)
            ? exports.FileAttrs.fromPartial(object.attrs)
            : undefined;
        return message;
    },
};
function createBaseAITMetadata() {
    return {
        isGame: false,
        platform: 0,
        runtimeVersion: "",
        bundleFiles: [],
        packageJson: undefined,
        extra: undefined,
        sdkVersion: undefined,
    };
}
exports.AITMetadata = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.isGame !== false) {
            writer.uint32(8).bool(message.isGame);
        }
        if (message.platform !== 0) {
            writer.uint32(16).int32(message.platform);
        }
        if (message.runtimeVersion !== "") {
            writer.uint32(26).string(message.runtimeVersion);
        }
        for (const v of message.bundleFiles) {
            writer.uint32(34).string(v);
        }
        if (message.packageJson !== undefined) {
            struct_1.Struct.encode(struct_1.Struct.wrap(message.packageJson), writer.uint32(42).fork()).ldelim();
        }
        if (message.extra !== undefined) {
            struct_1.Struct.encode(struct_1.Struct.wrap(message.extra), writer.uint32(50).fork()).ldelim();
        }
        if (message.sdkVersion !== undefined) {
            writer.uint32(58).string(message.sdkVersion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAITMetadata();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.isGame = reader.bool();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.platform = reader.int32();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.runtimeVersion = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.bundleFiles.push(reader.string());
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.packageJson = struct_1.Struct.unwrap(struct_1.Struct.decode(reader, reader.uint32()));
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.extra = struct_1.Struct.unwrap(struct_1.Struct.decode(reader, reader.uint32()));
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.sdkVersion = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            isGame: isSet(object.isGame) ? globalThis.Boolean(object.isGame) : false,
            platform: isSet(object.platform) ? platformTypeFromJSON(object.platform) : 0,
            runtimeVersion: isSet(object.runtimeVersion) ? globalThis.String(object.runtimeVersion) : "",
            bundleFiles: globalThis.Array.isArray(object?.bundleFiles)
                ? object.bundleFiles.map((e) => globalThis.String(e))
                : [],
            packageJson: isObject(object.packageJson) ? object.packageJson : undefined,
            extra: isObject(object.extra) ? object.extra : undefined,
            sdkVersion: isSet(object.sdkVersion) ? globalThis.String(object.sdkVersion) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.isGame !== false) {
            obj.isGame = message.isGame;
        }
        if (message.platform !== 0) {
            obj.platform = platformTypeToJSON(message.platform);
        }
        if (message.runtimeVersion !== "") {
            obj.runtimeVersion = message.runtimeVersion;
        }
        if (message.bundleFiles?.length) {
            obj.bundleFiles = message.bundleFiles;
        }
        if (message.packageJson !== undefined) {
            obj.packageJson = message.packageJson;
        }
        if (message.extra !== undefined) {
            obj.extra = message.extra;
        }
        if (message.sdkVersion !== undefined) {
            obj.sdkVersion = message.sdkVersion;
        }
        return obj;
    },
    create(base) {
        return exports.AITMetadata.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAITMetadata();
        message.isGame = object.isGame ?? false;
        message.platform = object.platform ?? 0;
        message.runtimeVersion = object.runtimeVersion ?? "";
        message.bundleFiles = object.bundleFiles?.map((e) => e) || [];
        message.packageJson = object.packageJson ?? undefined;
        message.extra = object.extra ?? undefined;
        message.sdkVersion = object.sdkVersion ?? undefined;
        return message;
    },
};
function createBaseAITBundle() {
    return {
        formatVersion: 0,
        deploymentId: "",
        appName: "",
        metadata: undefined,
        permissions: [],
        index: [],
        createdBy: "",
        createdAtMs: 0n,
    };
}
exports.AITBundle = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.formatVersion !== 0) {
            writer.uint32(8).int32(message.formatVersion);
        }
        if (message.deploymentId !== "") {
            writer.uint32(18).string(message.deploymentId);
        }
        if (message.appName !== "") {
            writer.uint32(26).string(message.appName);
        }
        if (message.metadata !== undefined) {
            exports.AITMetadata.encode(message.metadata, writer.uint32(34).fork()).ldelim();
        }
        for (const v of message.permissions) {
            exports.Permission.encode(v, writer.uint32(42).fork()).ldelim();
        }
        for (const v of message.index) {
            exports.ZipIndexEntry.encode(v, writer.uint32(50).fork()).ldelim();
        }
        if (message.createdBy !== "") {
            writer.uint32(58).string(message.createdBy);
        }
        if (message.createdAtMs !== 0n) {
            if (BigInt.asIntN(64, message.createdAtMs) !== message.createdAtMs) {
                throw new globalThis.Error("value provided for field message.createdAtMs of type int64 too large");
            }
            writer.uint32(64).int64(message.createdAtMs.toString());
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAITBundle();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.formatVersion = reader.int32();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.deploymentId = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.appName = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.metadata = exports.AITMetadata.decode(reader, reader.uint32());
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.permissions.push(exports.Permission.decode(reader, reader.uint32()));
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.index.push(exports.ZipIndexEntry.decode(reader, reader.uint32()));
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.createdBy = reader.string();
                    continue;
                case 8:
                    if (tag !== 64) {
                        break;
                    }
                    message.createdAtMs = longToBigint(reader.int64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            formatVersion: isSet(object.formatVersion) ? globalThis.Number(object.formatVersion) : 0,
            deploymentId: isSet(object.deploymentId) ? globalThis.String(object.deploymentId) : "",
            appName: isSet(object.appName) ? globalThis.String(object.appName) : "",
            metadata: isSet(object.metadata) ? exports.AITMetadata.fromJSON(object.metadata) : undefined,
            permissions: globalThis.Array.isArray(object?.permissions)
                ? object.permissions.map((e) => exports.Permission.fromJSON(e))
                : [],
            index: globalThis.Array.isArray(object?.index) ? object.index.map((e) => exports.ZipIndexEntry.fromJSON(e)) : [],
            createdBy: isSet(object.createdBy) ? globalThis.String(object.createdBy) : "",
            createdAtMs: isSet(object.createdAtMs) ? BigInt(object.createdAtMs) : 0n,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.formatVersion !== 0) {
            obj.formatVersion = Math.round(message.formatVersion);
        }
        if (message.deploymentId !== "") {
            obj.deploymentId = message.deploymentId;
        }
        if (message.appName !== "") {
            obj.appName = message.appName;
        }
        if (message.metadata !== undefined) {
            obj.metadata = exports.AITMetadata.toJSON(message.metadata);
        }
        if (message.permissions?.length) {
            obj.permissions = message.permissions.map((e) => exports.Permission.toJSON(e));
        }
        if (message.index?.length) {
            obj.index = message.index.map((e) => exports.ZipIndexEntry.toJSON(e));
        }
        if (message.createdBy !== "") {
            obj.createdBy = message.createdBy;
        }
        if (message.createdAtMs !== 0n) {
            obj.createdAtMs = message.createdAtMs.toString();
        }
        return obj;
    },
    create(base) {
        return exports.AITBundle.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAITBundle();
        message.formatVersion = object.formatVersion ?? 0;
        message.deploymentId = object.deploymentId ?? "";
        message.appName = object.appName ?? "";
        message.metadata = (object.metadata !== undefined && object.metadata !== null)
            ? exports.AITMetadata.fromPartial(object.metadata)
            : undefined;
        message.permissions = object.permissions?.map((e) => exports.Permission.fromPartial(e)) || [];
        message.index = object.index?.map((e) => exports.ZipIndexEntry.fromPartial(e)) || [];
        message.createdBy = object.createdBy ?? "";
        message.createdAtMs = object.createdAtMs ?? 0n;
        return message;
    },
};
function longToBigint(long) {
    return BigInt(long.toString());
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
