// @granite-js/mpack/dist/bundler/plugins/transformPlugin/steps/createTransformCodegenStep.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var createTransformCodegenStep_exports = {};
__export(createTransformCodegenStep_exports, {
  createTransformCodegenStep: () => createTransformCodegenStep
});
module.exports = __toCommonJS(createTransformCodegenStep_exports);
var import_core = require("@babel/core");
var import_defineStepName = require("../../../../utils/defineStepName");
function createTransformCodegenStep() {
  const codegenStep = async function codegen(code, args) {
    const transformedResult = await (0, import_core.transformAsync)(code, {
      test: /(?:^|[\\/])(?:Native\w+|(\w+)NativeComponent)\.[jt]sx?$/,
      filename: args.path,
      minified: false,
      compact: false,
      babelrc: false,
      configFile: false,
      plugins: [
        require.resolve("babel-plugin-syntax-hermes-parser"),
        require.resolve("@babel/plugin-transform-flow-strip-types"),
        [require.resolve("@babel/plugin-syntax-typescript"), false],
        require.resolve("@react-native/babel-plugin-codegen")
      ],
      overrides: [
        {
          test: /\.ts$/,
          plugins: [[require.resolve("@babel/plugin-syntax-typescript"), { isTSX: false, allowNamespaces: true }]]
        },
        {
          test: /\.tsx$/,
          plugins: [[require.resolve("@babel/plugin-syntax-typescript"), { isTSX: true, allowNamespaces: true }]]
        }
      ]
    });
    return { code: transformedResult?.code ?? code };
  };
  (0, import_defineStepName.defineStepName)(codegenStep, "transform-codegen");
  return codegenStep;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createTransformCodegenStep
});
