// @granite-js/mpack/dist/operations/OpenDebuggerKeyboardHandler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OpenDebuggerKeyboardHandler_exports = {};
__export(OpenDebuggerKeyboardHandler_exports, {
  default: () => OpenDebuggerKeyboardHandler
});
module.exports = __toCommonJS(OpenDebuggerKeyboardHandler_exports);
var import_util = require("util");
class OpenDebuggerKeyboardHandler {
  #devServerUrl;
  #reporter;
  #targetsShownForSelection = null;
  constructor({ devServerUrl, reporter }) {
    this.#devServerUrl = devServerUrl;
    this.#reporter = reporter;
  }
  async #tryOpenDebuggerForTarget(target) {
    this.#targetsShownForSelection = null;
    this.#clearTerminalMenu();
    try {
      const fetchFn = globalThis.fetch;
      if (fetchFn == null) {
        throw new Error("Global fetch is not available");
      }
      await fetchFn(new URL(`/open-debugger?target=${encodeURIComponent(target.id)}`, this.#devServerUrl).href, {
        method: "POST"
      });
    } catch (error) {
      this.#log("error", "Failed to open debugger for %s (%s): %s", target.title, target.description, "Network error");
      if (error instanceof Error && error.cause != null) {
        this.#log("error", "Cause: %s", error.cause);
      }
      this.#clearTerminalMenu();
    }
  }
  async handleOpenDebugger() {
    this.#setTerminalMenu("Fetching available debugging targets...");
    this.#targetsShownForSelection = null;
    try {
      const fetchFn = globalThis.fetch;
      if (fetchFn == null) {
        throw new Error("Global fetch is not available");
      }
      const response = await fetchFn(new URL("/json/list", this.#devServerUrl).href, { method: "POST" });
      if (response.status !== 200) {
        throw new Error(`Unexpected status code: ${response.status}`);
      }
      const targets = await response.json();
      if (!Array.isArray(targets)) {
        throw new Error("Expected array.");
      }
      if (targets.length === 0) {
        this.#log("warn", "No connected targets");
        this.#clearTerminalMenu();
      } else if (targets.length === 1) {
        const target = targets[0];
        if (target) {
          void this.#tryOpenDebuggerForTarget(target);
        } else {
          this.#clearTerminalMenu();
        }
      } else {
        this.#targetsShownForSelection = targets;
        if (targets.length > 9) {
          this.#log("warn", "10 or more debug targets available, showing the first 9.");
        }
        this.#setTerminalMenu(
          `Multiple debug targets available, please select:
  ${targets.slice(0, 9).map(({ title }, index) => `${(0, import_util.styleText)(["white", "inverse"], ` ${index + 1} `)} - "${title}"`).join("\n  ")}`
        );
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      this.#log("error", `Failed to fetch debug targets: ${message}`);
      this.#clearTerminalMenu();
    }
  }
  maybeHandleTargetSelection(keyName) {
    if (keyName >= "1" && keyName <= "9") {
      const targetIndex = Number(keyName) - 1;
      if (this.#targetsShownForSelection != null && targetIndex < this.#targetsShownForSelection.length) {
        const target = this.#targetsShownForSelection[targetIndex];
        if (target) {
          void this.#tryOpenDebuggerForTarget(target);
          return true;
        }
      }
    }
    return false;
  }
  dismiss() {
    this.#clearTerminalMenu();
    this.#targetsShownForSelection = null;
  }
  #log(level, ...data) {
    this.#reporter.update({
      type: "unstable_server_log",
      level,
      data
    });
  }
  #setTerminalMenu(message) {
    this.#reporter.update({
      type: "unstable_server_menu_updated",
      message
    });
  }
  #clearTerminalMenu() {
    this.#reporter.update({
      type: "unstable_server_menu_cleared"
    });
  }
}
