// @granite-js/image/dist/module/GraniteImage.js
"use strict";

import React, { useCallback } from 'react';
import { NativeModules } from 'react-native';
import GraniteImageNativeComponent from './GraniteImageNativeComponent';
import { jsx as _jsx } from "react/jsx-runtime";
const {
  GraniteImageModule
} = NativeModules;

// Source types matching GraniteImage

// Map resizeMode to contentMode for native
const resizeModeToContentMode = resizeMode => {
  return resizeMode || 'cover';
};

// Map GraniteImage cache to native cachePolicy
const mapCachePolicy = cache => {
  switch (cache) {
    case 'cacheOnly':
      return 'disk';
    case 'web':
      return 'none';
    case 'immutable':
    default:
      return 'disk';
  }
};

// Type declarations for static methods

// Static methods for cache management
const clearMemoryCache = () => {
  if (GraniteImageModule?.clearMemoryCache) {
    return GraniteImageModule.clearMemoryCache();
  }
  console.warn('GraniteImage.clearMemoryCache: Native module not available');
  return Promise.resolve();
};
const clearDiskCache = () => {
  if (GraniteImageModule?.clearDiskCache) {
    return GraniteImageModule.clearDiskCache();
  }
  console.warn('GraniteImage.clearDiskCache: Native module not available');
  return Promise.resolve();
};
const preload = sources => {
  if (GraniteImageModule?.preload) {
    const sourcesJson = JSON.stringify(sources);
    return GraniteImageModule.preload(sourcesJson);
  } else {
    console.warn('GraniteImage.preload: Native module not available');
    return Promise.resolve();
  }
};
const GraniteImageBase = ({
  source,
  resizeMode = 'cover',
  tintColor,
  style,
  defaultSource,
  fallbackSource,
  priority = 'normal',
  cachePolicy = 'disk',
  onLoadStart,
  onProgress,
  onLoad,
  onError,
  onLoadEnd,
  ...viewProps
}) => {
  // Parse source
  const uri = typeof source === 'string' ? source : (source && typeof source === 'object' ? source.uri : '') ?? '';
  const headers = typeof source === 'object' && source.headers ? JSON.stringify(source.headers) : undefined;
  const sourcePriority = typeof source === 'object' ? source.priority : undefined;
  const sourceCache = typeof source === 'object' ? source.cache : undefined;

  // Handle defaultSource
  const defaultSourceUri = typeof defaultSource === 'string' ? defaultSource : typeof defaultSource === 'number' ? undefined // Local require() - need native handling
  : undefined;

  // Handle fallbackSource
  const fallbackSourceUri = typeof fallbackSource === 'string' ? fallbackSource : typeof fallbackSource === 'number' ? undefined // Local require() - need native handling
  : undefined;

  // Event handlers
  const handleLoadStart = useCallback(event => {
    onLoadStart?.(event);
  }, [onLoadStart]);
  const handleProgress = useCallback(event => {
    onProgress?.(event);
  }, [onProgress]);
  const handleLoad = useCallback(event => {
    onLoad?.(event);
  }, [onLoad]);
  const handleError = useCallback(event => {
    onError?.(event);
  }, [onError]);
  const handleLoadEnd = useCallback(event => {
    onLoadEnd?.(event);
  }, [onLoadEnd]);
  return /*#__PURE__*/_jsx(GraniteImageNativeComponent, {
    ...viewProps,
    uri: uri,
    headers: headers,
    contentMode: resizeModeToContentMode(resizeMode),
    tintColor: tintColor,
    defaultSource: defaultSourceUri,
    fallbackSource: fallbackSourceUri,
    priority: sourcePriority || priority,
    cachePolicy: sourceCache ? mapCachePolicy(sourceCache) : cachePolicy,
    style: style,
    onGraniteLoadStart: onLoadStart ? handleLoadStart : undefined,
    onGraniteProgress: onProgress ? handleProgress : undefined,
    onGraniteLoad: onLoad ? handleLoad : undefined,
    onGraniteError: onError ? handleError : undefined,
    onGraniteLoadEnd: onLoadEnd ? handleLoadEnd : undefined
  });
};
export const GraniteImage = Object.assign(GraniteImageBase, {
  clearMemoryCache,
  clearDiskCache,
  preload
});
export default GraniteImage;
