// @apps-in-toss/native-modules/src/MiniAppModule/native-modules/index.ts
// AdMob V2
import {
  loadAppsInTossAdMob,
  showAppsInTossAdMob,
  type LoadAdMobParams,
  type LoadAdMobOptions,
  type LoadAdMobEvent,
  type ShowAdMobParams,
  type ShowAdMobEvent,
  type ShowAdMobOptions,
  isAppsInTossAdMobLoaded,
  type IsAdMobLoadedOptions,
} from './ads/googleAdMobV2';
// TossPay
import { checkoutPayment, type CheckoutPaymentOptions, type CheckoutPaymentResult } from './checkoutPayment';

export * from './appLogin';
export * from './eventLog';

/** permission functions */
export * from './permissions/fetchAlbumPhotos/fetchAlbumPhotos';
export * from './permissions/fetchContacts/fetchContacts';
export * from './permissions/getClipboardText/getClipboardText';
export * from './permissions/getCurrentLocation/getCurrentLocation';
export * from './permissions/setClipboardText/setClipboardText';
export * from './permissions/openCamera/openCamera';

export * from './getDeviceId';
export * from './getOperationalEnvironment';
export * from './getTossAppVersion';
export * from './getTossShareLink';
export * from './iap';
export * from './isMinVersionSupported';
export * from './saveBase64Data';
export * from './setDeviceOrientation';
export * from './storage';
export * from './openGameCenterLeaderboard';
export * from './getGameCenterGameProfile';
export * from './submitGameCenterLeaderBoardScore';
export * from './getAnonymousKey';
export * from './getUserKeyForGame';
export * from './grantPromotionReward';
export * from './grantPromotionRewardForGame';
export * from './getIsTossLoginIntegratedService';
export * from '../native-event-emitter/contactsViral';
export * from './appsInTossSignTossCert';
export * from './getGroupId';
export * from './shareWithScheme';
export * from './getServerTime';
export * from './requestReview';

export type {
  CheckoutPaymentOptions,
  CheckoutPaymentResult,
  LoadAdMobParams,
  LoadAdMobOptions,
  LoadAdMobEvent,
  ShowAdMobParams,
  ShowAdMobEvent,
  ShowAdMobOptions,
  IsAdMobLoadedOptions,
};

/**
 * @public
 * @category 토스페이
 * @name TossPay
 * @description 토스페이 결제 관련 함수를 모아둔 객체예요.
 * @property {typeof checkoutPayment} [checkoutPayment] 토스페이 결제를 인증하는 함수예요. 자세한 내용은 [checkoutPayment](/react-native/reference/native-modules/토스페이/checkoutPayment)를 참고하세요.
 */
export const TossPay = {
  checkoutPayment,
};

export * from './checkoutPayment';

/**
 * @public
 * @category 광고
 * @name GoogleAdMob
 * @description Google AdMob 광고 관련 함수를 모아둔 객체예요.
 * @property {typeof loadAppsInTossAdMob} [loadAppsInTossAdMob] 광고를 로드하는 함수예요.
 * @property {typeof showAppsInTossAdMob} [showAppsInTossAdMob] 로드한 광고를 보여주는 함수예요.
 * @property {typeof isAppsInTossAdMobLoaded} [isAppsInTossAdMobLoaded] 광고가 로드되었는지 확인하는 함수예요.
 */
export const GoogleAdMob = {
  loadAppsInTossAdMob,
  showAppsInTossAdMob,
  isAppsInTossAdMobLoaded,
};
