// @granite-js/mpack/dist/experimental/operations/serve.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var serve_exports = {};
__export(serve_exports, {
  EXPERIMENTAL__server: () => EXPERIMENTAL__server
});
module.exports = __toCommonJS(serve_exports);
var import_plugin_core = require("@granite-js/plugin-core");
var import_constants = require("../../constants");
var import_attachKeyHandlers = __toESM(require("../../operations/attachKeyHandlers"));
var import_keyReporter = require("../../operations/keyReporter");
var import_printLogo = require("../../utils/printLogo");
var import_printServerUrl = require("../../utils/printServerUrl");
var import_DevServer = require("../server/DevServer");
let keyHandlersAttached = false;
async function EXPERIMENTAL__server({
  config,
  host = import_constants.DEV_SERVER_DEFAULT_HOST,
  port = import_constants.DEV_SERVER_DEFAULT_PORT,
  onServerReady
}) {
  const driver = (0, import_plugin_core.createPluginHooksDriver)(config);
  await driver.devServer.pre({ host, port });
  const rootDir = config.cwd;
  const context = { command: "serve" };
  const { metro: _, devServer, ...buildConfig } = await (0, import_plugin_core.resolveConfig)(config, context) ?? {};
  const server = new import_DevServer.DevServer({
    buildConfig: { entry: config.entryFile, ...buildConfig },
    middlewares: devServer?.middlewares ?? [],
    host,
    port,
    rootDir
  });
  (0, import_printLogo.printLogo)();
  await server.initialize();
  await server.listen();
  await driver.devServer.post({ host, port });
  (0, import_printServerUrl.printServerUrl)({ host, port });
  if (!keyHandlersAttached) {
    keyHandlersAttached = true;
    const devServerHostname = host === "0.0.0.0" ? "localhost" : host;
    const devServerUrl = new URL(`http://${devServerHostname}:${port}`).origin;
    (0, import_attachKeyHandlers.default)({
      devServerUrl,
      messageSocket: {
        broadcast: (command, params) => server.broadcastCommand(command, params ?? void 0)
      },
      reporter: import_keyReporter.keyReporter
    });
    await onServerReady?.();
  }
  return {
    cleanup: async () => {
      await server.close();
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EXPERIMENTAL__server
});
