// @granite-js/plugin-micro-frontend/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let fs = require("fs");
fs = __toESM(fs);
let path = require("path");
path = __toESM(path);
let _granite_js_utils = require("@granite-js/utils");
let picocolors = require("picocolors");
picocolors = __toESM(picocolors);

//#region src/prelude.ts
function getPreludeConfig(options) {
	const registerStatements = Object.entries(options.shared ?? {}).filter(([_, config]) => config.eager === true).map(([libName], index) => {
		const identifier = `__mod${index}`;
		return `
    // ${libName}
    import * as ${identifier} from '${libName}';
    registerShared('${libName}', ${identifier});
    `;
	});
	const exposeStatements = Object.entries(options.exposes ?? {}).map(([exposeName, modulePath], index) => {
		const identifier = `__expose${index}`;
		return `
    import * as ${identifier} from '${path.default.resolve(modulePath)}';
    exposeModule(__container, '${exposeName}', ${identifier});
    `;
	});
	return {
		banner: `
    if (global.__MICRO_FRONTEND__ == null) {
      global.__MICRO_FRONTEND__ = {
        __SHARED__: {},
        __INSTANCES__: [],
      };
    }
    `,
		preludeScript: [
			`import { registerShared, createContainer, exposeModule } from '@granite-js/plugin-micro-frontend/runtime';`,
			`const __container = createContainer('${options.name}', ${JSON.stringify({
				remote: options.remote,
				shared: options.shared
			})});`,
			...registerStatements,
			...exposeStatements
		].join("\n")
	};
}

//#endregion
//#region src/log.ts
const tag = picocolors.default.bold(picocolors.default.bgCyan(picocolors.default.black(" MICRO FRONTEND ")));
function log(...args) {
	console.log(tag, ...args);
}

//#endregion
//#region src/remote.ts
const FALLBACK_SCRIPT = `console.warn('[MICRO FRONTEND] Failed to fetch remote bundles. Please check if the remote dev server is running')`;
async function fetchRemoteBundle(remote) {
	globalThis.remoteBundles = {
		android: FALLBACK_SCRIPT,
		ios: FALLBACK_SCRIPT
	};
	try {
		log("Prefetching remote bundles for development environment...");
		const [androidBundle, iosBundle] = await Promise.all([fetchBundle(remote, "android"), fetchBundle(remote, "ios")]);
		globalThis.remoteBundles = {
			android: androidBundle,
			ios: iosBundle
		};
		log("Fetch complete");
	} catch {
		log("Failed to fetch remote bundles. Please check if the remote dev server is running");
	}
}
async function fetchBundle(remote, platform) {
	return await (await fetch(`http://${remote.host}:${remote.port}/index.bundle?dev=true&platform=${platform}`)).text();
}

//#endregion
//#region src/resolver.ts
const VIRTUAL_SHARED_PROTOCOL = "virtual-shared";
function virtualSharedConfig(moduleEntries) {
	const alias = moduleEntries.map(([libName]) => ({
		from: libName,
		to: `${VIRTUAL_SHARED_PROTOCOL}:${libName}`,
		exact: true
	}));
	const protocols = { [VIRTUAL_SHARED_PROTOCOL]: { load: function virtualSharedProtocolLoader(args) {
		return {
			loader: "js",
			contents: `
          var sharedModule = global.__MICRO_FRONTEND__.__SHARED__['${args.path}'];

          if (sharedModule == null) {
            throw new Error("'${args.path}' is not registered in the shared registry");
          }

          module.exports = sharedModule.get();
          `
		};
	} } };
	return {
		alias,
		protocols: alias.length > 0 ? protocols : void 0
	};
}

//#endregion
//#region src/utils/intoShared.ts
const SHARED_PRESETS = {
	"react-native": [
		"react-native/Libraries/BatchedBridge/BatchedBridge",
		"react-native/Libraries/BatchedBridge/NativeModules",
		"react-native/Libraries/BatchedBridge/MessageQueue",
		"react-native/Libraries/NativeComponent/NativeComponentRegistry",
		"react-native/Libraries/NativeComponent/NativeComponentRegistryUnstable",
		"react-native/Libraries/NativeComponent/ViewConfigIgnore",
		"react-native/Libraries/ReactNative/RendererProxy",
		"react-native/Libraries/StyleSheet/PlatformColorValueTypes",
		"react-native/Libraries/StyleSheet/normalizeColor",
		"react-native/Libraries/StyleSheet/processColor",
		"react-native/Libraries/TurboModule/TurboModuleRegistry",
		"react-native/Libraries/Utilities/NativePlatformConstantsIOS",
		"react-native/Libraries/Utilities/Platform",
		"react-native/Libraries/Utilities/defineLazyObjectProperty"
	],
	react: ["react/jsx-runtime", "react/jsx-dev-runtime"]
};
function intoShared(shared) {
	if (shared == null) return;
	const normalized = Array.isArray(shared) ? shared.reduce((acc, lib) => {
		acc[lib] = {};
		return acc;
	}, {}) : { ...shared };
	for (const [lib, subpaths] of Object.entries(SHARED_PRESETS)) {
		const libConfig = normalized[lib];
		if (libConfig == null) continue;
		for (const subpath of subpaths) if (normalized[subpath] == null) normalized[subpath] = { ...libConfig };
	}
	return normalized;
}

//#endregion
//#region src/microFrontendPlugin.ts
const microFrontendPlugin = async (options) => {
	const sharedConfig = intoShared(options.shared);
	const nonEagerEntries = Object.entries(sharedConfig ?? {}).filter(([_, config]) => config.eager !== true);
	const rootDir = process.cwd();
	const preludeConfig = getPreludeConfig({
		...options,
		shared: sharedConfig
	});
	const localDir = (0, _granite_js_utils.prepareLocalDirectory)(rootDir);
	const preludePath = path.join(localDir, "micro-frontend-runtime.js");
	fs.writeFileSync(preludePath, preludeConfig.preludeScript);
	/**
	* @TODO `MPACK_DEV_SERVER` flag should be removed after next version of bundle loader is released and load bundle dynamically at JS runtime.
	*/
	if (process.env.MPACK_DEV_SERVER === "true" && options.remote) await fetchRemoteBundle(options.remote);
	/**
	* If importing `react-native` from the shared registry, skip its prelude scripts
	* to ensure the core is loaded only once per runtime by the host.
	*/
	const isReactNativeShared = Boolean(nonEagerEntries.find(([libName]) => libName === "react-native"));
	const virtualShared = virtualSharedConfig(nonEagerEntries);
	return {
		name: "micro-frontend-plugin",
		config: {
			extra: isReactNativeShared ? {
				skipReactNativePolyfills: true,
				skipReactNativeInitializeCore: true
			} : void 0,
			resolver: {
				alias: virtualShared.alias,
				protocols: virtualShared.protocols
			},
			esbuild: {
				prelude: [preludePath],
				banner: { js: preludeConfig.banner }
			}
		}
	};
};

//#endregion
exports.microFrontend = microFrontendPlugin;
