// @apps-in-toss/native-modules/src/spec/CommonModule.brick.ts
import { BrickModule, BrickModuleSpec } from 'brick-module';

interface ScreenAwakeModeParameter {
  enabled: boolean;
}

interface SecureScreenParameter {
  enabled: boolean;
}

interface HapticParameter {
  type: string;
}

interface IosSwipeGestureParameters {
  isEnabled: boolean;
}

interface ShareParameter {
  message: string;
}

interface LeftEdgeTouchEnabledParameter {
  isEnabled: boolean;
}

/**
 * CommonModule Spec
 * Unified type definition for CommonModule (formerly BedrockModule)
 */
export interface CommonModuleSpec extends BrickModuleSpec {
  readonly moduleName: 'CommonModule';

  getConstants(): {
    schemeUri: string;
    locale: string;
  };

  getNetworkStatus(): string;
  closeView(): void;
  generateHapticFeedback(parameter: HapticParameter): Promise<void>;
  share(parameter: ShareParameter): void;
  setSecureScreen(parameter: SecureScreenParameter): Promise<{ enabled: boolean }>;
  setScreenAwakeMode(parameter: ScreenAwakeModeParameter): Promise<{ enabled: boolean }>;
  setIosSwipeGestureEnabled(parameter: IosSwipeGestureParameters): Promise<void>;

  setLeftEdgeTouchEnabled(parameter: LeftEdgeTouchEnabledParameter): void;
}

export const CommonModule = BrickModule.get<CommonModuleSpec>('CommonModule');
