// @apps-in-toss/plugins/dist/index.js
import { createRequire } from 'module'; const require = createRequire(import.meta.url);
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __commonJS = (cb, mod) => function __require2() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// ../../../.yarn/__virtual__/tsup-virtual-5946fc3428/0/cache/tsup-npm-8.4.0-f78d2622c9-c6636ffd6a.zip/node_modules/tsup/assets/esm_shims.js
var init_esm_shims = __esm({
  "../../../.yarn/__virtual__/tsup-virtual-5946fc3428/0/cache/tsup-npm-8.4.0-f78d2622c9-c6636ffd6a.zip/node_modules/tsup/assets/esm_shims.js"() {
    "use strict";
  }
});

// ../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_isFormatUuid.js
var require_isFormatUuid = __commonJS({
  "../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_isFormatUuid.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports._isFormatUuid = void 0;
    var _isFormatUuid2 = (str) => PATTERN.test(str);
    exports._isFormatUuid = _isFormatUuid2;
    var PATTERN = /^(?:urn:uuid:)?[0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12}$/i;
  }
});

// ../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_validateReport.js
var require_validateReport = __commonJS({
  "../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_validateReport.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports._validateReport = void 0;
    var _validateReport2 = (array) => {
      const reportable = (path9) => {
        if (array.length === 0)
          return true;
        const last = array[array.length - 1].path;
        return path9.length > last.length || last.substring(0, path9.length) !== path9;
      };
      return (exceptable, error2) => {
        var _a;
        if (exceptable && reportable(error2.path)) {
          if (error2.value === void 0)
            (_a = error2.description) !== null && _a !== void 0 ? _a : error2.description = [
              "The value at this path is `undefined`.",
              "",
              `Please fill the \`${error2.expected}\` typed value next time.`
            ].join("\n");
          array.push(error2);
        }
        return false;
      };
    };
    exports._validateReport = _validateReport2;
  }
});

// ../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_createStandardSchema.js
var require_createStandardSchema = __commonJS({
  "../../../.yarn/__virtual__/typia-virtual-6e1bbef80b/0/cache/typia-npm-9.7.2-987d15a303-1b19baddfe.zip/node_modules/typia/lib/internal/_createStandardSchema.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports._createStandardSchema = void 0;
    var _createStandardSchema2 = (fn) => Object.assign(fn, {
      "~standard": {
        version: 1,
        vendor: "typia",
        validate: (input) => {
          const result = fn(input);
          if (result.success) {
            return {
              value: result.data
            };
          } else {
            return {
              issues: result.errors.map((error2) => ({
                message: `expected ${error2.expected}, got ${error2.value}`,
                path: typiaPathToStandardSchemaPath(error2.path)
              }))
            };
          }
        }
      }
    });
    exports._createStandardSchema = _createStandardSchema2;
    var PathParserState;
    (function(PathParserState2) {
      PathParserState2[PathParserState2["Start"] = 0] = "Start";
      PathParserState2[PathParserState2["Property"] = 1] = "Property";
      PathParserState2[PathParserState2["StringKey"] = 2] = "StringKey";
      PathParserState2[PathParserState2["NumberKey"] = 3] = "NumberKey";
    })(PathParserState || (PathParserState = {}));
    var typiaPathToStandardSchemaPath = (path9) => {
      if (!path9.startsWith("$input")) {
        throw new Error(`Invalid path: ${JSON.stringify(path9)}`);
      }
      const segments = [];
      let currentSegment = "";
      let state = PathParserState.Start;
      let index = "$input".length - 1;
      while (index < path9.length - 1) {
        index++;
        const char = path9[index];
        if (state === PathParserState.Property) {
          if (char === "." || char === "[") {
            segments.push({
              key: currentSegment
            });
            state = PathParserState.Start;
          } else if (index === path9.length - 1) {
            currentSegment += char;
            segments.push({
              key: currentSegment
            });
            index++;
            state = PathParserState.Start;
          } else {
            currentSegment += char;
          }
        } else if (state === PathParserState.StringKey) {
          if (char === '"') {
            segments.push({
              key: JSON.parse(currentSegment + char)
            });
            index += 2;
            state = PathParserState.Start;
          } else if (char === "\\") {
            currentSegment += path9[index];
            index++;
            currentSegment += path9[index];
          } else {
            currentSegment += char;
          }
        } else if (state === PathParserState.NumberKey) {
          if (char === "]") {
            segments.push({
              key: Number.parseInt(currentSegment)
            });
            index++;
            state = PathParserState.Start;
          } else {
            currentSegment += char;
          }
        }
        if (state === PathParserState.Start && index < path9.length - 1) {
          const newChar = path9[index];
          currentSegment = "";
          if (newChar === "[") {
            if (path9[index + 1] === '"') {
              state = PathParserState.StringKey;
              index++;
              currentSegment = '"';
            } else {
              state = PathParserState.NumberKey;
            }
          } else if (newChar === ".") {
            state = PathParserState.Property;
          } else {
            throw new Error("Unreachable: pointer points invalid character");
          }
        }
      }
      if (state !== PathParserState.Start) {
        throw new Error(`Failed to parse path: ${JSON.stringify(path9)}`);
      }
      return segments;
    };
  }
});

// package.json
var require_package = __commonJS({
  "package.json"(exports, module) {
    module.exports = {
      name: "@apps-in-toss/plugins",
      type: "module",
      version: "2.4.7",
      description: "The plugins for Apps In Toss",
      scripts: {
        test: "vitest --run",
        typecheck: "tsc --noEmit",
        lint: "eslint .",
        build: "tsup"
      },
      main: "./dist/index.cjs",
      module: "./dist/index.js",
      types: "./dist/index.d.ts",
      exports: {
        ".": {
          types: "./dist/index.d.ts",
          import: "./dist/index.js",
          require: "./dist/index.cjs"
        },
        "./internal": {
          types: "./dist/internal.d.ts",
          import: "./dist/internal.js",
          require: "./dist/internal.cjs"
        },
        "./package.json": "./package.json"
      },
      files: [
        "dist/**/*"
      ],
      devDependencies: {
        "@ryoppippi/unplugin-typia": "^2.6.5",
        "@types/archiver": "^6.0.3",
        "@types/connect": "^3",
        tsup: "^8.4.0",
        typescript: "5.8.3",
        typia: "^9.3.0",
        vitest: "^3.2.4"
      },
      dependencies: {
        "@apps-in-toss/ait-format": "1.0.0",
        "@apps-in-toss/plugin-compat": "workspace:*",
        "@granite-js/plugin-core": "catalog:granite-js",
        "@granite-js/plugin-micro-frontend": "catalog:granite-js",
        "@granite-js/plugin-sentry": "catalog:granite-js",
        "@granite-js/utils": "catalog:granite-js",
        archiver: "^7.0.1",
        connect: "^3.7.0",
        esbuild: "catalog:tooling",
        execa: "catalog:tooling",
        picocolors: "^1.1.1",
        uuidv7: "^1.0.2"
      }
    };
  }
});

// src/index.ts
init_esm_shims();

// src/appsInToss.ts
init_esm_shims();
import path8 from "path";
import {
  reactCompatibilityPlugin,
  reactNativeCompatibilityPlugin,
  reactNativeReverseCompatibilityPlugin
} from "@apps-in-toss/plugin-compat";
import { microFrontend } from "@granite-js/plugin-micro-frontend";
import { getPackageRoot as getPackageRoot6 } from "@granite-js/utils";

// src/constants.ts
init_esm_shims();
var REACT_NATIVE_VERSION = "0.84.0";
var APP_MANIFEST_NAME = "app.json";
var MICRO_FRONTEND_HOST_NAME = "apps-in-toss-host";
var MICRO_FRONTEND_REMOTE_NAME = "apps-in-toss-service";
var MICRO_FRONTEND_REMOTE_EXPOSE_PATH = "./ServiceEntry";
var HOST_CONTEXT_IDENTIFIER = "__appsInTossHost";
var SERVICE_CONTEXT_IDENTIFIER = "__appsInToss";
var getSharedPackages = ({ reactNativeVersion }) => [
  "@react-native-community/blur",
  "@react-navigation/native",
  "@react-navigation/native-stack",
  "@sentry/react-native",
  "@shopify/flash-list",
  "react-native-safe-area-context",
  "react-native-screens",
  "react-native-svg",
  "react-native-gesture-handler",
  "react-native",
  "react",
  "react-native-webview",
  "@granite-js/image",
  "@granite-js/lottie",
  "@granite-js/video",
  ...reactNativeVersion === "0.84" ? ["brick-module", "@react-navigation/elements"] : [],
  ...reactNativeVersion === "0.72" ? ["react-native-fast-image", "lottie-react-native", "react-native-video"] : []
];

// src/log.ts
init_esm_shims();
import pc from "picocolors";
var TAG = pc.bold(pc.cyan("[AppsInToss]"));
function log(...args) {
  console.log(TAG, ...args);
}
function error(...args) {
  console.error(TAG, ...args);
}

// src/plugins/index.ts
init_esm_shims();

// src/plugins/appJson.ts
init_esm_shims();
import fs from "node:fs";
import path from "path";
import { getPackageRoot } from "@granite-js/utils";
async function appsInTossAppJson(options) {
  const packageRoot = getPackageRoot();
  const handleAppJson = async (appName) => {
    const appJsonPath = path.join(packageRoot, ".granite", APP_MANIFEST_NAME);
    const appJsonObject = {
      appName,
      permissions: options.permissions,
      ...options.appType != null ? { appType: options.appType } : {}
    };
    await fs.promises.mkdir(path.dirname(appJsonPath), { recursive: true });
    try {
      const existingAppJson = await fs.promises.readFile(appJsonPath, "utf8");
      const existingAppJsonObject = JSON.parse(existingAppJson);
      Object.assign(appJsonObject, existingAppJsonObject, {
        appName,
        permissions: appJsonObject.permissions,
        ...options.appType != null ? { appType: options.appType } : {}
      });
    } catch {
    }
    await fs.promises.writeFile(appJsonPath, JSON.stringify(appJsonObject, null, 2), "utf8");
  };
  return {
    name: "apps-in-toss:app-json",
    dev: {
      order: "pre",
      handler: async ({ appName }) => {
        handleAppJson(appName);
      }
    },
    build: {
      order: "pre",
      handler: async ({ appName }) => {
        handleAppJson(appName);
      }
    }
  };
}

// src/plugins/artifact.ts
init_esm_shims();
import path4 from "path";
import { isBuildFailure, isBuildSuccess } from "@granite-js/plugin-core";
import { getPackageRoot as getPackageRoot2 } from "@granite-js/utils";
import pc2 from "picocolors";

// src/utils/collectDependencyVersions.ts
init_esm_shims();
import * as fs2 from "fs/promises";
import * as path2 from "path";
import * as esbuild from "esbuild";
async function collectDependencyVersions(rootDir) {
  const packageJsonPath = path2.join(rootDir, "package.json");
  const packageJson = JSON.parse(await fs2.readFile(packageJsonPath, "utf8"));
  const [dependencies, devDependencies] = await Promise.all([
    resolvePackageVersions(rootDir, Object.keys(packageJson.dependencies)),
    resolvePackageVersions(rootDir, Object.keys(packageJson.devDependencies))
  ]);
  return { dependencies, devDependencies };
}
async function resolvePackageVersions(rootDir, packageNames) {
  const results = {};
  await esbuild.build({
    stdin: { contents: createVirtualEntry(packageNames) },
    bundle: true,
    write: false,
    logLevel: "silent",
    plugins: [
      {
        name: "collect-package-version",
        setup(build2) {
          const RESOLVING = Symbol();
          build2.onResolve({ filter: /.*/ }, async (args) => {
            if (args.pluginData === RESOLVING) {
              return null;
            }
            const resolveOptions = {
              importer: args.importer,
              kind: args.kind,
              resolveDir: rootDir,
              pluginData: RESOLVING
            };
            let result = await build2.resolve(path2.join(args.path, "package.json"), resolveOptions);
            if (result.errors.length) {
              result = await build2.resolve(args.path, resolveOptions);
            }
            if (result.errors.length) {
              return result;
            }
            const packageName = args.path;
            const packagePath = extractPackagePath(result.path, packageName);
            if (packagePath) {
              results[packageName] = await getPackageVersion(packagePath);
            }
            return result;
          });
          build2.onLoad({ filter: /.*/ }, () => ({ contents: "// empty source" }));
        }
      }
    ]
  });
  return results;
}
function createVirtualEntry(packageNames) {
  return packageNames.map((packageName) => `import '${packageName}';`).join("\n");
}
function extractPackagePath(path9, packageName) {
  const normalizedPath = normalizePath(path9);
  if (normalizedPath.endsWith("/package.json")) {
    return normalizedPath.replace(/\/package\.json$/, "");
  }
  const match = normalizedPath.match(new RegExp(`(.*?node_modules/${packageName})/.*$`));
  if (match) {
    return match[1];
  }
  throw new Error(`Failed to extract path: ${packageName}`);
}
function normalizePath(path9) {
  return path9.replace(/\\/g, "/");
}
async function getPackageVersion(packagePath) {
  const packageJson = JSON.parse(await fs2.readFile(path2.join(packagePath, "package.json"), "utf-8"));
  return packageJson.version;
}

// src/utils/createArtifact.ts
init_esm_shims();
import * as fs3 from "fs/promises";
import path3 from "path";
import { AppsInTossBundle, PlatformType } from "@apps-in-toss/ait-format";

// src/utils/getUnityMetaDataEnv.ts
init_esm_shims();
function getUnityMetaDataEnv() {
  const metaData = process.env.UNITY_METADATA;
  if (!metaData) return {};
  try {
    return JSON.parse(metaData);
  } catch (e) {
    console.warn("UNITY_METADATA is not a valid JSON:", e);
    return {};
  }
}

// src/utils/createArtifact.ts
async function createArtifact(options) {
  const { bundleFiles, outfile, appJsonPath, reactNativeVersion, sdkVersion, isGame, isWeb } = options;
  const unityMetaData = getUnityMetaDataEnv();
  const appJsonContent = await fs3.readFile(appJsonPath, "utf-8");
  const appJson = JSON.parse(appJsonContent);
  const namedBundleFiles = bundleFiles.map((bundle) => {
    const originalName = path3.basename(bundle.path);
    const extension = path3.extname(originalName);
    const runtime = (bundle.runtimeVersion ?? reactNativeVersion).replace(new RegExp(/\./g), "_");
    return {
      path: bundle.path,
      name: `bundle.${bundle.platform}.${runtime}${extension === ".map" ? ".js.map" : extension}`
    };
  });
  const writer = AppsInTossBundle.writer({
    deploymentId: options.deploymentId,
    appName: appJson.appName,
    createdBy: "@apps-in-toss/plugins"
  });
  writer.setMetadata({
    platform: isWeb ? PlatformType.WEB : PlatformType.REACT_NATIVE,
    runtimeVersion: reactNativeVersion,
    isGame,
    sdkVersion,
    bundleFiles: namedBundleFiles.map(({ name }) => name),
    packageJson: options.packageJson,
    extra: { unityMetaData, ...options.extra }
  });
  for (const permission of appJson.permissions ?? []) {
    writer.addPermission(permission.name, permission.access);
  }
  for (const bundle of namedBundleFiles) {
    const content = await fs3.readFile(bundle.path);
    writer.addFile(bundle.name, new Uint8Array(content));
  }
  for (const file of options.additionalFilesToZip ?? []) {
    const content = await fs3.readFile(file.path);
    writer.addFile(file.name, new Uint8Array(content));
  }
  const buffer = await writer.toBuffer();
  await fs3.writeFile(outfile, buffer);
  return outfile;
}

// src/plugins/artifact.ts
function appsInTossCreateArtifact(deploymentId, buildOption, reactNativeVersion = REACT_NATIVE_VERSION) {
  const packageRoot = getPackageRoot2();
  return {
    name: "apps-in-toss:create-artifact",
    build: {
      order: "post",
      handler: async ({ buildResults, appName, cwd }) => {
        const buildFailed = buildResults.some(isBuildFailure);
        if (buildFailed) {
          throw new Error("\uBC88\uB4E4 \uBE4C\uB4DC \uC2E4\uD328");
        }
        log("\uC571 \uBE4C\uB4DC \uC911...");
        const { dependencies, devDependencies } = await collectDependencyVersions(cwd);
        const artifactOutfile = await createArtifact({
          reactNativeVersion,
          deploymentId,
          sdkVersion: buildOption.sdkVersion,
          isGame: buildOption.isGame,
          packageJson: { dependencies, devDependencies },
          bundleFiles: buildResults.filter(isBuildSuccess).map(({ outfile, sourcemapOutfile, platform }) => [
            {
              path: outfile,
              platform
            },
            {
              path: sourcemapOutfile,
              platform
            }
          ]).flat(),
          outfile: path4.join(cwd, `${appName}.ait`),
          appJsonPath: path4.join(packageRoot, ".granite", APP_MANIFEST_NAME),
          isWeb: buildOption.isWeb
        });
        if (!artifactOutfile) {
          throw new Error("\uC544\uD2F0\uD329\uD2B8 \uC0DD\uC131\uC5D0 \uC2E4\uD328\uD588\uC5B4\uC694.");
        }
        const filename = path4.basename(artifactOutfile);
        log(`\u2705 ${pc2.green(`${filename}`)} \uBE4C\uB4DC \uC644\uB8CC`);
      }
    }
  };
}

// src/plugins/build.ts
init_esm_shims();
function appsInTossEsbuildConfig(envScript) {
  return {
    name: "apps-in-toss:esbuild-config",
    config: {
      esbuild: {
        banner: {
          js: envScript
        }
      }
    }
  };
}
function appsInTossMetroConfig(envScriptPath) {
  return {
    name: "apps-in-toss:metro-config",
    config: {
      metro: {
        serializer: {
          getPolyfills: () => [envScriptPath]
        }
      }
    }
  };
}

// src/plugins/compat.ts
init_esm_shims();
import fs4 from "fs";
import path5 from "path";
import { getLocalTempDirectoryPath, getPackageRoot as getPackageRoot3 } from "@granite-js/utils";
var nativeModuleProxyContent = `
(function () {
  global.nativeModuleProxy = new Proxy(global.nativeModuleProxy, {
    get: function (target, name) {
      if (name === 'GraniteModule') {
        return target['BedrockModule'] || target[name];
      }

      if (name === 'GraniteCoreModule') {
        return target['BedrockCoreModule'] || target[name];
      }

      return target[name];
    }
  });

  global.__nativeModuleProxyConfigured = true;
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
var reactNativeModuleProxyContent = `
(function (global) {
  if (global.__nativeModuleProxyConfigured) {
    return;
  }

  function getCustomTurboModuleRegistry(registry) {
    var remappedModules = {
      'GraniteModule': 'BedrockModule',
      'GraniteCoreModule': 'BedrockCoreModule',
    };

    return {
      get: function (name) {
        var mod;
        var remappedName = remappedModules[name];

        if (remappedName) {
          mod = registry.get(remappedName);
        }

        return mod || registry.get(name);
      },
      getEnforcing: function (name) {
        var mod;
        var remappedName = remappedModules[name];

        if (remappedName) {
          mod = registry.get(remappedName);
        }

        return mod || registry.getEnforcing(name);
      }
    };
  }

  function createReactNativeProxy(reactNative) {
    return new Proxy(reactNative, {
      get: function (target, name) {
        var origin = target[name];
        return name === 'TurboModuleRegistry' ? getCustomTurboModuleRegistry(origin) : origin;
      }
    });
  }

  var reactNative;

  if (typeof global.__MICRO_FRONTEND__ !== 'undefined') {
    var mod = global.__MICRO_FRONTEND__.__SHARED__['react-native'];
    reactNative = mod && mod.get();
  }

  if (reactNative == null && typeof __bedrock_require__ === 'function') {
    reactNative = global.__bedrock_require__('react-native');
  }

  if (reactNative == null) {
    throw new Error('cannot get react-native in the global registry');
  }

  global.__reactNativeProxy = createReactNativeProxy(reactNative);
  global.__MICRO_FRONTEND__.__SHARED__['react-native'] = {
    get: function () {
      return global.__reactNativeProxy;
    },
    loaded: true,
  };
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
function bedrockCompat({ isHost }) {
  const metroPolyfillContent = path5.join(getLocalTempDirectoryPath(getPackageRoot3()), "metro-native-module-proxy.js");
  fs4.writeFileSync(metroPolyfillContent, nativeModuleProxyContent);
  return {
    name: "bedrock-compat-plugin",
    config: {
      esbuild: {
        banner: {
          js: isHost ? nativeModuleProxyContent : reactNativeModuleProxyContent
        }
      },
      metro: {
        serializer: {
          getPolyfills: () => [metroPolyfillContent]
        }
      }
    }
  };
}

// src/plugins/crypto.ts
init_esm_shims();
import fs5 from "fs";
import path6 from "path";
import { getLocalTempDirectoryPath as getLocalTempDirectoryPath2, getPackageRoot as getPackageRoot4 } from "@granite-js/utils";
var cryptoPolyfillContent = `
(function (global) {
  if (!global.__decrypt &&
      global.nativeModuleProxy &&
      global.nativeModuleProxy.TossCoreCryptoModule &&
      global.nativeModuleProxy.TossCoreCryptoModule.decrypt) {
    global.__decrypt = global.nativeModuleProxy.TossCoreCryptoModule.decrypt;
  }
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
function appsInTossCrypto({ isHost }) {
  const metroPolyfillPath = path6.join(getLocalTempDirectoryPath2(getPackageRoot4()), "metro-crypto-polyfill.js");
  fs5.writeFileSync(metroPolyfillPath, cryptoPolyfillContent);
  return {
    name: "apps-in-toss-crypto-plugin",
    config: {
      ...isHost ? {
        esbuild: {
          banner: {
            js: cryptoPolyfillContent
          }
        }
      } : {},
      metro: {
        serializer: {
          // Metro 단일 번들 환경에서는 Host/Remote 구분 없이 항상 polyfill 필요
          getPolyfills: () => [metroPolyfillPath]
        }
      }
    }
  };
}

// src/plugins/devServer.ts
init_esm_shims();

// src/utils/createServerPermissionsMiddleware.ts
init_esm_shims();
function createServerPermissionsMiddleware(permissions) {
  const parsedPermissions = parsePermissions(permissions);
  return (req, res, next) => {
    if (req.method === "GET" && req.url === "/permissions") {
      res.writeHead(200, {
        "Content-Length": Buffer.byteLength(parsedPermissions),
        "Content-Type": "text/plain"
      }).end(parsedPermissions);
      return;
    }
    next();
  };
}
function parsePermissions(permissions) {
  const permissionMap = /* @__PURE__ */ new Map();
  for (const permission of permissions) {
    const { name, access } = permission;
    if (!permissionMap.has(name)) {
      permissionMap.set(name, /* @__PURE__ */ new Set());
    }
    let accessValue = null;
    switch (access) {
      case "read":
        accessValue = "r";
        break;
      case "write":
        accessValue = "w";
        break;
      case "access":
        accessValue = "a";
        break;
      default:
        accessValue = null;
    }
    if (!accessValue) {
      continue;
    }
    permissionMap.get(name).add(accessValue);
  }
  return Array.from(permissionMap.entries()).map(([name, accesses]) => `${name}:${Array.from(accesses).sort().join("")}`).sort().join(",");
}

// src/plugins/devServer.ts
function appsInTossDevServer(options) {
  return {
    name: "apps-in-toss:dev-server",
    config: {
      metro: {
        middlewares: [createServerPermissionsMiddleware(options.permissions)]
      }
    }
  };
}

// src/plugins/notice.ts
init_esm_shims();
function appsInTossPostNotice() {
  const phase = "post";
  return {
    name: "apps-in-toss:post-notice",
    dev: {
      order: phase,
      handler() {
      }
    },
    build: {
      order: phase,
      async handler({ appName }) {
        if (this.meta.sentry) {
          const command = [
            "ait sentry upload-sourcemap",
            `--app-name ${appName}`,
            "--api-key <API_KEY>",
            "--deployment-id <DEPLOYMENT_ID>"
          ].join(" \\\n");
          log(
            `Sentry \uAD6C\uC131\uC774 \uAC10\uC9C0\uB410\uC5B4\uC694. \uC815\uD655\uD55C \uC624\uB958 \uCD94\uC801\uC744 \uC704\uD574\uC11C\uB294 \uBC30\uD3EC\uAC00 \uB05D\uB09C \uB4A4 \uB2E4\uC74C \uBA85\uB839\uC5B4\uB85C \uC18C\uC2A4\uB9F5\uC744 \uC5C5\uB85C\uB4DC\uD574 \uC8FC\uC138\uC694.

${command}
`
          );
        }
      }
    }
  };
}

// src/plugins/resolve.ts
init_esm_shims();
import Module from "module";
function getRequire() {
  return typeof __require === "function" ? __require : Module.createRequire(import.meta.url);
}
function requireMicroFrontendRuntime() {
  const require2 = getRequire();
  const runtimePath = require2.resolve("@granite-js/plugin-micro-frontend/runtime");
  return {
    name: "apps-in-toss-micro-frontend",
    config: {
      resolver: {
        alias: [
          {
            from: "@granite-js/plugin-micro-frontend/runtime",
            to: runtimePath
          }
        ]
      }
    }
  };
}

// src/types.ts
init_esm_shims();
var __typia_transform__isFormatUuid = __toESM(require_isFormatUuid(), 1);
var __typia_transform__validateReport = __toESM(require_validateReport(), 1);
var __typia_transform__createStandardSchema = __toESM(require_createStandardSchema(), 1);
var validateAppManifest = (() => {
  const _io0 = (input) => "string" === typeof input.appName && (void 0 === input.appType || "general" === input.appType || "game" === input.appType) && (Array.isArray(input.permissions) && input.permissions.every((elem) => "object" === typeof elem && null !== elem && _iu0(elem))) && ("object" === typeof input._metadata && null !== input._metadata && _io7(input._metadata));
  const _io1 = (input) => "clipboard" === input.name && ("read" === input.access || "write" === input.access);
  const _io2 = (input) => "geolocation" === input.name && "access" === input.access;
  const _io3 = (input) => "contacts" === input.name && ("read" === input.access || "write" === input.access);
  const _io4 = (input) => "photos" === input.name && ("read" === input.access || "write" === input.access);
  const _io5 = (input) => "camera" === input.name && "access" === input.access;
  const _io6 = (input) => "microphone" === input.name && "access" === input.access;
  const _io7 = (input) => Array.isArray(input.bundleFiles) && input.bundleFiles.every((elem) => "string" === typeof elem) && ("string" === typeof input.deploymentId && __typia_transform__isFormatUuid._isFormatUuid(input.deploymentId));
  const _iu0 = (input) => (() => {
    if ("microphone" === input.name)
      return _io6(input);
    else if ("camera" === input.name)
      return _io5(input);
    else if ("photos" === input.name)
      return _io4(input);
    else if ("contacts" === input.name)
      return _io3(input);
    else if ("geolocation" === input.name)
      return _io2(input);
    else if ("clipboard" === input.name)
      return _io1(input);
    else
      return false;
  })();
  const _vo0 = (input, _path, _exceptionable = true) => ["string" === typeof input.appName || _report(_exceptionable, {
    path: _path + ".appName",
    expected: "string",
    value: input.appName
  }), void 0 === input.appType || "general" === input.appType || "game" === input.appType || _report(_exceptionable, {
    path: _path + ".appType",
    expected: '("game" | "general" | undefined)',
    value: input.appType
  }), (Array.isArray(input.permissions) || _report(_exceptionable, {
    path: _path + ".permissions",
    expected: "Array<Permission>",
    value: input.permissions
  })) && input.permissions.map((elem, _index3) => ("object" === typeof elem && null !== elem || _report(_exceptionable, {
    path: _path + ".permissions[" + _index3 + "]",
    expected: "(CameraPermission | ClipboardPermission | ContactsPermission | GeolocationPermission | MicrophonePermission | PhotosPermission)",
    value: elem
  })) && _vu0(elem, _path + ".permissions[" + _index3 + "]", _exceptionable) || _report(_exceptionable, {
    path: _path + ".permissions[" + _index3 + "]",
    expected: "(CameraPermission | ClipboardPermission | ContactsPermission | GeolocationPermission | MicrophonePermission | PhotosPermission)",
    value: elem
  })).every((flag) => flag) || _report(_exceptionable, {
    path: _path + ".permissions",
    expected: "Array<Permission>",
    value: input.permissions
  }), ("object" === typeof input._metadata && null !== input._metadata || _report(_exceptionable, {
    path: _path + "._metadata",
    expected: "__type",
    value: input._metadata
  })) && _vo7(input._metadata, _path + "._metadata", _exceptionable) || _report(_exceptionable, {
    path: _path + "._metadata",
    expected: "__type",
    value: input._metadata
  })].every((flag) => flag);
  const _vo1 = (input, _path, _exceptionable = true) => ["clipboard" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"clipboard"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo2 = (input, _path, _exceptionable = true) => ["geolocation" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"geolocation"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo3 = (input, _path, _exceptionable = true) => ["contacts" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"contacts"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo4 = (input, _path, _exceptionable = true) => ["photos" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"photos"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo5 = (input, _path, _exceptionable = true) => ["camera" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"camera"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo6 = (input, _path, _exceptionable = true) => ["microphone" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"microphone"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo7 = (input, _path, _exceptionable = true) => [(Array.isArray(input.bundleFiles) || _report(_exceptionable, {
    path: _path + ".bundleFiles",
    expected: "Array<string>",
    value: input.bundleFiles
  })) && input.bundleFiles.map((elem, _index4) => "string" === typeof elem || _report(_exceptionable, {
    path: _path + ".bundleFiles[" + _index4 + "]",
    expected: "string",
    value: elem
  })).every((flag) => flag) || _report(_exceptionable, {
    path: _path + ".bundleFiles",
    expected: "Array<string>",
    value: input.bundleFiles
  }), "string" === typeof input.deploymentId && (__typia_transform__isFormatUuid._isFormatUuid(input.deploymentId) || _report(_exceptionable, {
    path: _path + ".deploymentId",
    expected: 'string & Format<"uuid">',
    value: input.deploymentId
  })) || _report(_exceptionable, {
    path: _path + ".deploymentId",
    expected: '(string & Format<"uuid">)',
    value: input.deploymentId
  })].every((flag) => flag);
  const _vu0 = (input, _path, _exceptionable = true) => (() => {
    if ("microphone" === input.name)
      return _vo6(input, _path, _exceptionable);
    else if ("camera" === input.name)
      return _vo5(input, _path, _exceptionable);
    else if ("photos" === input.name)
      return _vo4(input, _path, _exceptionable);
    else if ("contacts" === input.name)
      return _vo3(input, _path, _exceptionable);
    else if ("geolocation" === input.name)
      return _vo2(input, _path, _exceptionable);
    else if ("clipboard" === input.name)
      return _vo1(input, _path, _exceptionable);
    else
      return _report(_exceptionable, {
        path: _path,
        expected: "(MicrophonePermission | CameraPermission | PhotosPermission | ContactsPermission | GeolocationPermission | ClipboardPermission)",
        value: input
      });
  })();
  const __is = (input) => "object" === typeof input && null !== input && _io0(input);
  let errors;
  let _report;
  return __typia_transform__createStandardSchema._createStandardSchema((input) => {
    if (false === __is(input)) {
      errors = [];
      _report = __typia_transform__validateReport._validateReport(errors);
      ((input2, _path, _exceptionable = true) => ("object" === typeof input2 && null !== input2 || _report(true, {
        path: _path + "",
        expected: "AppManifest",
        value: input2
      })) && _vo0(input2, _path + "", true) || _report(true, {
        path: _path + "",
        expected: "AppManifest",
        value: input2
      }))(input, "$input", true);
      const success = 0 === errors.length;
      return success ? {
        success,
        data: input
      } : {
        success,
        errors,
        data: input
      };
    }
    return {
      success: true,
      data: input
    };
  });
})();
var validateAppInTossPluginOptions = (() => {
  const _io0 = (input) => (void 0 === input.target || "0.72.6" === input.target || "0.84.0" === input.target) && (void 0 === input.appType || "general" === input.appType || "game" === input.appType) && ("object" === typeof input.brand && null !== input.brand && _io1(input.brand)) && (Array.isArray(input.permissions) && input.permissions.every((elem) => "object" === typeof elem && null !== elem && _iu0(elem))) && (void 0 === input.navigationBar || "object" === typeof input.navigationBar && null !== input.navigationBar && false === Array.isArray(input.navigationBar) && _io8(input.navigationBar));
  const _io1 = (input) => "string" === typeof input.displayName && "string" === typeof input.primaryColor && "string" === typeof input.icon;
  const _io2 = (input) => "clipboard" === input.name && ("read" === input.access || "write" === input.access);
  const _io3 = (input) => "geolocation" === input.name && "access" === input.access;
  const _io4 = (input) => "contacts" === input.name && ("read" === input.access || "write" === input.access);
  const _io5 = (input) => "photos" === input.name && ("read" === input.access || "write" === input.access);
  const _io6 = (input) => "camera" === input.name && "access" === input.access;
  const _io7 = (input) => "microphone" === input.name && "access" === input.access;
  const _io8 = (input) => (void 0 === input.withBackButton || "boolean" === typeof input.withBackButton) && (void 0 === input.withHomeButton || "boolean" === typeof input.withHomeButton) && (void 0 === input.initialAccessoryButton || "object" === typeof input.initialAccessoryButton && null !== input.initialAccessoryButton && _io9(input.initialAccessoryButton));
  const _io9 = (input) => "string" === typeof input.id && "string" === typeof input.title && ("object" === typeof input.icon && null !== input.icon && _iu1(input.icon));
  const _io10 = (input) => "object" === typeof input.source && null !== input.source && _io11(input.source) && (null !== input.name && void 0 === input.name);
  const _io11 = (input) => "string" === typeof input.uri;
  const _io12 = (input) => "string" === typeof input.name && (null !== input.source && void 0 === input.source);
  const _iu0 = (input) => (() => {
    if ("microphone" === input.name)
      return _io7(input);
    else if ("camera" === input.name)
      return _io6(input);
    else if ("photos" === input.name)
      return _io5(input);
    else if ("contacts" === input.name)
      return _io4(input);
    else if ("geolocation" === input.name)
      return _io3(input);
    else if ("clipboard" === input.name)
      return _io2(input);
    else
      return false;
  })();
  const _iu1 = (input) => (() => {
    if ("string" === typeof input.name)
      return _io12(input);
    else if ("object" === typeof input.source && null !== input.source && _io11(input.source))
      return _io10(input);
    else
      return false;
  })();
  const _vo0 = (input, _path, _exceptionable = true) => [void 0 === input.target || "0.72.6" === input.target || "0.84.0" === input.target || _report(_exceptionable, {
    path: _path + ".target",
    expected: '("0.72.6" | "0.84.0" | undefined)',
    value: input.target
  }), void 0 === input.appType || "general" === input.appType || "game" === input.appType || _report(_exceptionable, {
    path: _path + ".appType",
    expected: '("game" | "general" | undefined)',
    value: input.appType
  }), ("object" === typeof input.brand && null !== input.brand || _report(_exceptionable, {
    path: _path + ".brand",
    expected: "__type",
    value: input.brand
  })) && _vo1(input.brand, _path + ".brand", _exceptionable) || _report(_exceptionable, {
    path: _path + ".brand",
    expected: "__type",
    value: input.brand
  }), (Array.isArray(input.permissions) || _report(_exceptionable, {
    path: _path + ".permissions",
    expected: "Array<Permission>",
    value: input.permissions
  })) && input.permissions.map((elem, _index2) => ("object" === typeof elem && null !== elem || _report(_exceptionable, {
    path: _path + ".permissions[" + _index2 + "]",
    expected: "(CameraPermission | ClipboardPermission | ContactsPermission | GeolocationPermission | MicrophonePermission | PhotosPermission)",
    value: elem
  })) && _vu0(elem, _path + ".permissions[" + _index2 + "]", _exceptionable) || _report(_exceptionable, {
    path: _path + ".permissions[" + _index2 + "]",
    expected: "(CameraPermission | ClipboardPermission | ContactsPermission | GeolocationPermission | MicrophonePermission | PhotosPermission)",
    value: elem
  })).every((flag) => flag) || _report(_exceptionable, {
    path: _path + ".permissions",
    expected: "Array<Permission>",
    value: input.permissions
  }), void 0 === input.navigationBar || ("object" === typeof input.navigationBar && null !== input.navigationBar && false === Array.isArray(input.navigationBar) || _report(_exceptionable, {
    path: _path + ".navigationBar",
    expected: "(__type.o1 | undefined)",
    value: input.navigationBar
  })) && _vo8(input.navigationBar, _path + ".navigationBar", _exceptionable) || _report(_exceptionable, {
    path: _path + ".navigationBar",
    expected: "(__type.o1 | undefined)",
    value: input.navigationBar
  })].every((flag) => flag);
  const _vo1 = (input, _path, _exceptionable = true) => ["string" === typeof input.displayName || _report(_exceptionable, {
    path: _path + ".displayName",
    expected: "string",
    value: input.displayName
  }), "string" === typeof input.primaryColor || _report(_exceptionable, {
    path: _path + ".primaryColor",
    expected: "string",
    value: input.primaryColor
  }), "string" === typeof input.icon || _report(_exceptionable, {
    path: _path + ".icon",
    expected: "string",
    value: input.icon
  })].every((flag) => flag);
  const _vo2 = (input, _path, _exceptionable = true) => ["clipboard" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"clipboard"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo3 = (input, _path, _exceptionable = true) => ["geolocation" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"geolocation"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo4 = (input, _path, _exceptionable = true) => ["contacts" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"contacts"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo5 = (input, _path, _exceptionable = true) => ["photos" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"photos"',
    value: input.name
  }), "read" === input.access || "write" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '("read" | "write")',
    value: input.access
  })].every((flag) => flag);
  const _vo6 = (input, _path, _exceptionable = true) => ["camera" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"camera"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo7 = (input, _path, _exceptionable = true) => ["microphone" === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: '"microphone"',
    value: input.name
  }), "access" === input.access || _report(_exceptionable, {
    path: _path + ".access",
    expected: '"access"',
    value: input.access
  })].every((flag) => flag);
  const _vo8 = (input, _path, _exceptionable = true) => [void 0 === input.withBackButton || "boolean" === typeof input.withBackButton || _report(_exceptionable, {
    path: _path + ".withBackButton",
    expected: "(boolean | undefined)",
    value: input.withBackButton
  }), void 0 === input.withHomeButton || "boolean" === typeof input.withHomeButton || _report(_exceptionable, {
    path: _path + ".withHomeButton",
    expected: "(boolean | undefined)",
    value: input.withHomeButton
  }), void 0 === input.initialAccessoryButton || ("object" === typeof input.initialAccessoryButton && null !== input.initialAccessoryButton || _report(_exceptionable, {
    path: _path + ".initialAccessoryButton",
    expected: "(InitialAccessoryButton | undefined)",
    value: input.initialAccessoryButton
  })) && _vo9(input.initialAccessoryButton, _path + ".initialAccessoryButton", _exceptionable) || _report(_exceptionable, {
    path: _path + ".initialAccessoryButton",
    expected: "(InitialAccessoryButton | undefined)",
    value: input.initialAccessoryButton
  })].every((flag) => flag);
  const _vo9 = (input, _path, _exceptionable = true) => ["string" === typeof input.id || _report(_exceptionable, {
    path: _path + ".id",
    expected: "string",
    value: input.id
  }), "string" === typeof input.title || _report(_exceptionable, {
    path: _path + ".title",
    expected: "string",
    value: input.title
  }), ("object" === typeof input.icon && null !== input.icon || _report(_exceptionable, {
    path: _path + ".icon",
    expected: "(__type.o2 | __type.o4)",
    value: input.icon
  })) && _vu1(input.icon, _path + ".icon", _exceptionable) || _report(_exceptionable, {
    path: _path + ".icon",
    expected: "(__type.o2 | __type.o4)",
    value: input.icon
  })].every((flag) => flag);
  const _vo10 = (input, _path, _exceptionable = true) => [("object" === typeof input.source && null !== input.source || _report(_exceptionable, {
    path: _path + ".source",
    expected: "__type.o3",
    value: input.source
  })) && _vo11(input.source, _path + ".source", _exceptionable) || _report(_exceptionable, {
    path: _path + ".source",
    expected: "__type.o3",
    value: input.source
  }), (null !== input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: "undefined",
    value: input.name
  })) && (void 0 === input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: "undefined",
    value: input.name
  }))].every((flag) => flag);
  const _vo11 = (input, _path, _exceptionable = true) => ["string" === typeof input.uri || _report(_exceptionable, {
    path: _path + ".uri",
    expected: "string",
    value: input.uri
  })].every((flag) => flag);
  const _vo12 = (input, _path, _exceptionable = true) => ["string" === typeof input.name || _report(_exceptionable, {
    path: _path + ".name",
    expected: "string",
    value: input.name
  }), (null !== input.source || _report(_exceptionable, {
    path: _path + ".source",
    expected: "undefined",
    value: input.source
  })) && (void 0 === input.source || _report(_exceptionable, {
    path: _path + ".source",
    expected: "undefined",
    value: input.source
  }))].every((flag) => flag);
  const _vu0 = (input, _path, _exceptionable = true) => (() => {
    if ("microphone" === input.name)
      return _vo7(input, _path, _exceptionable);
    else if ("camera" === input.name)
      return _vo6(input, _path, _exceptionable);
    else if ("photos" === input.name)
      return _vo5(input, _path, _exceptionable);
    else if ("contacts" === input.name)
      return _vo4(input, _path, _exceptionable);
    else if ("geolocation" === input.name)
      return _vo3(input, _path, _exceptionable);
    else if ("clipboard" === input.name)
      return _vo2(input, _path, _exceptionable);
    else
      return _report(_exceptionable, {
        path: _path,
        expected: "(MicrophonePermission | CameraPermission | PhotosPermission | ContactsPermission | GeolocationPermission | ClipboardPermission)",
        value: input
      });
  })();
  const _vu1 = (input, _path, _exceptionable = true) => (() => {
    if ("string" === typeof input.name)
      return _vo12(input, _path, _exceptionable);
    else if ("object" === typeof input.source && null !== input.source && _vo11(input.source, _path + ".source", false))
      return _vo10(input, _path, _exceptionable);
    else
      return _report(_exceptionable, {
        path: _path,
        expected: "(__type.o4 | __type.o2)",
        value: input
      });
  })();
  const __is = (input) => "object" === typeof input && null !== input && _io0(input);
  let errors;
  let _report;
  return __typia_transform__createStandardSchema._createStandardSchema((input) => {
    if (false === __is(input)) {
      errors = [];
      _report = __typia_transform__validateReport._validateReport(errors);
      ((input2, _path, _exceptionable = true) => ("object" === typeof input2 && null !== input2 || _report(true, {
        path: _path + "",
        expected: "AppsInTossPluginOptions",
        value: input2
      })) && _vo0(input2, _path + "", true) || _report(true, {
        path: _path + "",
        expected: "AppsInTossPluginOptions",
        value: input2
      }))(input, "$input", true);
      const success = 0 === errors.length;
      return success ? {
        success,
        data: input
      } : {
        success,
        errors,
        data: input
      };
    }
    return {
      success: true,
      data: input
    };
  });
})();

// src/utils/generateDeploymentId.ts
init_esm_shims();
import { uuidv7 } from "uuidv7";
function generateDeploymentId() {
  if (process.env.AIT_DEPLOYMENT_ID != null && process.env.AIT_DEPLOYMENT_ID.length > 0) {
    return process.env.AIT_DEPLOYMENT_ID;
  }
  return uuidv7();
}

// src/utils/getReactNativeTarget.ts
init_esm_shims();
function getReactNativeTarget(target) {
  return target ?? REACT_NATIVE_VERSION;
}

// src/utils/setupRuntimeSetupScript.ts
init_esm_shims();
import fs6 from "node:fs";
import path7 from "node:path";
import { getPackageRoot as getPackageRoot5 } from "@granite-js/utils";
import { transformSync } from "esbuild";
function setupRuntimeSetupScript(config) {
  const script = getRuntimeSetupScript(
    {
      appType: config.appType,
      deploymentId: config.deploymentId,
      brandDisplayName: config.brand.displayName,
      brandPrimaryColor: config.brand.primaryColor,
      brandIcon: config.brand.icon,
      navigationBar: config.navigationBar ?? {}
    },
    SERVICE_CONTEXT_IDENTIFIER
  );
  const setupScriptPath = writeRuntimeSetupScript(script);
  return { contents: script, path: setupScriptPath };
}
function setupHostRuntimeSetupScript(metadata) {
  const script = getRuntimeSetupScript(metadata, HOST_CONTEXT_IDENTIFIER);
  const setupScriptPath = writeRuntimeSetupScript(script);
  return { contents: script, path: setupScriptPath };
}
function getRuntimeSetupScript(metadata, identifier) {
  const script = `
  (function (global) {
    ${getBedrockCompatScript()}
    ${identifier === HOST_CONTEXT_IDENTIFIER ? "" : getMicroFrontendCompatScript()}
    global.${identifier} = ${JSON.stringify(metadata)};
  })(
    typeof globalThis !== 'undefined'
      ? globalThis
      : typeof global !== 'undefined'
      ? global
      : typeof window !== 'undefined'
      ? window
      : this
  );
  `;
  return transformSync(script, {
    minify: true,
    sourcemap: false,
    banner: "// Runtime setup script"
  }).code;
}
function writeRuntimeSetupScript(script) {
  const packageRoot = getPackageRoot5();
  const granitePath = path7.join(packageRoot, ".granite");
  try {
    fs6.accessSync(granitePath);
  } catch {
    fs6.mkdirSync(granitePath, { recursive: true });
  }
  const envFilePath = path7.join(granitePath, ".apps-in-toss.env.js");
  fs6.writeFileSync(envFilePath, script, "utf-8");
  return envFilePath;
}
function getBedrockCompatScript() {
  return `
  if (global.__bedrock_require__ == null) {
    global.__bedrock_require__ = function (id) {
      return global.__MICRO_FRONTEND__.__SHARED__[id].get();
    };
  }
  `;
}
function getMicroFrontendCompatScript() {
  return `
  function getHostContainer() {
    var microFrontendContext = global.__MICRO_FRONTEND__;
    var microFrontendInstances = microFrontendContext == null ? [] : microFrontendContext.__INSTANCES__;
    var sharedContainer = microFrontendInstances.find(function (instance) {
      return instance.name === '${MICRO_FRONTEND_HOST_NAME}';
    });
    return sharedContainer;
  }

  if (getHostContainer() == null) {
    var hostContainerShim = { name: '${MICRO_FRONTEND_HOST_NAME}', config: {} };
    var sharedModuleProxy = new Proxy({}, {
      get(target, prop) {
        var mod;

        try {
          mod = target[prop] && target[prop].get();
        } catch {}

        try {
          mod = mod || global.__bedrock_require__(prop);
        } catch {}

        if (mod == null) {
          return;
        }

        return {
          get: function () {
            return mod;
          },
          loaded: true,
        };
      },
      set(target, prop, value) {
        target[prop] = value;
        return true;
      }
    });

    global.__MICRO_FRONTEND__ = {
      __INSTANCES__: [hostContainerShim],
      __SHARED__: sharedModuleProxy,
    };
    global.__MICRO_FRONTEND__.__INSTANCES__['${MICRO_FRONTEND_HOST_NAME}'] = 0;
  }
  `;
}

// src/appsInToss.ts
function getAppsInTossBuildPlugins(options) {
  const plugins = [appsInTossDevServer(options), appsInTossPostNotice()];
  return plugins;
}
function withAppsInTossCommon(plugins, options) {
  const { contents, path: envScriptPath } = setupRuntimeSetupScript({
    appType: options.appType,
    brand: options.brand,
    deploymentId: options.deploymentId,
    navigationBar: options.navigationBar
  });
  const target = getReactNativeTarget(options.target);
  const reactNativeVersion = target.startsWith("0.72") ? "0.72" : "0.84";
  return [
    requireMicroFrontendRuntime(),
    microFrontend({
      name: MICRO_FRONTEND_REMOTE_NAME,
      exposes: {
        [MICRO_FRONTEND_REMOTE_EXPOSE_PATH]: options.exposePath
      },
      shared: getSharedPackages({
        reactNativeVersion
      }),
      reactNativeBasePath: options.reactNativeBasePath
    }),
    appsInTossAppJson(options),
    reactCompatibilityPlugin({ target, rootDir: options.root }),
    reactNativeCompatibilityPlugin({ target, rootDir: options.root }),
    reactNativeReverseCompatibilityPlugin({ target, rootDir: options.root, isHost: false }),
    appsInTossEsbuildConfig(contents),
    appsInTossMetroConfig(envScriptPath),
    bedrockCompat({ isHost: false }),
    reactNativeVersion === "0.84" ? appsInTossCrypto({ isHost: false }) : null,
    ...plugins
  ].filter((plugin) => plugin !== null);
}
function appsInToss(options) {
  const result = validateAppInTossPluginOptions(options);
  if (!result.success) {
    error("[Apps In Toss Plugin] \uD50C\uB7EC\uADF8\uC778 \uC635\uC158\uC774 \uC62C\uBC14\uB974\uC9C0 \uC54A\uC2B5\uB2C8\uB2E4. granite.config.ts \uAD6C\uC131\uC744 \uD655\uC778\uD574\uC8FC\uC138\uC694.");
    process.exit(1);
  }
  const deploymentId = generateDeploymentId();
  const root = getPackageRoot6();
  return withAppsInTossCommon(getAppsInTossBuildPlugins(options), {
    ...options,
    root,
    deploymentId,
    exposePath: path8.join(root, "./src/_app.tsx"),
    navigationBar: result.data.navigationBar
  });
}

// src/analytics.ts
init_esm_shims();
var INTERNAL_MEMBER_IDENT = "_reactInternalFiber";
var REACT_NATIVE_RENDERER_DEV_MODULE = {
  modulePath: "react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js",
  transform: (code) => {
    return code.replace(
      "this._internalFiberInstanceHandleDEV = internalInstanceHandleDEV;",
      `this._internalFiberInstanceHandleDEV = this.${INTERNAL_MEMBER_IDENT} = internalInstanceHandleDEV;`
    );
  }
};
var REACT_NATIVE_RENDERER_PROD_MODULE = {
  modulePath: "react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js",
  transform: (code) => {
    return code.replace(
      `    function ReactNativeFiberHostComponent(tag, viewConfig) {
      this._nativeTag = tag;
      this._children = [];
      this.viewConfig = viewConfig;
    }`,
      `    function ReactNativeFiberHostComponent(tag, viewConfig, fiber) {
      this._nativeTag = tag;
      this._children = [];
      this.viewConfig = viewConfig;
      this.${INTERNAL_MEMBER_IDENT} = fiber;
    }`
    );
  }
};
function analytics() {
  let transformed = false;
  return {
    name: "apps-in-toss:analytics",
    build: {
      order: "pre",
      handler: () => {
        transformed = false;
      }
    },
    config: {
      transformer: {
        transformSync: (id, code) => {
          if (transformed) {
            return code;
          }
          if (id.endsWith(REACT_NATIVE_RENDERER_DEV_MODULE.modulePath)) {
            transformed = true;
            return REACT_NATIVE_RENDERER_DEV_MODULE.transform(code);
          }
          if (id.endsWith(REACT_NATIVE_RENDERER_PROD_MODULE.modulePath)) {
            transformed = true;
            return REACT_NATIVE_RENDERER_PROD_MODULE.transform(code);
          }
          return code;
        }
      }
    }
  };
}

// src/utils/getSdkVersion.ts
init_esm_shims();
function getSdkVersion() {
  const packageJson = require_package();
  return packageJson.version;
}
export {
  analytics,
  appsInToss,
  appsInTossAppJson,
  appsInTossCreateArtifact,
  appsInTossCrypto,
  appsInTossDevServer,
  appsInTossEsbuildConfig,
  appsInTossMetroConfig,
  appsInTossPostNotice,
  bedrockCompat,
  collectDependencyVersions,
  createArtifact,
  cryptoPolyfillContent,
  getAppsInTossBuildPlugins,
  getRuntimeSetupScript,
  getSdkVersion,
  nativeModuleProxyContent,
  reactNativeModuleProxyContent,
  requireMicroFrontendRuntime,
  setupHostRuntimeSetupScript,
  setupRuntimeSetupScript,
  validateAppInTossPluginOptions,
  validateAppManifest,
  withAppsInTossCommon
};
