// @apps-in-toss/plugins/dist/internal.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/internal.ts
var internal_exports = {};
__export(internal_exports, {
  appsInTossHost: () => appsInTossHost,
  getAppsInTossHostBuildPlugins: () => getAppsInTossHostBuildPlugins
});
module.exports = __toCommonJS(internal_exports);

// ../../../.yarn/__virtual__/tsup-virtual-5946fc3428/0/cache/tsup-npm-8.4.0-f78d2622c9-c6636ffd6a.zip/node_modules/tsup/assets/cjs_shims.js
var getImportMetaUrl = () => typeof document === "undefined" ? new URL(`file:${__filename}`).href : document.currentScript && document.currentScript.src || new URL("main.js", document.baseURI).href;
var importMetaUrl = /* @__PURE__ */ getImportMetaUrl();

// src/appsInTossHost.ts
var import_plugin_compat = require("@apps-in-toss/plugin-compat");
var import_plugin_micro_frontend = require("@granite-js/plugin-micro-frontend");
var import_plugin_sentry = require("@granite-js/plugin-sentry");
var import_utils6 = require("@granite-js/utils");

// src/constants.ts
var REACT_NATIVE_VERSION = "0.84.0";
var APP_MANIFEST_NAME = "app.json";
var MICRO_FRONTEND_HOST_NAME = "apps-in-toss-host";
var HOST_CONTEXT_IDENTIFIER = "__appsInTossHost";
var getSharedPackages = ({ reactNativeVersion }) => [
  "@react-native-community/blur",
  "@react-navigation/native",
  "@react-navigation/native-stack",
  "@sentry/react-native",
  "@shopify/flash-list",
  "react-native-safe-area-context",
  "react-native-screens",
  "react-native-svg",
  "react-native-gesture-handler",
  "react-native",
  "react",
  "react-native-webview",
  "@granite-js/image",
  "@granite-js/lottie",
  "@granite-js/video",
  ...reactNativeVersion === "0.84" ? ["brick-module", "@react-navigation/elements"] : [],
  ...reactNativeVersion === "0.72" ? ["react-native-fast-image", "lottie-react-native", "react-native-video"] : []
];

// src/plugins/appJson.ts
var import_node_fs = __toESM(require("fs"), 1);
var import_path = __toESM(require("path"), 1);
var import_utils = require("@granite-js/utils");
async function appsInTossAppJson(options) {
  const packageRoot = (0, import_utils.getPackageRoot)();
  const handleAppJson = async (appName) => {
    const appJsonPath = import_path.default.join(packageRoot, ".granite", APP_MANIFEST_NAME);
    const appJsonObject = {
      appName,
      permissions: options.permissions,
      ...options.appType != null ? { appType: options.appType } : {}
    };
    await import_node_fs.default.promises.mkdir(import_path.default.dirname(appJsonPath), { recursive: true });
    try {
      const existingAppJson = await import_node_fs.default.promises.readFile(appJsonPath, "utf8");
      const existingAppJsonObject = JSON.parse(existingAppJson);
      Object.assign(appJsonObject, existingAppJsonObject, {
        appName,
        permissions: appJsonObject.permissions,
        ...options.appType != null ? { appType: options.appType } : {}
      });
    } catch {
    }
    await import_node_fs.default.promises.writeFile(appJsonPath, JSON.stringify(appJsonObject, null, 2), "utf8");
  };
  return {
    name: "apps-in-toss:app-json",
    dev: {
      order: "pre",
      handler: async ({ appName }) => {
        handleAppJson(appName);
      }
    },
    build: {
      order: "pre",
      handler: async ({ appName }) => {
        handleAppJson(appName);
      }
    }
  };
}

// src/plugins/artifact.ts
var import_plugin_core = require("@granite-js/plugin-core");
var import_utils2 = require("@granite-js/utils");
var import_picocolors2 = __toESM(require("picocolors"), 1);

// src/log.ts
var import_picocolors = __toESM(require("picocolors"), 1);
var TAG = import_picocolors.default.bold(import_picocolors.default.cyan("[AppsInToss]"));

// src/utils/collectDependencyVersions.ts
var esbuild = __toESM(require("esbuild"), 1);

// src/utils/createArtifact.ts
var import_ait_format = require("@apps-in-toss/ait-format");

// src/plugins/build.ts
function appsInTossEsbuildConfig(envScript) {
  return {
    name: "apps-in-toss:esbuild-config",
    config: {
      esbuild: {
        banner: {
          js: envScript
        }
      }
    }
  };
}
function appsInTossMetroConfig(envScriptPath) {
  return {
    name: "apps-in-toss:metro-config",
    config: {
      metro: {
        serializer: {
          getPolyfills: () => [envScriptPath]
        }
      }
    }
  };
}

// src/plugins/compat.ts
var import_fs = __toESM(require("fs"), 1);
var import_path2 = __toESM(require("path"), 1);
var import_utils3 = require("@granite-js/utils");
var nativeModuleProxyContent = `
(function () {
  global.nativeModuleProxy = new Proxy(global.nativeModuleProxy, {
    get: function (target, name) {
      if (name === 'GraniteModule') {
        return target['BedrockModule'] || target[name];
      }

      if (name === 'GraniteCoreModule') {
        return target['BedrockCoreModule'] || target[name];
      }

      return target[name];
    }
  });

  global.__nativeModuleProxyConfigured = true;
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
var reactNativeModuleProxyContent = `
(function (global) {
  if (global.__nativeModuleProxyConfigured) {
    return;
  }

  function getCustomTurboModuleRegistry(registry) {
    var remappedModules = {
      'GraniteModule': 'BedrockModule',
      'GraniteCoreModule': 'BedrockCoreModule',
    };

    return {
      get: function (name) {
        var mod;
        var remappedName = remappedModules[name];

        if (remappedName) {
          mod = registry.get(remappedName);
        }

        return mod || registry.get(name);
      },
      getEnforcing: function (name) {
        var mod;
        var remappedName = remappedModules[name];

        if (remappedName) {
          mod = registry.get(remappedName);
        }

        return mod || registry.getEnforcing(name);
      }
    };
  }

  function createReactNativeProxy(reactNative) {
    return new Proxy(reactNative, {
      get: function (target, name) {
        var origin = target[name];
        return name === 'TurboModuleRegistry' ? getCustomTurboModuleRegistry(origin) : origin;
      }
    });
  }

  var reactNative;

  if (typeof global.__MICRO_FRONTEND__ !== 'undefined') {
    var mod = global.__MICRO_FRONTEND__.__SHARED__['react-native'];
    reactNative = mod && mod.get();
  }

  if (reactNative == null && typeof __bedrock_require__ === 'function') {
    reactNative = global.__bedrock_require__('react-native');
  }

  if (reactNative == null) {
    throw new Error('cannot get react-native in the global registry');
  }

  global.__reactNativeProxy = createReactNativeProxy(reactNative);
  global.__MICRO_FRONTEND__.__SHARED__['react-native'] = {
    get: function () {
      return global.__reactNativeProxy;
    },
    loaded: true,
  };
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
function bedrockCompat({ isHost }) {
  const metroPolyfillContent = import_path2.default.join((0, import_utils3.getLocalTempDirectoryPath)((0, import_utils3.getPackageRoot)()), "metro-native-module-proxy.js");
  import_fs.default.writeFileSync(metroPolyfillContent, nativeModuleProxyContent);
  return {
    name: "bedrock-compat-plugin",
    config: {
      esbuild: {
        banner: {
          js: isHost ? nativeModuleProxyContent : reactNativeModuleProxyContent
        }
      },
      metro: {
        serializer: {
          getPolyfills: () => [metroPolyfillContent]
        }
      }
    }
  };
}

// src/plugins/crypto.ts
var import_fs2 = __toESM(require("fs"), 1);
var import_path3 = __toESM(require("path"), 1);
var import_utils4 = require("@granite-js/utils");
var cryptoPolyfillContent = `
(function (global) {
  if (!global.__decrypt &&
      global.nativeModuleProxy &&
      global.nativeModuleProxy.TossCoreCryptoModule &&
      global.nativeModuleProxy.TossCoreCryptoModule.decrypt) {
    global.__decrypt = global.nativeModuleProxy.TossCoreCryptoModule.decrypt;
  }
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
function appsInTossCrypto({ isHost }) {
  const metroPolyfillPath = import_path3.default.join((0, import_utils4.getLocalTempDirectoryPath)((0, import_utils4.getPackageRoot)()), "metro-crypto-polyfill.js");
  import_fs2.default.writeFileSync(metroPolyfillPath, cryptoPolyfillContent);
  return {
    name: "apps-in-toss-crypto-plugin",
    config: {
      ...isHost ? {
        esbuild: {
          banner: {
            js: cryptoPolyfillContent
          }
        }
      } : {},
      metro: {
        serializer: {
          // Metro 단일 번들 환경에서는 Host/Remote 구분 없이 항상 polyfill 필요
          getPolyfills: () => [metroPolyfillPath]
        }
      }
    }
  };
}

// src/utils/createServerPermissionsMiddleware.ts
function createServerPermissionsMiddleware(permissions) {
  const parsedPermissions = parsePermissions(permissions);
  return (req, res, next) => {
    if (req.method === "GET" && req.url === "/permissions") {
      res.writeHead(200, {
        "Content-Length": Buffer.byteLength(parsedPermissions),
        "Content-Type": "text/plain"
      }).end(parsedPermissions);
      return;
    }
    next();
  };
}
function parsePermissions(permissions) {
  const permissionMap = /* @__PURE__ */ new Map();
  for (const permission of permissions) {
    const { name, access } = permission;
    if (!permissionMap.has(name)) {
      permissionMap.set(name, /* @__PURE__ */ new Set());
    }
    let accessValue = null;
    switch (access) {
      case "read":
        accessValue = "r";
        break;
      case "write":
        accessValue = "w";
        break;
      case "access":
        accessValue = "a";
        break;
      default:
        accessValue = null;
    }
    if (!accessValue) {
      continue;
    }
    permissionMap.get(name).add(accessValue);
  }
  return Array.from(permissionMap.entries()).map(([name, accesses]) => `${name}:${Array.from(accesses).sort().join("")}`).sort().join(",");
}

// src/plugins/devServer.ts
function appsInTossDevServer(options) {
  return {
    name: "apps-in-toss:dev-server",
    config: {
      metro: {
        middlewares: [createServerPermissionsMiddleware(options.permissions)]
      }
    }
  };
}

// src/plugins/resolve.ts
var import_module = __toESM(require("module"), 1);
function getRequire() {
  return typeof require === "function" ? require : import_module.default.createRequire(importMetaUrl);
}
function requireMicroFrontendRuntime() {
  const require2 = getRequire();
  const runtimePath = require2.resolve("@granite-js/plugin-micro-frontend/runtime");
  return {
    name: "apps-in-toss-micro-frontend",
    config: {
      resolver: {
        alias: [
          {
            from: "@granite-js/plugin-micro-frontend/runtime",
            to: runtimePath
          }
        ]
      }
    }
  };
}

// src/utils/generateDeploymentId.ts
var import_uuidv7 = require("uuidv7");
function generateDeploymentId() {
  if (process.env.AIT_DEPLOYMENT_ID != null && process.env.AIT_DEPLOYMENT_ID.length > 0) {
    return process.env.AIT_DEPLOYMENT_ID;
  }
  return (0, import_uuidv7.uuidv7)();
}

// src/utils/getReactNativeTarget.ts
function getReactNativeTarget(target) {
  return target ?? REACT_NATIVE_VERSION;
}

// src/utils/setupRuntimeSetupScript.ts
var import_node_fs2 = __toESM(require("fs"), 1);
var import_node_path = __toESM(require("path"), 1);
var import_utils5 = require("@granite-js/utils");
var import_esbuild = require("esbuild");
function setupHostRuntimeSetupScript(metadata) {
  const script = getRuntimeSetupScript(metadata, HOST_CONTEXT_IDENTIFIER);
  const setupScriptPath = writeRuntimeSetupScript(script);
  return { contents: script, path: setupScriptPath };
}
function getRuntimeSetupScript(metadata, identifier) {
  const script = `
  (function (global) {
    ${getBedrockCompatScript()}
    ${identifier === HOST_CONTEXT_IDENTIFIER ? "" : getMicroFrontendCompatScript()}
    global.${identifier} = ${JSON.stringify(metadata)};
  })(
    typeof globalThis !== 'undefined'
      ? globalThis
      : typeof global !== 'undefined'
      ? global
      : typeof window !== 'undefined'
      ? window
      : this
  );
  `;
  return (0, import_esbuild.transformSync)(script, {
    minify: true,
    sourcemap: false,
    banner: "// Runtime setup script"
  }).code;
}
function writeRuntimeSetupScript(script) {
  const packageRoot = (0, import_utils5.getPackageRoot)();
  const granitePath = import_node_path.default.join(packageRoot, ".granite");
  try {
    import_node_fs2.default.accessSync(granitePath);
  } catch {
    import_node_fs2.default.mkdirSync(granitePath, { recursive: true });
  }
  const envFilePath = import_node_path.default.join(granitePath, ".apps-in-toss.env.js");
  import_node_fs2.default.writeFileSync(envFilePath, script, "utf-8");
  return envFilePath;
}
function getBedrockCompatScript() {
  return `
  if (global.__bedrock_require__ == null) {
    global.__bedrock_require__ = function (id) {
      return global.__MICRO_FRONTEND__.__SHARED__[id].get();
    };
  }
  `;
}
function getMicroFrontendCompatScript() {
  return `
  function getHostContainer() {
    var microFrontendContext = global.__MICRO_FRONTEND__;
    var microFrontendInstances = microFrontendContext == null ? [] : microFrontendContext.__INSTANCES__;
    var sharedContainer = microFrontendInstances.find(function (instance) {
      return instance.name === '${MICRO_FRONTEND_HOST_NAME}';
    });
    return sharedContainer;
  }

  if (getHostContainer() == null) {
    var hostContainerShim = { name: '${MICRO_FRONTEND_HOST_NAME}', config: {} };
    var sharedModuleProxy = new Proxy({}, {
      get(target, prop) {
        var mod;

        try {
          mod = target[prop] && target[prop].get();
        } catch {}

        try {
          mod = mod || global.__bedrock_require__(prop);
        } catch {}

        if (mod == null) {
          return;
        }

        return {
          get: function () {
            return mod;
          },
          loaded: true,
        };
      },
      set(target, prop, value) {
        target[prop] = value;
        return true;
      }
    });

    global.__MICRO_FRONTEND__ = {
      __INSTANCES__: [hostContainerShim],
      __SHARED__: sharedModuleProxy,
    };
    global.__MICRO_FRONTEND__.__INSTANCES__['${MICRO_FRONTEND_HOST_NAME}'] = 0;
  }
  `;
}

// src/appsInTossHost.ts
var PERMISSIONS = [];
function getAppsInTossHostBuildPlugins() {
  const plugins = [appsInTossDevServer({ permissions: PERMISSIONS })];
  return plugins;
}
function appsInTossHost(options) {
  const deploymentId = generateDeploymentId();
  const reactNativeTarget = getReactNativeTarget(options.target);
  const reactNativeVersion = reactNativeTarget.startsWith("0.72") ? "0.72" : "0.84";
  const { contents, path: envFilePath } = setupHostRuntimeSetupScript({
    deploymentId,
    environment: getHostEnvironment()
  });
  return [
    requireMicroFrontendRuntime(),
    (0, import_plugin_micro_frontend.microFrontend)({
      name: MICRO_FRONTEND_HOST_NAME,
      remote: {
        host: options?.remote?.host ?? "localhost",
        port: options?.remote?.port ?? 8082
      },
      shared: getSharedPackages({
        reactNativeVersion
      }).reduce((prev, curr) => ({ ...prev, [curr]: { eager: true } }), {})
    }),
    (0, import_plugin_compat.reactNativeCompatibilityPlugin)({ target: reactNativeTarget, rootDir: (0, import_utils6.getPackageRoot)() }),
    (0, import_plugin_compat.reactNativeReverseCompatibilityPlugin)({ target: reactNativeTarget, rootDir: (0, import_utils6.getPackageRoot)(), isHost: true }),
    appsInTossAppJson({ permissions: PERMISSIONS }),
    ...getAppsInTossHostBuildPlugins(),
    appsInTossEsbuildConfig(contents),
    appsInTossMetroConfig(envFilePath),
    bedrockCompat({ isHost: true }),
    reactNativeVersion === "0.84" ? appsInTossCrypto({ isHost: true }) : null,
    (0, import_plugin_sentry.sentry)({ useClient: false })
  ].filter((plugin) => plugin !== null);
}
function getHostEnvironment() {
  const environment = process.env.HOST_ENV;
  switch (environment) {
    case "alpha":
    case "live":
      return environment;
    default:
      return "live";
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  appsInTossHost,
  getAppsInTossHostBuildPlugins
});
