// @apps-in-toss/plugins/dist/internal.js
import { createRequire } from 'module'; const require = createRequire(import.meta.url);
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/appsInTossHost.ts
import { reactNativeCompatibilityPlugin, reactNativeReverseCompatibilityPlugin } from "@apps-in-toss/plugin-compat";
import { microFrontend } from "@granite-js/plugin-micro-frontend";
import { sentry } from "@granite-js/plugin-sentry";
import { getPackageRoot as getPackageRoot6 } from "@granite-js/utils";

// src/constants.ts
var REACT_NATIVE_VERSION = "0.84.0";
var APP_MANIFEST_NAME = "app.json";
var MICRO_FRONTEND_HOST_NAME = "apps-in-toss-host";
var HOST_CONTEXT_IDENTIFIER = "__appsInTossHost";
var getSharedPackages = ({ reactNativeVersion }) => [
  "@react-native-community/blur",
  "@react-navigation/native",
  "@react-navigation/native-stack",
  "@sentry/react-native",
  "@shopify/flash-list",
  "react-native-safe-area-context",
  "react-native-screens",
  "react-native-svg",
  "react-native-gesture-handler",
  "react-native",
  "react",
  "react-native-webview",
  "@granite-js/image",
  "@granite-js/lottie",
  "@granite-js/video",
  ...reactNativeVersion === "0.84" ? ["brick-module", "@react-navigation/elements"] : [],
  ...reactNativeVersion === "0.72" ? ["react-native-fast-image", "lottie-react-native", "react-native-video"] : []
];

// src/plugins/appJson.ts
import fs from "node:fs";
import path from "path";
import { getPackageRoot } from "@granite-js/utils";
async function appsInTossAppJson(options) {
  const packageRoot = getPackageRoot();
  const handleAppJson = async (appName) => {
    const appJsonPath = path.join(packageRoot, ".granite", APP_MANIFEST_NAME);
    const appJsonObject = {
      appName,
      permissions: options.permissions,
      ...options.appType != null ? { appType: options.appType } : {}
    };
    await fs.promises.mkdir(path.dirname(appJsonPath), { recursive: true });
    try {
      const existingAppJson = await fs.promises.readFile(appJsonPath, "utf8");
      const existingAppJsonObject = JSON.parse(existingAppJson);
      Object.assign(appJsonObject, existingAppJsonObject, {
        appName,
        permissions: appJsonObject.permissions,
        ...options.appType != null ? { appType: options.appType } : {}
      });
    } catch {
    }
    await fs.promises.writeFile(appJsonPath, JSON.stringify(appJsonObject, null, 2), "utf8");
  };
  return {
    name: "apps-in-toss:app-json",
    dev: {
      order: "pre",
      handler: async ({ appName }) => {
        handleAppJson(appName);
      }
    },
    build: {
      order: "pre",
      handler: async ({ appName }) => {
        handleAppJson(appName);
      }
    }
  };
}

// src/plugins/artifact.ts
import { isBuildFailure, isBuildSuccess } from "@granite-js/plugin-core";
import { getPackageRoot as getPackageRoot2 } from "@granite-js/utils";
import pc2 from "picocolors";

// src/log.ts
import pc from "picocolors";
var TAG = pc.bold(pc.cyan("[AppsInToss]"));

// src/utils/collectDependencyVersions.ts
import * as esbuild from "esbuild";

// src/utils/createArtifact.ts
import { AppsInTossBundle, PlatformType } from "@apps-in-toss/ait-format";

// src/plugins/build.ts
function appsInTossEsbuildConfig(envScript) {
  return {
    name: "apps-in-toss:esbuild-config",
    config: {
      esbuild: {
        banner: {
          js: envScript
        }
      }
    }
  };
}
function appsInTossMetroConfig(envScriptPath) {
  return {
    name: "apps-in-toss:metro-config",
    config: {
      metro: {
        serializer: {
          getPolyfills: () => [envScriptPath]
        }
      }
    }
  };
}

// src/plugins/compat.ts
import fs2 from "fs";
import path2 from "path";
import { getLocalTempDirectoryPath, getPackageRoot as getPackageRoot3 } from "@granite-js/utils";
var nativeModuleProxyContent = `
(function () {
  global.nativeModuleProxy = new Proxy(global.nativeModuleProxy, {
    get: function (target, name) {
      if (name === 'GraniteModule') {
        return target['BedrockModule'] || target[name];
      }

      if (name === 'GraniteCoreModule') {
        return target['BedrockCoreModule'] || target[name];
      }

      return target[name];
    }
  });

  global.__nativeModuleProxyConfigured = true;
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
var reactNativeModuleProxyContent = `
(function (global) {
  if (global.__nativeModuleProxyConfigured) {
    return;
  }

  function getCustomTurboModuleRegistry(registry) {
    var remappedModules = {
      'GraniteModule': 'BedrockModule',
      'GraniteCoreModule': 'BedrockCoreModule',
    };

    return {
      get: function (name) {
        var mod;
        var remappedName = remappedModules[name];

        if (remappedName) {
          mod = registry.get(remappedName);
        }

        return mod || registry.get(name);
      },
      getEnforcing: function (name) {
        var mod;
        var remappedName = remappedModules[name];

        if (remappedName) {
          mod = registry.get(remappedName);
        }

        return mod || registry.getEnforcing(name);
      }
    };
  }

  function createReactNativeProxy(reactNative) {
    return new Proxy(reactNative, {
      get: function (target, name) {
        var origin = target[name];
        return name === 'TurboModuleRegistry' ? getCustomTurboModuleRegistry(origin) : origin;
      }
    });
  }

  var reactNative;

  if (typeof global.__MICRO_FRONTEND__ !== 'undefined') {
    var mod = global.__MICRO_FRONTEND__.__SHARED__['react-native'];
    reactNative = mod && mod.get();
  }

  if (reactNative == null && typeof __bedrock_require__ === 'function') {
    reactNative = global.__bedrock_require__('react-native');
  }

  if (reactNative == null) {
    throw new Error('cannot get react-native in the global registry');
  }

  global.__reactNativeProxy = createReactNativeProxy(reactNative);
  global.__MICRO_FRONTEND__.__SHARED__['react-native'] = {
    get: function () {
      return global.__reactNativeProxy;
    },
    loaded: true,
  };
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
function bedrockCompat({ isHost }) {
  const metroPolyfillContent = path2.join(getLocalTempDirectoryPath(getPackageRoot3()), "metro-native-module-proxy.js");
  fs2.writeFileSync(metroPolyfillContent, nativeModuleProxyContent);
  return {
    name: "bedrock-compat-plugin",
    config: {
      esbuild: {
        banner: {
          js: isHost ? nativeModuleProxyContent : reactNativeModuleProxyContent
        }
      },
      metro: {
        serializer: {
          getPolyfills: () => [metroPolyfillContent]
        }
      }
    }
  };
}

// src/plugins/crypto.ts
import fs3 from "fs";
import path3 from "path";
import { getLocalTempDirectoryPath as getLocalTempDirectoryPath2, getPackageRoot as getPackageRoot4 } from "@granite-js/utils";
var cryptoPolyfillContent = `
(function (global) {
  if (!global.__decrypt &&
      global.nativeModuleProxy &&
      global.nativeModuleProxy.TossCoreCryptoModule &&
      global.nativeModuleProxy.TossCoreCryptoModule.decrypt) {
    global.__decrypt = global.nativeModuleProxy.TossCoreCryptoModule.decrypt;
  }
})(
  typeof globalThis !== 'undefined'
    ? globalThis
    : typeof global !== 'undefined'
    ? global
    : typeof window !== 'undefined'
    ? window
    : this
);
`;
function appsInTossCrypto({ isHost }) {
  const metroPolyfillPath = path3.join(getLocalTempDirectoryPath2(getPackageRoot4()), "metro-crypto-polyfill.js");
  fs3.writeFileSync(metroPolyfillPath, cryptoPolyfillContent);
  return {
    name: "apps-in-toss-crypto-plugin",
    config: {
      ...isHost ? {
        esbuild: {
          banner: {
            js: cryptoPolyfillContent
          }
        }
      } : {},
      metro: {
        serializer: {
          // Metro 단일 번들 환경에서는 Host/Remote 구분 없이 항상 polyfill 필요
          getPolyfills: () => [metroPolyfillPath]
        }
      }
    }
  };
}

// src/utils/createServerPermissionsMiddleware.ts
function createServerPermissionsMiddleware(permissions) {
  const parsedPermissions = parsePermissions(permissions);
  return (req, res, next) => {
    if (req.method === "GET" && req.url === "/permissions") {
      res.writeHead(200, {
        "Content-Length": Buffer.byteLength(parsedPermissions),
        "Content-Type": "text/plain"
      }).end(parsedPermissions);
      return;
    }
    next();
  };
}
function parsePermissions(permissions) {
  const permissionMap = /* @__PURE__ */ new Map();
  for (const permission of permissions) {
    const { name, access } = permission;
    if (!permissionMap.has(name)) {
      permissionMap.set(name, /* @__PURE__ */ new Set());
    }
    let accessValue = null;
    switch (access) {
      case "read":
        accessValue = "r";
        break;
      case "write":
        accessValue = "w";
        break;
      case "access":
        accessValue = "a";
        break;
      default:
        accessValue = null;
    }
    if (!accessValue) {
      continue;
    }
    permissionMap.get(name).add(accessValue);
  }
  return Array.from(permissionMap.entries()).map(([name, accesses]) => `${name}:${Array.from(accesses).sort().join("")}`).sort().join(",");
}

// src/plugins/devServer.ts
function appsInTossDevServer(options) {
  return {
    name: "apps-in-toss:dev-server",
    config: {
      metro: {
        middlewares: [createServerPermissionsMiddleware(options.permissions)]
      }
    }
  };
}

// src/plugins/resolve.ts
import Module from "module";
function getRequire() {
  return typeof __require === "function" ? __require : Module.createRequire(import.meta.url);
}
function requireMicroFrontendRuntime() {
  const require2 = getRequire();
  const runtimePath = require2.resolve("@granite-js/plugin-micro-frontend/runtime");
  return {
    name: "apps-in-toss-micro-frontend",
    config: {
      resolver: {
        alias: [
          {
            from: "@granite-js/plugin-micro-frontend/runtime",
            to: runtimePath
          }
        ]
      }
    }
  };
}

// src/utils/generateDeploymentId.ts
import { uuidv7 } from "uuidv7";
function generateDeploymentId() {
  if (process.env.AIT_DEPLOYMENT_ID != null && process.env.AIT_DEPLOYMENT_ID.length > 0) {
    return process.env.AIT_DEPLOYMENT_ID;
  }
  return uuidv7();
}

// src/utils/getReactNativeTarget.ts
function getReactNativeTarget(target) {
  return target ?? REACT_NATIVE_VERSION;
}

// src/utils/setupRuntimeSetupScript.ts
import fs4 from "node:fs";
import path4 from "node:path";
import { getPackageRoot as getPackageRoot5 } from "@granite-js/utils";
import { transformSync } from "esbuild";
function setupHostRuntimeSetupScript(metadata) {
  const script = getRuntimeSetupScript(metadata, HOST_CONTEXT_IDENTIFIER);
  const setupScriptPath = writeRuntimeSetupScript(script);
  return { contents: script, path: setupScriptPath };
}
function getRuntimeSetupScript(metadata, identifier) {
  const script = `
  (function (global) {
    ${getBedrockCompatScript()}
    ${identifier === HOST_CONTEXT_IDENTIFIER ? "" : getMicroFrontendCompatScript()}
    global.${identifier} = ${JSON.stringify(metadata)};
  })(
    typeof globalThis !== 'undefined'
      ? globalThis
      : typeof global !== 'undefined'
      ? global
      : typeof window !== 'undefined'
      ? window
      : this
  );
  `;
  return transformSync(script, {
    minify: true,
    sourcemap: false,
    banner: "// Runtime setup script"
  }).code;
}
function writeRuntimeSetupScript(script) {
  const packageRoot = getPackageRoot5();
  const granitePath = path4.join(packageRoot, ".granite");
  try {
    fs4.accessSync(granitePath);
  } catch {
    fs4.mkdirSync(granitePath, { recursive: true });
  }
  const envFilePath = path4.join(granitePath, ".apps-in-toss.env.js");
  fs4.writeFileSync(envFilePath, script, "utf-8");
  return envFilePath;
}
function getBedrockCompatScript() {
  return `
  if (global.__bedrock_require__ == null) {
    global.__bedrock_require__ = function (id) {
      return global.__MICRO_FRONTEND__.__SHARED__[id].get();
    };
  }
  `;
}
function getMicroFrontendCompatScript() {
  return `
  function getHostContainer() {
    var microFrontendContext = global.__MICRO_FRONTEND__;
    var microFrontendInstances = microFrontendContext == null ? [] : microFrontendContext.__INSTANCES__;
    var sharedContainer = microFrontendInstances.find(function (instance) {
      return instance.name === '${MICRO_FRONTEND_HOST_NAME}';
    });
    return sharedContainer;
  }

  if (getHostContainer() == null) {
    var hostContainerShim = { name: '${MICRO_FRONTEND_HOST_NAME}', config: {} };
    var sharedModuleProxy = new Proxy({}, {
      get(target, prop) {
        var mod;

        try {
          mod = target[prop] && target[prop].get();
        } catch {}

        try {
          mod = mod || global.__bedrock_require__(prop);
        } catch {}

        if (mod == null) {
          return;
        }

        return {
          get: function () {
            return mod;
          },
          loaded: true,
        };
      },
      set(target, prop, value) {
        target[prop] = value;
        return true;
      }
    });

    global.__MICRO_FRONTEND__ = {
      __INSTANCES__: [hostContainerShim],
      __SHARED__: sharedModuleProxy,
    };
    global.__MICRO_FRONTEND__.__INSTANCES__['${MICRO_FRONTEND_HOST_NAME}'] = 0;
  }
  `;
}

// src/appsInTossHost.ts
var PERMISSIONS = [];
function getAppsInTossHostBuildPlugins() {
  const plugins = [appsInTossDevServer({ permissions: PERMISSIONS })];
  return plugins;
}
function appsInTossHost(options) {
  const deploymentId = generateDeploymentId();
  const reactNativeTarget = getReactNativeTarget(options.target);
  const reactNativeVersion = reactNativeTarget.startsWith("0.72") ? "0.72" : "0.84";
  const { contents, path: envFilePath } = setupHostRuntimeSetupScript({
    deploymentId,
    environment: getHostEnvironment()
  });
  return [
    requireMicroFrontendRuntime(),
    microFrontend({
      name: MICRO_FRONTEND_HOST_NAME,
      remote: {
        host: options?.remote?.host ?? "localhost",
        port: options?.remote?.port ?? 8082
      },
      shared: getSharedPackages({
        reactNativeVersion
      }).reduce((prev, curr) => ({ ...prev, [curr]: { eager: true } }), {})
    }),
    reactNativeCompatibilityPlugin({ target: reactNativeTarget, rootDir: getPackageRoot6() }),
    reactNativeReverseCompatibilityPlugin({ target: reactNativeTarget, rootDir: getPackageRoot6(), isHost: true }),
    appsInTossAppJson({ permissions: PERMISSIONS }),
    ...getAppsInTossHostBuildPlugins(),
    appsInTossEsbuildConfig(contents),
    appsInTossMetroConfig(envFilePath),
    bedrockCompat({ isHost: true }),
    reactNativeVersion === "0.84" ? appsInTossCrypto({ isHost: true }) : null,
    sentry({ useClient: false })
  ].filter((plugin) => plugin !== null);
}
function getHostEnvironment() {
  const environment = process.env.HOST_ENV;
  switch (environment) {
    case "alpha":
    case "live":
      return environment;
    default:
      return "live";
  }
}
export {
  appsInTossHost,
  getAppsInTossHostBuildPlugins
};
