// @apps-in-toss/plugin-compat/dist/index.js
import { createRequire } from 'module'; const require = createRequire(import.meta.url);
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// ../../../.yarn/__virtual__/tsup-virtual-5946fc3428/0/cache/tsup-npm-8.4.0-f78d2622c9-c6636ffd6a.zip/node_modules/tsup/assets/esm_shims.js
import { fileURLToPath } from "url";
import path from "path";
var getFilename = () => fileURLToPath(import.meta.url);
var getDirname = () => path.dirname(getFilename());
var __dirname = /* @__PURE__ */ getDirname();

// src/constants.ts
import path2 from "path";
var DEPRECATED_NATIVE_PACKAGES = [
  "react-native-google-mobile-ads",
  "react-native-applovin-max",
  "react-native-mmkv"
];
var getTossReactNativePackages = ({ target }) => [
  "@react-navigation/native-stack",
  "@react-navigation/native",
  "@react-native-async-storage/async-storage",
  "@sentry/react-native",
  "@shopify/flash-list",
  "@granite-js/image",
  "@granite-js/lottie",
  "@granite-js/video",
  "react-native-safe-area-context",
  "react-native-screens",
  "react-native-svg",
  "react-native-webview",
  "react-native",
  "react",
  "react-native-gesture-handler",
  "react-native-pager-view",
  "@react-native-community/blur",
  "brick-module",
  ...target === "0.72.6" ? ["react-native-fast-image", "lottie-react-native", "react-native-video"] : [],
  ...DEPRECATED_NATIVE_PACKAGES
];
var DEFAULT_REACT_NATIVE_COMPATIBILITY = "0.84";
function getHermesOSBin() {
  switch (process.platform) {
    case "win32":
      return "win64-bin";
    case "darwin":
      return "osx-bin";
    default:
      return "linux64-bin";
  }
}
function getHermesOSExe() {
  const hermesExecutableName = "hermesc";
  return process.platform === "win32" ? `${hermesExecutableName}.exe` : hermesExecutableName;
}
var REACT_NATIVE_COMPATIBILITIES = {
  "0.84.0": {
    dependencyAliases: null,
    getReactNativeBasePath: null,
    getHermesBinaryPath: null
  },
  "0.72.6": {
    dependencyAliases: {
      react: "react-18.2.0",
      "react-dom": "react-dom-18.2.0",
      "@react-native/normalize-colors": "@react-native/normalize-colors-0.72.0",
      "@react-native-async-storage/async-storage": "@react-native-async-storage/async-storage-1.18.2",
      "@react-native-community/blur": "@react-native-community/blur-4.3.2",
      "@shopify/flash-list": "@shopify/flash-list-1.6.2",
      "lottie-react-native": "lottie-react-native-6.4.0",
      "@granite-js/lottie": "lottie-react-native-6.4.0",
      "react-native-fast-image": "react-native-fast-image-8.6.3",
      "@granite-js/image": "react-native-fast-image-8.6.3",
      "react-native": "react-native-0.72.6",
      "react-native-safe-area-context": "react-native-safe-area-context-4.7.4",
      "react-native-gesture-handler": "react-native-gesture-handler-2.8.0",
      "react-native-screens": "react-native-screens-3.27.0",
      "react-native-svg": "react-native-svg-13.14.0",
      "react-native-webview": "react-native-webview-13.6.3",
      "react-native-video": "react-native-video-6.0.0-alpha.6",
      "@granite-js/video": "react-native-video-6.0.0-alpha.6",
      "react-native-pager-view": "react-native-pager-view-6.1.2",
      "@react-navigation/elements": "@react-navigation/elements-1.3.9",
      "@react-navigation/native": "@react-navigation/native-6.0.13",
      "@react-navigation/native-stack": "@react-navigation/native-stack-6.9.0",
      "@react-native/js-polyfills": "@react-native/js-polyfills-0.72.1"
    },
    getReactNativeBasePath: () => {
      return path2.dirname(__require.resolve("react-native-0.72.6"));
    },
    getHermesBinaryPath: () => {
      const rnBasePath = path2.dirname(__require.resolve("react-native-0.72.6"));
      const hermesEngine = path2.join(rnBasePath, "sdks", "hermesc", getHermesOSBin(), getHermesOSExe());
      return hermesEngine;
    }
  }
};

// src/reactNativeCompatibilityPlugin.ts
import path5 from "path";
import { isNotNil } from "es-toolkit";

// src/resolver/customMetroResolver.ts
import Module from "module";
import path3 from "path";
import enhancedResolve from "enhanced-resolve";
var SINGLETON_MODULES = ["@babel/runtime"];
var NATIVE_MODULES = ["react-native", "react"];
var SUPPORTED_BUILTIN_FALLBACKS = {
  http: "stream-http",
  https: "https-browserify",
  zlib: "browserify-zlib",
  url: "url",
  util: "util",
  path: "path-browserify",
  buffer: "buffer",
  stream: "stream-browserify",
  assert: "assert",
  events: "events"
};
var builtinModules = new Set(Module.builtinModules);
var resolvers = /* @__PURE__ */ new Map();
function createCustomResolver(rootPath, options) {
  const resolvedAliasConfig = resolveAliasConfig(options?.alias ?? []);
  function createResolverImpl(context, platform, rootPath2) {
    const baseExtensions = context.sourceExts.map((extension) => `.${extension}`);
    let finalExtensions = [...baseExtensions];
    if (context.preferNativePlatform) {
      finalExtensions = [...baseExtensions.map((extension) => `.native${extension}`), ...finalExtensions];
    }
    if (platform) {
      finalExtensions = [...baseExtensions.map((extension) => `.${platform}${extension}`), ...finalExtensions];
    }
    const resolver = enhancedResolve.create.sync({
      extensions: finalExtensions,
      mainFields: context.mainFields,
      conditionNames: ["react-native", "require", "node", "default"],
      mainFiles: ["index"],
      modules: ["node_modules", path3.join(rootPath2, "src")]
    });
    function resolve(context2, request) {
      const matchedAlias = resolvedAliasConfig.find((alias) => alias.filter.test(request));
      if (matchedAlias != null) {
        const resolveResult = resolver(
          {},
          matchedAlias.resolveDir ?? path3.dirname(context2.originModulePath),
          matchedAlias.resolveRequest(request)
        );
        return {
          type: "sourceFile",
          filePath: resolveResult
        };
      }
      for (const nativeModule of NATIVE_MODULES) {
        if (request === nativeModule) {
          return {
            type: "sourceFile",
            filePath: resolver({}, rootPath2, request)
          };
        }
      }
      for (const singletonModule of SINGLETON_MODULES) {
        if (request.startsWith(singletonModule)) {
          return {
            type: "sourceFile",
            filePath: resolver({}, rootPath2, request)
          };
        }
      }
      if (builtinModules.has(request)) {
        if (request in SUPPORTED_BUILTIN_FALLBACKS) {
          const source = SUPPORTED_BUILTIN_FALLBACKS[request];
          return {
            type: "sourceFile",
            filePath: __require.resolve(`${source}/`)
          };
        }
        request = `${request}/`;
      }
      try {
        const resolveResult = resolver({}, path3.dirname(context2.originModulePath), request);
        return {
          type: "sourceFile",
          filePath: resolveResult
        };
      } catch (error) {
        const typedError = error;
        if ("code" in typedError && typedError.code !== "QUALIFIED_PATH_RESOLUTION_FAILED") {
          typedError.message = `"${context2.originModulePath}"\uC5D0\uC11C "${request}"\uB97C \uCC3E\uC744 \uC218 \uC5C6\uC5C8\uC2B5\uB2C8\uB2E4.
${typedError.message}`;
        }
        throw typedError;
      }
    }
    return resolve;
  }
  return function resolve(context, request, platform) {
    let resolver = resolvers.get(platform);
    if (resolver == null) {
      resolver = createResolverImpl(context, platform, rootPath);
      resolvers.set(platform, resolver);
    }
    return resolver(context, request, platform);
  };
}
function resolveAliasConfig(alias) {
  return alias.map((alias2) => {
    const escapedFrom = escapeRegExpString(alias2.from);
    const isResolvedPath = path3.isAbsolute(alias2.to);
    return {
      ...alias2,
      filter: new RegExp(alias2.exact ? `^${escapedFrom}$` : `^${escapedFrom}(?:$|/)`),
      resolveRequest: (request) => {
        return isResolvedPath ? alias2.to : replaceModulePath(request, alias2.from, alias2.to);
      }
    };
  });
}
function escapeRegExpString(str) {
  return str.replace(/[|\\{}()[\]^$+*?.]/g, "\\$&");
}
function replaceModulePath(path8, from, to) {
  const subpath = path8.slice(from.length);
  return `${to}${subpath}`;
}

// src/shims/async-storage-compat.ts
import fs from "fs";
import path4 from "path";
var ASYNC_STORAGE_PACKAGE_NAME = "@react-native-async-storage/async-storage";
var GRANITE_DIRECTORY_NAME = ".granite";
var ASYNC_STORAGE_COMPAT_FILE_NAME = ".ait-async-storage-compat.js";
var ASYNC_STORAGE_COMPAT_CONTENT = `
const __appsInTossAsyncStorageWarning = \`
[apps-in-toss] @react-native-async-storage/async-storage is not available in AppsInToss.
Use Storage from the AppsInToss SDK instead.

import { Storage } from '@apps-in-toss/framework';

const KEY = 'my-key';

async function handleSetStorageItem(value) {
  await Storage.setItem(KEY, value);
}

async function handleGetStorageItem() {
  return Storage.getItem(KEY);
}

async function handleRemoveStorageItem() {
  await Storage.removeItem(KEY);
}
\`.trim();

let __appsInTossDidWarn = false;

function warnOnce(methodName) {
  if (__appsInTossDidWarn) {
    return;
  }

  __appsInTossDidWarn = true;

  if (typeof console !== "undefined" && typeof console.warn === "function") {
    console.warn([__appsInTossAsyncStorageWarning, 'Called method: ' + methodName].join('\\n'));
  }
}

function invokeCallback(callback, value) {
  if (typeof callback === "function") {
    callback(null, value);
  }
}

export function getItem(_key, callback) {
  warnOnce('getItem');
  invokeCallback(callback, null);
  return Promise.resolve(null);
}

export function setItem(_key, _value, callback) {
  warnOnce('setItem');
  invokeCallback(callback);
  return Promise.resolve();
}

export function removeItem(_key, callback) {
  warnOnce('removeItem');
  invokeCallback(callback);
  return Promise.resolve();
}

export function mergeItem(_key, _value, callback) {
  warnOnce('mergeItem');
  invokeCallback(callback);
  return Promise.resolve();
}

export function clear(callback) {
  warnOnce('clear');
  invokeCallback(callback);
  return Promise.resolve();
}

export function getAllKeys(callback) {
  warnOnce('getAllKeys');
  const keys = [];
  invokeCallback(callback, keys);
  return Promise.resolve(keys);
}

export function multiGet(keys, callback) {
  warnOnce('multiGet');
  const values = keys.map((key) => [key, null]);
  invokeCallback(callback, values);
  return Promise.resolve(values);
}

export function multiSet(_keyValuePairs, callback) {
  warnOnce('multiSet');
  invokeCallback(callback);
  return Promise.resolve();
}

export function multiRemove(_keys, callback) {
  warnOnce('multiRemove');
  invokeCallback(callback);
  return Promise.resolve();
}

export function multiMerge(_keyValuePairs, callback) {
  warnOnce('multiMerge');
  invokeCallback(callback);
  return Promise.resolve();
}

export function flushGetRequests() {
  warnOnce('flushGetRequests');
}

export function useAsyncStorage(key) {
  return {
    getItem(callback) {
      return getItem(key, callback);
    },
    setItem(value, callback) {
      return setItem(key, value, callback);
    },
    mergeItem(value, callback) {
      return mergeItem(key, value, callback);
    },
    removeItem(callback) {
      return removeItem(key, callback);
    },
  };
}

const AsyncStorage = {
  getItem,
  setItem,
  removeItem,
  mergeItem,
  clear,
  getAllKeys,
  multiGet,
  multiSet,
  multiRemove,
  multiMerge,
  flushGetRequests,
  useAsyncStorage,
};

export default AsyncStorage;
`.trimStart();
function createAsyncStorageCompatContent() {
  return ASYNC_STORAGE_COMPAT_CONTENT;
}
function ensureAsyncStorageCompatFile(rootDir) {
  const graniteDirectoryPath = path4.join(rootDir, GRANITE_DIRECTORY_NAME);
  const asyncStorageCompatPath = path4.join(graniteDirectoryPath, ASYNC_STORAGE_COMPAT_FILE_NAME);
  try {
    fs.mkdirSync(graniteDirectoryPath, { recursive: true });
    fs.writeFileSync(asyncStorageCompatPath, createAsyncStorageCompatContent(), "utf8");
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error);
    throw new Error(
      [
        "[react-native-compatibility-plugin] Failed to create async-storage compatibility temp file.",
        `rootDir="${rootDir}"`,
        `asyncStorageCompatPath="${asyncStorageCompatPath}"`,
        `reason: ${reason}`
      ].join(" ")
    );
  }
  return { asyncStorageCompatPath };
}

// src/reactNativeCompatibilityPlugin.ts
function reactNativeCompatibilityPlugin(options) {
  const { target } = options;
  const compatibilityConfig = REACT_NATIVE_COMPATIBILITIES[target];
  const compatibilityAliases = compatibilityConfig?.dependencyAliases ?? {};
  const compatibilityReactNativePackage = compatibilityAliases["react-native"];
  const isReactNativeVersionOverridden = compatibilityReactNativePackage != null;
  const initializeCoreAliasConfig = null;
  const polyfillPreludeScripts = [];
  let compatibilityReactNativePath = void 0;
  if (isReactNativeVersionOverridden) {
    compatibilityReactNativePath = path5.dirname(__require.resolve(`${compatibilityReactNativePackage}/package.json`));
  }
  const { asyncStorageCompatPath } = ensureAsyncStorageCompatFile(options.rootDir);
  const dependencyAliasResolveCaches = /* @__PURE__ */ new Map();
  const dependencyAliasConfigs = [
    {
      from: ASYNC_STORAGE_PACKAGE_NAME,
      to: asyncStorageCompatPath,
      exact: true
    },
    ...Object.entries(compatibilityAliases).map(
      ([from, to]) => ({
        from,
        to: async (context) => {
          const subpath = context.args.path.slice(from.length);
          const lookupPath = `${to}${subpath}`;
          if (dependencyAliasResolveCaches.has(lookupPath)) {
            return dependencyAliasResolveCaches.get(lookupPath);
          }
          const result = await context.resolve(lookupPath, {
            kind: context.args.kind,
            namespace: context.args.namespace,
            pluginData: context.args.pluginData,
            resolveDir: __dirname,
            with: context.args.with
          });
          if (result.errors.length > 0 || result.path.length === 0) {
            const details = result.errors.map(({ text }) => text).join(" | ");
            throw new Error(
              [
                `[react-native-compatibility-plugin] Failed to resolve alias`,
                `from "${from}" to "${to}"`,
                `for "${context.args.path}"`,
                `(lookup: "${lookupPath}")`,
                details.length > 0 ? `errors: ${details}` : "errors: empty resolved path"
              ].join(" ")
            );
          }
          dependencyAliasResolveCaches.set(lookupPath, result.path);
          return result.path;
        }
      })
    )
  ];
  const metroDependencyAliasConfig = {
    "0.72.6": [
      {
        from: ASYNC_STORAGE_PACKAGE_NAME,
        to: asyncStorageCompatPath,
        exact: true
      },
      ...Object.entries(compatibilityAliases).map(([from, to]) => ({
        from,
        to,
        resolveDir: __dirname
      }))
    ],
    "0.84.0": [
      {
        from: ASYNC_STORAGE_PACKAGE_NAME,
        to: asyncStorageCompatPath,
        exact: true
      }
    ]
  };
  return {
    name: "react-native-compatibility-plugin",
    config: {
      reactNativePath: compatibilityReactNativePath,
      resolver: {
        alias: [initializeCoreAliasConfig, ...dependencyAliasConfigs].filter(isNotNil)
      },
      esbuild: {
        prelude: polyfillPreludeScripts
      },
      metro: {
        resolver: {
          resolveRequest: createCustomResolver(options.rootDir, {
            alias: metroDependencyAliasConfig[target]
          })
        }
      }
    }
  };
}

// src/reactNativeReverseCompatibilityPlugin.ts
import fs2 from "fs";
import path6 from "path";

// src/shims/native-modules-compat.ts
var MINI_APP_CONSTANT_KEYS = ["groupId", "operationalEnvironment", "tossAppVersion", "deviceId"];
var MINI_APP_EVENT_NAMES = [
  "updateLocation",
  "visibilityChangedByTransparentServiceWeb",
  "appBridgeCallback"
];
var EXCEPT_METHODS = [
  "loadAppsInTossAdmob",
  "showAppsInTossAdmob",
  "getCachedStatusAppsInTossAdmob",
  "loadTossAdOrAdmob",
  "showTossAdOrAdmob",
  "fetchTossAd",
  // RN84에서는 모든 메시지가 params가 하나로 들어갑니다.
  // 아래 있는 것들은 RN72에서 params.params 형식으로 들어가던 메소드들이라 예외처리가 필요합니다.
  "getTossShareLink",
  "appsInTossSignTossCert",
  "checkoutPayment",
  "completeProductGrant",
  "grantPromotionRewardForGame",
  "shareWithScheme"
];
function isRecord(value) {
  return value != null && typeof value === "object";
}
function getReactNative(globalRef) {
  function getReactNativeFromSharedOrProxy() {
    if (isRecord(globalRef.__reactNativeProxy)) {
      return globalRef.__reactNativeProxy;
    }
    const sharedReactNative = globalRef.__MICRO_FRONTEND__?.__SHARED__?.["react-native"]?.get?.();
    if (isRecord(sharedReactNative)) {
      return sharedReactNative;
    }
    return void 0;
  }
  function getReactNativeFromNativeModuleProxy() {
    const nativeModuleProxy = isRecord(globalRef.nativeModuleProxy) ? globalRef.nativeModuleProxy : void 0;
    const turboModuleProxy = typeof globalRef.__turboModuleProxy === "function" ? globalRef.__turboModuleProxy : void 0;
    if (nativeModuleProxy != null || turboModuleProxy != null) {
      const turboRegistry = {
        get(name) {
          if (nativeModuleProxy?.[name] != null) {
            return nativeModuleProxy[name];
          }
          if (turboModuleProxy != null) {
            try {
              return turboModuleProxy(name);
            } catch {
              return void 0;
            }
          }
          return void 0;
        },
        getEnforcing(name) {
          const moduleRef = turboRegistry.get?.(name);
          if (moduleRef != null) {
            return moduleRef;
          }
          throw new Error(
            `[react-native-reverse-compatibility-plugin] TurboModule "${name}" is not available through nativeModuleProxy`
          );
        }
      };
      return {
        NativeModules: nativeModuleProxy,
        TurboModuleRegistry: turboRegistry
      };
    }
    return void 0;
  }
  if (isRecord(__appsInTossReverseCompatPreloadedReactNative)) {
    return __appsInTossReverseCompatPreloadedReactNative;
  }
  if (__appsInTossReverseCompatIsHost) {
    const reactNativeFromNativeModuleProxy = getReactNativeFromNativeModuleProxy();
    if (reactNativeFromNativeModuleProxy != null) {
      return reactNativeFromNativeModuleProxy;
    }
    const reactNativeFromSharedOrProxy = getReactNativeFromSharedOrProxy();
    if (reactNativeFromSharedOrProxy != null) {
      return reactNativeFromSharedOrProxy;
    }
  } else {
    const reactNativeFromSharedOrProxy = getReactNativeFromSharedOrProxy();
    if (reactNativeFromSharedOrProxy != null) {
      return reactNativeFromSharedOrProxy;
    }
    const reactNativeFromNativeModuleProxy = getReactNativeFromNativeModuleProxy();
    if (reactNativeFromNativeModuleProxy != null) {
      return reactNativeFromNativeModuleProxy;
    }
  }
  throw new Error(
    `[react-native-reverse-compatibility-plugin] cannot resolve react-native runtime (isHost=${__appsInTossReverseCompatIsHost})`
  );
}
function getTurboModule(reactNative, name) {
  const registry = reactNative.TurboModuleRegistry;
  if (!isRecord(registry)) {
    return void 0;
  }
  if (typeof registry.get === "function") {
    const mod = registry.get(name);
    if (mod != null) {
      return mod;
    }
  }
  if (typeof registry.getEnforcing === "function") {
    try {
      const mod = registry.getEnforcing(name);
      if (mod != null) {
        return mod;
      }
    } catch {
    }
  }
  return void 0;
}
function getNativeModule(reactNative, names) {
  const nativeModules = isRecord(reactNative.NativeModules) ? reactNative.NativeModules : void 0;
  for (const name of names) {
    const mod = nativeModules?.[name] ?? getTurboModule(reactNative, name);
    if (mod != null) {
      return mod;
    }
  }
  return void 0;
}
function getRequiredNativeModule(reactNative, names, label) {
  const mod = getNativeModule(reactNative, names);
  if (mod == null) {
    throw new Error(
      `[react-native-reverse-compatibility-plugin] ${label} is not available in NativeModules/TurboModuleRegistry`
    );
  }
  return mod;
}
function extractAppBridgePayload(input) {
  if (!isRecord(input)) {
    return null;
  }
  const keys = Object.keys(input);
  const hasParams = keys.includes("params");
  const hasCallbacks = keys.includes("callbacks");
  if (!hasParams && !hasCallbacks) {
    return null;
  }
  const isLegacyEnvelope = keys.every((key) => key === "params" || key === "callbacks");
  if (!isLegacyEnvelope) {
    return null;
  }
  const payload = input;
  return {
    params: payload.params ?? {},
    callbacks: payload.callbacks ?? {}
  };
}
function getMiniAppConstants(moduleRef) {
  if (typeof moduleRef.getConstants === "function") {
    const constants = moduleRef.getConstants();
    return isRecord(constants) ? constants : {};
  }
  return MINI_APP_CONSTANT_KEYS.reduce((prev, key) => {
    if (key in moduleRef) {
      prev[key] = moduleRef[key];
    }
    return prev;
  }, {});
}
function callMiniAppMethod(moduleRef, methodName, params, callbacks) {
  const payload = extractAppBridgePayload(params);
  const payloadParams = payload?.params ?? params;
  const payloadCallbacks = callbacks != null && isRecord(callbacks) ? { ...isRecord(payload?.callbacks) ? payload.callbacks : {}, ...callbacks } : payload?.callbacks ?? {};
  const directMethod = moduleRef[methodName];
  if (typeof directMethod !== "function") {
    throw new Error(
      `[react-native-reverse-compatibility-plugin] MiniAppModule.${methodName} is not available in AppsInTossModule`
    );
  }
  const hasCallbacks = isRecord(payloadCallbacks) && Object.keys(payloadCallbacks).length > 0;
  const isAppBridgePayload = payload != null;
  if (hasCallbacks || isAppBridgePayload || EXCEPT_METHODS.includes(methodName)) {
    return directMethod({
      params: payloadParams ?? {},
      callbacks: payloadCallbacks
    });
  }
  return directMethod(payloadParams);
}
function createNoopSubscription() {
  return {
    remove() {
    }
  };
}
function getMiniAppEventEmitter(reactNative, moduleRef) {
  const nativeEventEmitterCtor = reactNative.NativeEventEmitter;
  if (typeof nativeEventEmitterCtor === "function") {
    try {
      const emitter = new nativeEventEmitterCtor(moduleRef);
      if (isRecord(emitter) && typeof emitter.addListener === "function") {
        return emitter;
      }
    } catch {
    }
  }
  const deviceEventEmitter = reactNative.DeviceEventEmitter;
  if (isRecord(deviceEventEmitter) && typeof deviceEventEmitter.addListener === "function") {
    return deviceEventEmitter;
  }
  return null;
}
function createOnSendEventCompat(moduleRef, reactNative) {
  return (listener) => {
    if (typeof listener !== "function") {
      return createNoopSubscription();
    }
    const onSendEvent = moduleRef.onSendEvent;
    if (typeof onSendEvent === "function") {
      const subscription = onSendEvent(listener);
      if (isRecord(subscription) && typeof subscription.remove === "function") {
        return subscription;
      }
      return createNoopSubscription();
    }
    const emitter = getMiniAppEventEmitter(reactNative, moduleRef);
    if (emitter == null) {
      return createNoopSubscription();
    }
    const subscriptions = [];
    for (const eventName of MINI_APP_EVENT_NAMES) {
      try {
        const subscription = emitter.addListener(eventName, (body) => {
          listener({
            eventName,
            body
          });
        });
        if (isRecord(subscription) && typeof subscription.remove === "function") {
          subscriptions.push(subscription);
        }
      } catch {
      }
    }
    return {
      remove() {
        for (const subscription of subscriptions) {
          try {
            subscription.remove?.();
          } catch {
          }
        }
      }
    };
  };
}
function createMiniAppModuleCompat(moduleRef, reactNative) {
  const onSendEventCompat = createOnSendEventCompat(moduleRef, reactNative);
  return new Proxy(moduleRef, {
    get(target, prop) {
      if (prop === "moduleName") {
        return "MiniAppModule";
      }
      if (prop === "getConstants") {
        return () => getMiniAppConstants(target);
      }
      if (prop === "postMessage") {
        return (methodName, params, callbacks) => {
          if (typeof methodName !== "string" || methodName.length === 0) {
            throw new Error(
              "[react-native-reverse-compatibility-plugin] MiniAppModule.postMessage requires methodName"
            );
          }
          return Promise.resolve(callMiniAppMethod(target, methodName, params, callbacks));
        };
      }
      if (prop === "postMessageSync") {
        return (methodName, params, callbacks) => {
          if (typeof methodName !== "string" || methodName.length === 0) {
            throw new Error(
              "[react-native-reverse-compatibility-plugin] MiniAppModule.postMessageSync requires methodName"
            );
          }
          return callMiniAppMethod(target, methodName, params, callbacks);
        };
      }
      if (prop === "onSendEvent") {
        return onSendEventCompat;
      }
      if (typeof prop === "string" && MINI_APP_CONSTANT_KEYS.includes(prop)) {
        const constants = getMiniAppConstants(target);
        return constants[prop];
      }
      if (typeof prop === "string") {
        const current = target[prop];
        if (typeof current === "function") {
          return current.bind(target);
        }
        if (current !== void 0) {
          return current;
        }
        return (params) => callMiniAppMethod(target, prop, params);
      }
      return Reflect.get(target, prop);
    }
  });
}
function createBedrockModuleCompat(moduleRef) {
  return new Proxy(moduleRef, {
    get(target, prop) {
      if (prop === "moduleName") {
        return "CommonModule";
      }
      if (prop === "DeviceInfo") {
        if (typeof target.getConstants === "function") {
          const constants = target.getConstants();
          const locale = isRecord(constants) ? constants.locale : void 0;
          return { locale: typeof locale === "string" ? locale : "" };
        }
        return { locale: "" };
      }
      if (prop === "schemeUri") {
        if (typeof target.getConstants === "function") {
          const constants = target.getConstants();
          return isRecord(constants) ? constants.schemeUri : void 0;
        }
        return void 0;
      }
      const current = Reflect.get(target, prop);
      return typeof current === "function" ? current.bind(target) : current;
    }
  });
}
function createGraniteBrownfieldModuleCompat(moduleRef, reactNative) {
  const bedrockCompat = createBedrockModuleCompat(moduleRef);
  return new Proxy(bedrockCompat, {
    get(target, prop) {
      if (prop === "moduleName") {
        return "GraniteBrownfieldModule";
      }
      if (prop === "onVisibilityChanged") {
        return (listener) => {
          if (typeof listener !== "function") {
            return createNoopSubscription();
          }
          const coreModule = getNativeModule(reactNative, ["BedrockCoreModule", "GraniteCoreModule"]);
          if (coreModule == null) {
            return createNoopSubscription();
          }
          const emitter = getMiniAppEventEmitter(reactNative, coreModule);
          if (emitter == null) {
            return createNoopSubscription();
          }
          const subscription = emitter.addListener("visibilityChanged", (nextVisible) => {
            listener({ visible: nextVisible });
          });
          if (isRecord(subscription) && typeof subscription.remove === "function") {
            return subscription;
          }
          return createNoopSubscription();
        };
      }
      if (prop === "closeView") {
        const closeView = Reflect.get(target, prop);
        if (typeof closeView === "function") {
          return (...args) => Promise.resolve(closeView(...args));
        }
      }
      const current = Reflect.get(target, prop);
      return typeof current === "function" ? current.bind(target) : current;
    }
  });
}
function createTossCoreModuleCompat(reactNative) {
  const tossCoreModule = getRequiredNativeModule(reactNative, ["TossCoreModule"], "TossCoreModule");
  return new Proxy(tossCoreModule, {
    get(target, prop) {
      if (prop === "moduleName") {
        return "TossCoreModule";
      }
      if (prop === "eventLog") {
        return (params) => {
          return target[prop]({ params });
        };
      }
      return Reflect.get(target, prop);
    }
  });
}
function createBrickModuleCompat(globalRef) {
  const cache = /* @__PURE__ */ new Map();
  let reactNativeCache = null;
  const getReactNativeRuntime = () => {
    if (reactNativeCache != null) {
      return reactNativeCache;
    }
    reactNativeCache = getReactNative(globalRef);
    return reactNativeCache;
  };
  const getMiniAppModule = () => getRequiredNativeModule(getReactNativeRuntime(), ["AppsInTossModule", "MiniAppModule"], "AppsInTossModule");
  const moduleFactories = {
    CommonModule: () => createBedrockModuleCompat(
      getRequiredNativeModule(
        getReactNativeRuntime(),
        ["BedrockModule", "GraniteModule"],
        "BedrockModule/GraniteModule"
      )
    ),
    GraniteBrownfieldModule: () => createGraniteBrownfieldModuleCompat(
      getRequiredNativeModule(
        getReactNativeRuntime(),
        ["BedrockModule", "GraniteModule"],
        "BedrockModule/GraniteModule"
      ),
      getReactNativeRuntime()
    ),
    MiniAppModule: () => createMiniAppModuleCompat(getMiniAppModule(), getReactNativeRuntime()),
    TossCoreModule: () => createTossCoreModuleCompat(getReactNativeRuntime()),
    MiniAppBundleLoader: () => getRequiredNativeModule(
      getReactNativeRuntime(),
      ["BedrockCoreModule", "GraniteCoreModule"],
      "BedrockCoreModule/GraniteCoreModule"
    )
  };
  return {
    get(moduleName) {
      if (cache.has(moduleName)) {
        return cache.get(moduleName);
      }
      const factory = moduleFactories[moduleName];
      if (typeof factory !== "function") {
        throw new Error(
          `[react-native-reverse-compatibility-plugin] Unsupported BrickModule.get("${moduleName}") in RN72 target`
        );
      }
      const moduleInstance = factory();
      cache.set(moduleName, moduleInstance);
      return moduleInstance;
    },
    getRegisteredModules() {
      return Object.keys(moduleFactories);
    },
    clearCache() {
      cache.clear();
    }
  };
}
function applyReverseBrickModuleCompat(globalObject, moduleObject) {
  const globalRef = globalObject;
  const brickModule = createBrickModuleCompat(globalRef);
  moduleObject.exports = {
    __esModule: true,
    default: brickModule,
    BrickModule: brickModule,
    BrickError: Error,
    BrickModuleError: Error,
    Any: void 0,
    AnyObject: void 0
  };
}
function createReverseBrickModuleCompatContent({ isHost }) {
  return [
    `const __appsInTossReverseCompatIsHost = ${String(isHost)};`,
    `const __appsInTossReverseCompatPreloadedReactNative = (() => { try { return require('react-native'); } catch { return undefined; } })();`,
    `const MINI_APP_CONSTANT_KEYS = ${JSON.stringify([...MINI_APP_CONSTANT_KEYS])};`,
    `const MINI_APP_EVENT_NAMES = ${JSON.stringify([...MINI_APP_EVENT_NAMES])};`,
    `const EXCEPT_METHODS = ${JSON.stringify([...EXCEPT_METHODS])};`,
    isRecord.toString(),
    getReactNative.toString(),
    getTurboModule.toString(),
    getNativeModule.toString(),
    getRequiredNativeModule.toString(),
    extractAppBridgePayload.toString(),
    getMiniAppConstants.toString(),
    callMiniAppMethod.toString(),
    createNoopSubscription.toString(),
    getMiniAppEventEmitter.toString(),
    createOnSendEventCompat.toString(),
    createMiniAppModuleCompat.toString(),
    createBedrockModuleCompat.toString(),
    createGraniteBrownfieldModuleCompat.toString(),
    createTossCoreModuleCompat.toString(),
    createBrickModuleCompat.toString(),
    applyReverseBrickModuleCompat.toString(),
    `(applyReverseBrickModuleCompat)(typeof globalThis !== 'undefined' ? globalThis : typeof global !== 'undefined' ? global : typeof window !== 'undefined' ? window : this, module);`
  ].join("\n");
}

// src/reactNativeReverseCompatibilityPlugin.ts
var REVERSE_COMPATIBILITY_TARGET = "0.72.6";
var BRICK_MODULE_NAME = "brick-module";
var GRANITE_DIRECTORY_NAME2 = ".granite";
var BRICK_MODULE_COMPAT_FILE_NAME = ".ait-rn72-brick-module-compat.js";
function ensureReverseCompatFiles({ rootDir, isHost }) {
  const graniteDirectoryPath = path6.join(rootDir, GRANITE_DIRECTORY_NAME2);
  const brickModuleCompatPath = path6.join(graniteDirectoryPath, BRICK_MODULE_COMPAT_FILE_NAME);
  try {
    const reverseBrickModuleCompatContent = createReverseBrickModuleCompatContent({ isHost });
    fs2.mkdirSync(graniteDirectoryPath, { recursive: true });
    fs2.writeFileSync(brickModuleCompatPath, reverseBrickModuleCompatContent, "utf8");
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error);
    throw new Error(
      [
        "[react-native-reverse-compatibility-plugin] Failed to create reverse brick-module compatibility temp file.",
        `rootDir="${rootDir}"`,
        `brickModuleCompatPath="${brickModuleCompatPath}"`,
        `reason: ${reason}`
      ].join(" ")
    );
  }
  return {
    brickModuleCompatPath
  };
}
function reactNativeReverseCompatibilityPlugin(options) {
  if (options.target !== REVERSE_COMPATIBILITY_TARGET) {
    return {
      name: "react-native-reverse-compatibility-plugin",
      config: {}
    };
  }
  const isHost = options.isHost ?? false;
  const compatibilityAliases = REACT_NATIVE_COMPATIBILITIES[REVERSE_COMPATIBILITY_TARGET]?.dependencyAliases ?? {};
  const { asyncStorageCompatPath } = ensureAsyncStorageCompatFile(options.rootDir);
  const { brickModuleCompatPath } = ensureReverseCompatFiles({
    rootDir: options.rootDir,
    isHost
  });
  const brickAliasConfig = {
    from: BRICK_MODULE_NAME,
    to: brickModuleCompatPath,
    exact: true
  };
  const asyncStorageAliasConfig = {
    from: ASYNC_STORAGE_PACKAGE_NAME,
    to: asyncStorageCompatPath,
    exact: true
  };
  const metroAliasConfig = [
    ...Object.entries(compatibilityAliases).map(([from, to]) => ({
      from,
      to,
      resolveDir: __dirname
    })),
    {
      from: ASYNC_STORAGE_PACKAGE_NAME,
      to: asyncStorageCompatPath,
      exact: true
    },
    {
      from: BRICK_MODULE_NAME,
      to: brickModuleCompatPath,
      exact: true
    }
  ];
  return {
    name: "react-native-reverse-compatibility-plugin",
    config: {
      resolver: {
        alias: [brickAliasConfig, asyncStorageAliasConfig]
      },
      metro: {
        resolver: {
          resolveRequest: createCustomResolver(options.rootDir, {
            alias: metroAliasConfig
          })
        }
      }
    }
  };
}

// src/reactCompatibilityPlugin.ts
import fs3 from "fs";
import path7 from "path";
import { getLocalTempDirectoryPath } from "@granite-js/utils";
function getReactCompatibilityByReactNativeVersion(reactNativeVersion) {
  switch (reactNativeVersion) {
    case "0.72.6":
      return "18.2.0";
    case "0.84.0":
      return "19.2.3";
    default:
      throw new Error(`Unsupported React Native version: ${reactNativeVersion}`);
  }
}
function getReactCompatScript() {
  const reactUsePolyfillPath = __require.resolve("react18-use");
  const reactEffectEventPolyfillPath = __require.resolve("use-effect-event");
  return `
    const React = require('react');
    const reactUsePolyfill = require('${reactUsePolyfillPath}');
    const reactEffectEventPolyfill = require('${reactEffectEventPolyfillPath}');
  
    function useOptimisticPolyfill(passthroughState, reducer) {
      const [optimisticState, setOptimisticState] = React.useState(passthroughState);
      const lastPassthroughState = React.useRef(passthroughState);
  
      if (passthroughState !== lastPassthroughState.current) {
        setOptimisticState(passthroughState);
        lastPassthroughState.current = passthroughState;
      }
  
      function addOptimistic(action) {
        setOptimisticState((current) => reducer(current, action));
      }
  
      return [optimisticState, addOptimistic];
    }
  
    module.exports = Object.assign(React, {
      use: reactUsePolyfill.use,
      useEffectEvent: reactEffectEventPolyfill.useEffectEvent,
      useOptimistic: useOptimisticPolyfill,
    });
    `;
}
function reactCompatibilityPlugin(options) {
  const { rootDir, target } = options;
  if (getReactCompatibilityByReactNativeVersion(target) === "19.2.3") {
    return null;
  }
  const localDir = getLocalTempDirectoryPath(rootDir);
  const compatScriptPath = path7.join(localDir, "react-19-compat.js");
  const compatScript = getReactCompatScript();
  fs3.writeFileSync(compatScriptPath, compatScript);
  return {
    name: "react-compatibility-plugin",
    config: {
      esbuild: {
        prelude: [compatScriptPath]
      },
      metro: {
        serializer: {
          getPolyfills: () => [compatScriptPath]
        }
      }
    }
  };
}
export {
  DEFAULT_REACT_NATIVE_COMPATIBILITY,
  DEPRECATED_NATIVE_PACKAGES,
  REACT_NATIVE_COMPATIBILITIES,
  createCustomResolver,
  getReactCompatibilityByReactNativeVersion,
  getTossReactNativePackages,
  reactCompatibilityPlugin,
  reactNativeCompatibilityPlugin,
  reactNativeReverseCompatibilityPlugin
};
